<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0"> 
    <meta name="author" content="The BioJulia Organisation"> 
    <link rel="canonical" href="https://biojulia.github.io/Bio.jl/man/seq/search/"> 
    <link rel="shortcut icon" href="../../../img/favicon.ico">

    <title>Searching - Bio.jl</title>

    <link href="../../../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../../../css/base.css" rel="stylesheet">
    <link href="../../../css/biojulia.css" rel="stylesheet">
    <link rel="stylesheet" href="../../../css/highlight.css">
    <link rel="stylesheet" href="../../../css/friendly.css" type="text/css">
    <link href="../../../assets/Documenter.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="../../..">Bio.jl</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../../..">Home</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Manual <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../../frontpage/">Front Page</a>
</li>

                        
                            
<li >
    <a href="../../install/">Installation</a>
</li>

                        
                            
<li >
    <a href="../../reading/">IO API</a>
</li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Biological Sequences</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../module/">Overview</a>
</li>

        
            
<li >
    <a href="../symbols/">Biological Symbols</a>
</li>

        
            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Sequence types</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../sequences/sequences/">Overview</a>
</li>

        
            
<li >
    <a href="../sequences/bioseq/">BioSequence</a>
</li>

        
            
<li >
    <a href="../sequences/refseq/">Reference Sequences</a>
</li>

        
            
<li >
    <a href="../sequences/kmer/">Nucleic acid k-mers</a>
</li>

        
    </ul>
  </li>

        
            
<li class="active">
    <a href="./">Searching</a>
</li>

        
            
<li >
    <a href="../seqrecords/">Sequence Records</a>
</li>

        
            
<li >
    <a href="../demultiplexer/">Demultiplexing</a>
</li>

        
    </ul>
  </li>

                        
                            
<li >
    <a href="../../alignments/">Alignments</a>
</li>

                        
                            
<li >
    <a href="../../intervals/">Intervals</a>
</li>

                        
                            
<li >
    <a href="../../var/">Genetic Variation</a>
</li>

                        
                            
<li >
    <a href="../../phylo/">Phylogenetic Trees</a>
</li>

                        
                            
<li >
    <a href="../../structure/">Structure</a>
</li>

                        
                            
<li >
    <a href="../../services/">Services</a>
</li>

                        
                            
<li >
    <a href="../../util/">Util</a>
</li>

                        
                            
<li >
    <a href="../../tools/">Tools</a>
</li>

                        
                            
<li >
    <a href="../../">Index</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li >
                        <a href="../../contributing/">Contributing</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../sequences/kmer/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../seqrecords/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/BioJulia/Bio.jl">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
                <li><a href="https://biojulia.github.io">BioJulia Home</a></li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="first-level active"><a href="#sequence-search">Sequence search</a></li>
        
            <li class="second-level"><a href="#exact-search">Exact search</a></li>
            
        
            <li class="second-level"><a href="#approximate-search">Approximate search</a></li>
            
        
            <li class="second-level"><a href="#regular-expression-search">Regular expression search</a></li>
            
        
            <li class="second-level"><a href="#sequence-composition">Sequence composition</a></li>
            
        
    
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<p><a id='Sequence-search-1'></a></p>
<h1 id="sequence-search">Sequence search</h1>
<p>Three kinds of on-line search functions are provided:</p>
<ol>
<li>Exact search</li>
<li>Approximate search</li>
<li>Regular expression search</li>
</ol>
<p>These are all specialized for biological sequences and ambiguities of symbols are considered.</p>
<p><a id='Exact-search-1'></a></p>
<h2 id="exact-search">Exact search</h2>
<p>Exact search functions search for an occurrence of the query symbol or sequence. Four functions, <code>search</code>, <code>searchindex</code>, <code>rsearch</code>, and <code>rsearchindex</code> are available:</p>
<div class="codehilite"><pre><span></span>julia&gt; seq = dna&quot;ACAGCGTAGCT&quot;;

julia&gt; search(seq, DNA_G)  # search a query symbol
4:4

julia&gt; query = dna&quot;AGC&quot;;

julia&gt; search(seq, query)  # search a query sequence
3:5

julia&gt; searchindex(seq, query)
3

julia&gt; rsearch(seq, query)  # similar to `search` but in the reverse direction
8:10

julia&gt; rsearchindex(seq, query)  # similar to `searchindex` but in the reverse direction
8
</pre></div>


<p>These search functions take ambiguous symbols into account. That is, if two symbols are compatible (e.g. <code>DNA_A</code> and <code>DNA_N</code>), they match when searching an occurrence. In the following example, 'N' is a wild card that matches any symbols:</p>
<div class="codehilite"><pre><span></span>julia&gt; search(dna&quot;ACNT&quot;, DNA_N)  # &#39;A&#39; matches &#39;N&#39;
1:1

julia&gt; search(dna&quot;ACNT&quot;, dna&quot;CGT&quot;)  # &#39;N&#39; matches &#39;G&#39;
2:4

julia&gt; search(dna&quot;ACGT&quot;, dna&quot;CNT&quot;)  # &#39;G&#39; matches &#39;N&#39;
2:4
</pre></div>


<p>The exact sequence search needs preprocessing phase of query sequence before searching phase. This would be enough fast for most search applications. But when searching a query sequence to large amounts of target sequences, caching the result of preprocessing may save time. The <code>ExactSearchQuery</code> creates such a preprocessed query object and is applicable to the search functions:</p>
<div class="codehilite"><pre><span></span>julia&gt; query = ExactSearchQuery(dna&quot;ATT&quot;);

julia&gt; search(dna&quot;ATTTATT&quot;, query)
1:3

julia&gt; rsearch(dna&quot;ATTTATT&quot;, query)
5:7
</pre></div>


<p><a id='Approximate-search-1'></a></p>
<h2 id="approximate-search">Approximate search</h2>
<p>The approximate search is similar to the exact search but allows a specific number of errors. That is, it tries to find a subsequence of the target sequence within a specific <a href="https://en.wikipedia.org/wiki/Levenshtein_distance">Levenshtein distance</a> of the query sequence:</p>
<div class="codehilite"><pre><span></span><span class="s s-Atom">julia</span><span class="o">&gt;</span> <span class="s s-Atom">seq</span> <span class="o">=</span> <span class="s s-Atom">dna</span><span class="s2">&quot;ACAGCGTAGCT&quot;</span><span class="p">;</span>

<span class="s s-Atom">julia</span><span class="o">&gt;</span> <span class="nf">approxsearch</span><span class="p">(</span><span class="s s-Atom">seq</span><span class="p">,</span> <span class="s s-Atom">dna</span><span class="s2">&quot;AGGG&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>  <span class="s s-Atom">#</span> <span class="s s-Atom">nothing</span> <span class="s s-Atom">matches</span> <span class="s s-Atom">with</span> <span class="s s-Atom">no</span> <span class="s s-Atom">errors</span>
<span class="mi">0</span><span class="p">:-</span><span class="mi">1</span>

<span class="s s-Atom">julia</span><span class="o">&gt;</span> <span class="nf">approxsearch</span><span class="p">(</span><span class="s s-Atom">seq</span><span class="p">,</span> <span class="s s-Atom">dna</span><span class="s2">&quot;AGGG&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>  <span class="s s-Atom">#</span> <span class="s s-Atom">seq</span><span class="p">[</span><span class="mi">3</span><span class="s s-Atom">:</span><span class="mi">5</span><span class="p">]</span> <span class="s s-Atom">matches</span> <span class="s s-Atom">with</span> <span class="s s-Atom">one</span> <span class="s s-Atom">error</span>
<span class="mi">3</span><span class="s s-Atom">:</span><span class="mi">6</span>

<span class="s s-Atom">julia</span><span class="o">&gt;</span> <span class="nf">approxsearch</span><span class="p">(</span><span class="s s-Atom">seq</span><span class="p">,</span> <span class="s s-Atom">dna</span><span class="s2">&quot;AGGG&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>  <span class="s s-Atom">#</span> <span class="s s-Atom">seq</span><span class="p">[</span><span class="mi">1</span><span class="s s-Atom">:</span><span class="mi">4</span><span class="p">]</span> <span class="s s-Atom">matches</span> <span class="s s-Atom">with</span> <span class="s s-Atom">two</span> <span class="s s-Atom">errors</span>
<span class="mi">1</span><span class="s s-Atom">:</span><span class="mi">4</span>
</pre></div>


<p>Like the exact search functions, four kinds of functions (<code>approxsearch</code>, <code>approxsearchindex</code>, <code>approxrsearch</code>, and <code>approxrsearchindex</code>) are available:</p>
<div class="codehilite"><pre><span></span>julia&gt; seq = dna&quot;ACAGCGTAGCT&quot;; pat = dna&quot;AGGG&quot;;

julia&gt; approxsearch(seq, pat, 2)        # return the range (forward)
1:4

julia&gt; approxsearchindex(seq, pat, 2)   # return the starting index (forward)
1

julia&gt; approxrsearch(seq, pat, 2)       # return the range (backward)
8:11

julia&gt; approxrsearchindex(seq, pat, 2)  # return the starting index (backward)
8
</pre></div>


<p>Preprocessing can be cached in an <code>ApproximateSearchQuery</code> object:</p>
<div class="codehilite"><pre><span></span>julia&gt; query = ApproximateSearchQuery(dna&quot;AGGG&quot;);

julia&gt; approxsearch(dna&quot;AAGAGG&quot;, query, 1)
2:5

julia&gt; approxsearch(dna&quot;ACTACGT&quot;, query, 2)
4:6
</pre></div>


<p><a id='Regular-expression-search-1'></a></p>
<h2 id="regular-expression-search">Regular expression search</h2>
<p>Query patterns can be described in regular expressions. The syntax supports a subset of Perl and PROSITE's notation.</p>
<p>The Perl-like syntax starts with <code>biore</code> (<strong>bio</strong>logical <strong>re</strong>gular expression) and ends with a symbol option: "dna", "rna" or "aa". For example, <code>biore"A+"dna</code> is a regular expression for DNA sequences and <code>biore"A+"aa</code> is for amino acid sequences. The symbol options can be abbreviated to its first character: "d", "r" or "a", respectively.</p>
<p>Here are examples of using the regular expression for <code>BioSequence</code>s:</p>
<div class="codehilite"><pre><span></span>julia&gt; match(biore&quot;A+C*&quot;dna, dna&quot;AAAACC&quot;)
Nullable{Bio.Seq.RE.RegexMatch{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}}}}(RegexMatch(&quot;AAAACC&quot;))

julia&gt; match(biore&quot;A+C*&quot;d, dna&quot;AAAACC&quot;)
Nullable{Bio.Seq.RE.RegexMatch{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}}}}(RegexMatch(&quot;AAAACC&quot;))

julia&gt; ismatch(biore&quot;A+C*&quot;dna, dna&quot;AAC&quot;)
true

julia&gt; ismatch(biore&quot;A+C*&quot;dna, dna&quot;C&quot;)
false
</pre></div>


<p><code>match</code> always returns a <code>Nullable</code> object and it should be null if no match is found.</p>
<p>The table below summarizes available syntax elements.</p>
<table>
<thead>
<tr>
<th align="center">Syntax</th>
<th align="left">Description</th>
<th align="left">Example</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><code>\|</code></td>
<td align="left">alternation</td>
<td align="left"><code>"A\|T"</code> matches <code>"A"</code> and <code>"T"</code></td>
</tr>
<tr>
<td align="center"><code>*</code></td>
<td align="left">zero or more times repeat</td>
<td align="left"><code>"TA*"</code> matches <code>"T"</code>, <code>"TA"</code> and <code>"TAA"</code></td>
</tr>
<tr>
<td align="center"><code>+</code></td>
<td align="left">one or more times repeat</td>
<td align="left"><code>"TA+"</code> matches <code>"TA"</code> and <code>"TAA"</code></td>
</tr>
<tr>
<td align="center"><code>?</code></td>
<td align="left">zero or one time</td>
<td align="left"><code>"TA?"</code> matches <code>"T"</code> and <code>"TA"</code></td>
</tr>
<tr>
<td align="center"><code>{n,}</code></td>
<td align="left"><code>n</code> or more times repeat</td>
<td align="left"><code>"A{3,}"</code> matches <code>"AAA"</code> and <code>"AAAA"</code></td>
</tr>
<tr>
<td align="center"><code>{n,m}</code></td>
<td align="left"><code>n</code>-<code>m</code> times repeat</td>
<td align="left"><code>"A{3,5}"</code> matches <code>"AAA"</code>, <code>"AAAA"</code> and <code>"AAAAA"</code></td>
</tr>
<tr>
<td align="center"><code>^</code></td>
<td align="left">the start of the sequence</td>
<td align="left"><code>"^TAN*"</code> matches <code>"TATGT"</code></td>
</tr>
<tr>
<td align="center"><code>$</code></td>
<td align="left">the end of the sequence</td>
<td align="left"><code>"N*TA$"</code> matches <code>"GCTA"</code></td>
</tr>
<tr>
<td align="center"><code>(...)</code></td>
<td align="left">pattern grouping</td>
<td align="left"><code>"(TA)+"</code> matches <code>"TA"</code> and <code>"TATA"</code></td>
</tr>
<tr>
<td align="center"><code>[...]</code></td>
<td align="left">one of symbols</td>
<td align="left"><code>"[ACG]+"</code> matches <code>"AGGC"</code></td>
</tr>
</tbody>
</table>
<p><code>eachmatch</code>, <code>matchall</code>, and <code>search</code> are also defined like usual strings:</p>
<div class="codehilite"><pre><span></span>julia&gt; matchall(biore&quot;TATA*?&quot;d, dna&quot;TATTATAATTA&quot;)  # overlap (default)
4-element Array{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}},1}:
 TAT  
 TAT  
 TATA
 TATAA

julia&gt; matchall(biore&quot;TATA*&quot;d, dna&quot;TATTATAATTA&quot;, false)  # no overlap
2-element Array{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}},1}:
 TAT  
 TATAA

julia&gt; search(dna&quot;TATTATAATTA&quot;, biore&quot;TATA*&quot;d)
1:3

julia&gt; search(dna&quot;TATTATAATTA&quot;, biore&quot;TATA*&quot;d, 2)
4:8
</pre></div>


<p>Notewothy differences from strings are:</p>
<ul>
<li>Ambiguous characters match any compatible characters (e.g. <code>biore"N"d</code> is equivalent to <code>biore"[ACGT]"d</code>).</li>
<li>Whitespaces are ignored (e.g. <code>biore"A C G"d</code> is equivalent to <code>biore"ACG"d</code>).</li>
</ul>
<p>The PROSITE notation is described in <a href="http://prosite.expasy.org/scanprosite/scanprosite_doc.html">ScanProsite - user manual</a>. The syntax supports almost all notations including the extended syntax. The PROSITE notation starts with <code>prosite</code> prefix and no symbol option is needed because it always describes patterns of amino acid sequences:</p>
<div class="codehilite"><pre><span></span>julia&gt; match(prosite&quot;[AC]-x-V-x(4)-{ED}&quot;, aa&quot;CPVPQARG&quot;)
Nullable{Bio.Seq.RE.RegexMatch{Bio.Seq.BioSequence{Bio.Seq.AminoAcidAlphabet}}}(RegexMatch(&quot;CPVPQARG&quot;))

julia&gt; match(prosite&quot;[AC]xVx(4){ED}&quot;, aa&quot;CPVPQARG&quot;)
Nullable{Bio.Seq.RE.RegexMatch{Bio.Seq.BioSequence{Bio.Seq.AminoAcidAlphabet}}}(RegexMatch(&quot;CPVPQARG&quot;))
</pre></div>


<p><a id='Sequence-composition-1'></a></p>
<h2 id="sequence-composition">Sequence composition</h2>
<p>Sequence composition can be easily calculated using the <code>composition</code> function:</p>
<div class="codehilite"><pre><span></span>julia&gt; comp = composition(dna&quot;ACGAG&quot;)
DNA Composition:
  DNA_Gap =&gt; 0
  DNA_A   =&gt; 2
  DNA_C   =&gt; 1
  DNA_M   =&gt; 0
  DNA_G   =&gt; 2
  DNA_R   =&gt; 0
  DNA_S   =&gt; 0
  DNA_V   =&gt; 0
  DNA_T   =&gt; 0
  DNA_W   =&gt; 0
  DNA_Y   =&gt; 0
  DNA_H   =&gt; 0
  DNA_K   =&gt; 0
  DNA_D   =&gt; 0
  DNA_B   =&gt; 0
  DNA_N   =&gt; 0

julia&gt; comp[DNA_A]
2

julia&gt; comp[DNA_T]
0
</pre></div>


<p>To accumulate composition statistics of multiple sequences, <code>merge!</code> can be used as follows:</p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="c"># initiaize an empty composition counter</span>
       <span class="n">comp</span> <span class="o">=</span> <span class="n">composition</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;&quot;</span><span class="p">);</span>
<span class="n">ERROR</span><span class="o">:</span> <span class="kt">UndefVarError</span><span class="o">:</span> <span class="nd">@dna_str</span> <span class="n">not</span> <span class="n">defined</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="c"># iterate over sequences and accumulate composition statistics into `comp`</span>
       <span class="k">for</span> <span class="n">seq</span> <span class="kp">in</span> <span class="n">seqs</span>
           <span class="n">merge!</span><span class="p">(</span><span class="n">comp</span><span class="p">,</span> <span class="n">composition</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
       <span class="k">end</span>
<span class="n">ERROR</span><span class="o">:</span> <span class="kt">UndefVarError</span><span class="o">:</span> <span class="n">seqs</span> <span class="n">not</span> <span class="n">defined</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="c"># or functional programming style in one line</span>
       <span class="n">foldl</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">merge</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">composition</span><span class="p">(</span><span class="n">y</span><span class="p">)),</span> <span class="n">composition</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;&quot;</span><span class="p">),</span> <span class="n">seqs</span><span class="p">)</span>
<span class="n">ERROR</span><span class="o">:</span> <span class="kt">UndefVarError</span><span class="o">:</span> <span class="nd">@dna_str</span> <span class="n">not</span> <span class="n">defined</span>
</pre></div>


<p><code>composition</code> is also applicable to a <em>k</em>-mer iterator:</p>
<div class="codehilite"><pre><span></span>julia&gt; comp = composition(each(DNAKmer{4}, dna&quot;ACGT&quot;^100));

julia&gt; comp[DNAKmer(&quot;ACGT&quot;)]
100

julia&gt; comp[DNAKmer(&quot;CGTA&quot;)]
99
</pre></div></div>
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Copyright (c) 2016 The BioJulia Organisation<br></small>
        
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../../../js/jquery-1.10.2.min.js"></script>
    <script src="../../../js/bootstrap-3.0.3.min.js"></script>
    <script src="../../../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '../../..';
    </script>
    <script data-main="../../../mkdocs/js/search.js" src="../../../mkdocs/js/require.js"></script>
    <script src="../../../js/base.js"></script>
    <script>
        ((window.gitter = {}).chat = {}).options = {
            room: 'BioJulia/Bio.jl'
        };
    </script>
    <script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <script src="../../../assets/mathjaxhelper.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
