<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0"> 
    <meta name="author" content="The BioJulia Organisation"> 
    <link rel="canonical" href="https://biojulia.github.io/Bio.jl/man/structure/"> 
    <link rel="shortcut icon" href="../../img/favicon.ico">

    <title>Structure - Bio.jl</title>

    <link href="../../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../../css/base.css" rel="stylesheet">
    <link href="../../css/biojulia.css" rel="stylesheet">
    <link rel="stylesheet" href="../../css/highlight.css">
    <link rel="stylesheet" href="../../css/friendly.css" type="text/css">
    <link href="../../assets/Documenter.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="../..">Bio.jl</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../..">Home</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Manual <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../install/">Installation</a>
</li>

                        
                            
<li >
    <a href="../reading/">IO API</a>
</li>

                        
                            
<li >
    <a href="../seq/">Biological Sequences</a>
</li>

                        
                            
<li >
    <a href="../alignments/">Alignments</a>
</li>

                        
                            
<li >
    <a href="../intervals/">Intervals</a>
</li>

                        
                            
<li >
    <a href="../var/">Biological Variation</a>
</li>

                        
                            
<li >
    <a href="../phylo/">Phylogenetic Trees</a>
</li>

                        
                            
<li class="active">
    <a href="./">Structure</a>
</li>

                        
                            
<li >
    <a href="../util/">Util</a>
</li>

                        
                            
<li >
    <a href="../tools/">Tools</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Reference <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../../lib/seq/">Seq</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li >
                        <a href="../contributing/">Contributing</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../phylo/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../util/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/BioJulia/Bio.jl">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="first-level active"><a href="#biostructure-macromolecular-structures">Bio.Structure: Macromolecular Structures</a></li>
        
            <li class="second-level"><a href="#parsing-pdb-files">Parsing PDB files</a></li>
            
        
            <li class="second-level"><a href="#manipulating-structures">Manipulating structures</a></li>
            
        
            <li class="second-level"><a href="#writing-pdb-files">Writing PDB files</a></li>
            
        
            <li class="second-level"><a href="#spatial-calculations">Spatial calculations</a></li>
            
        
            <li class="second-level"><a href="#examples">Examples</a></li>
            
        
    
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<p><a id='Bio.Structure:-Macromolecular-Structures-1'></a></p>
<h1 id="biostructure-macromolecular-structures">Bio.Structure: Macromolecular Structures</h1>
<p>The <code>Bio.Structure</code> module provides functionality to manipulate macromolecular structures, and in particular to read and write <a href="http://www.rcsb.org/pdb/home/home.do">Protein Data Bank</a> (PDB) files. It is designed to be used for standard structural analysis tasks, as well as acting as a platform on which others can build to create more specific tools.</p>
<p><a id='Parsing-PDB-files-1'></a></p>
<h2 id="parsing-pdb-files">Parsing PDB files</h2>
<p>To download a PDB file:</p>
<div class="codehilite"><pre><span></span><span class="n">downloadpdb</span><span class="p">(</span><span class="s">&quot;1EN2&quot;</span><span class="p">)</span>
</pre></div>


<p>To parse a PDB file into a Structure-Model-Chain-Residue-Atom framework:</p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">struc</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="n">filepath_1EN2</span><span class="p">,</span> <span class="n">PDB</span><span class="p">)</span>
<span class="n">Bio</span><span class="o">.</span><span class="n">Structure</span><span class="o">.</span><span class="n">ProteinStructure</span>
<span class="n">Name</span>                        <span class="o">-</span>  <span class="mi">1</span><span class="n">EN2</span><span class="o">.</span><span class="n">pdb</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">models</span>            <span class="o">-</span>  <span class="mi">1</span>
<span class="n">Chain</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>                    <span class="o">-</span>  <span class="n">A</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">residues</span>          <span class="o">-</span>  <span class="mi">85</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">point</span> <span class="n">mutations</span>   <span class="o">-</span>  <span class="mi">5</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">other</span> <span class="n">molecules</span>   <span class="o">-</span>  <span class="mi">5</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">water</span> <span class="n">molecules</span>   <span class="o">-</span>  <span class="mi">76</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">atoms</span>             <span class="o">-</span>  <span class="mi">614</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">hydrogens</span>         <span class="o">-</span>  <span class="mi">0</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">disordered</span> <span class="n">atoms</span>  <span class="o">-</span>  <span class="mi">27</span>
</pre></div>


<p>The elements of <code>struc</code> can be accessed as follows:</p>
<table>
<thead>
<tr>
<th align="left">Command</th>
<th align="left">Returns</th>
<th align="left">Return type</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>struc[1]</code></td>
<td align="left">Model 1</td>
<td align="left"><code>Model</code></td>
</tr>
<tr>
<td align="left"><code>struc[1]['A']</code></td>
<td align="left">Model 1, chain A</td>
<td align="left"><code>Chain</code></td>
</tr>
<tr>
<td align="left"><code>struc['A']</code></td>
<td align="left">The lowest model (model 1), chain A</td>
<td align="left"><code>Chain</code></td>
</tr>
<tr>
<td align="left"><code>struc['A']["50"]</code></td>
<td align="left">Model 1, chain A, residue 50</td>
<td align="left"><code>AbstractResidue</code></td>
</tr>
<tr>
<td align="left"><code>struc['A'][50]</code></td>
<td align="left">Shortcut to above if it is not a hetero residue and the insertion code is blank</td>
<td align="left"><code>AbstractResidue</code></td>
</tr>
<tr>
<td align="left"><code>struc['A']["H_90"]</code></td>
<td align="left">Model 1, chain A, hetero residue 90</td>
<td align="left"><code>AbstractResidue</code></td>
</tr>
<tr>
<td align="left"><code>struc['A'][50]["CA"]</code></td>
<td align="left">Model 1, chain A, residue 50, atom name CA</td>
<td align="left"><code>AbstractAtom</code></td>
</tr>
<tr>
<td align="left"><code>struc['A'][15]["CG"]['A']</code></td>
<td align="left">For disordered atoms, access a specific location</td>
<td align="left"><code>Atom</code></td>
</tr>
</tbody>
</table>
<p>Disordered atoms are stored in a <code>DisorderedAtom</code> container but calls fall back to the default atom, so disorder can be ignored if you are not interested in it.</p>
<p>Disordered residues (i.e. point mutations with different residue names) are stored in a <code>DisorderedResidue</code> container.</p>
<p>The idea is that disorder will only bother you if you want it to. See the <a href="http://biopython.org/wiki/The_Biopython_Structural_Bioinformatics_FAQ#How_is_disorder_handled.3F">Biopython discussion</a> for more.</p>
<p>Properties can be retrieved as follows:</p>
<table>
<thead>
<tr>
<th align="left">Function</th>
<th align="left">Returns</th>
<th align="left">Return type</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>serial</code></td>
<td align="left">Serial number of an atom</td>
<td align="left"><code>Int</code></td>
</tr>
<tr>
<td align="left"><code>atomname</code></td>
<td align="left">Name of an atom</td>
<td align="left"><code>String</code></td>
</tr>
<tr>
<td align="left"><code>altlocid</code></td>
<td align="left">Alternative location ID of an atom</td>
<td align="left"><code>Char</code></td>
</tr>
<tr>
<td align="left"><code>x</code></td>
<td align="left">x coordinate of an atom</td>
<td align="left"><code>Float64</code></td>
</tr>
<tr>
<td align="left"><code>y</code></td>
<td align="left">y coordinate of an atom</td>
<td align="left"><code>Float64</code></td>
</tr>
<tr>
<td align="left"><code>z</code></td>
<td align="left">z coordinate of an atom</td>
<td align="left"><code>Float64</code></td>
</tr>
<tr>
<td align="left"><code>coords</code></td>
<td align="left">coordinates of an atom</td>
<td align="left"><code>Array{Float64,1}</code></td>
</tr>
<tr>
<td align="left"><code>occupancy</code></td>
<td align="left">Occupancy of an atom (default is <code>1.0</code>)</td>
<td align="left"><code>Float64</code></td>
</tr>
<tr>
<td align="left"><code>tempfactor</code></td>
<td align="left">Temperature factor of an atom (default is <code>0.0</code>)</td>
<td align="left"><code>Float64</code></td>
</tr>
<tr>
<td align="left"><code>element</code></td>
<td align="left">Element of an atom (default is <code>"  "</code>)</td>
<td align="left"><code>String</code></td>
</tr>
<tr>
<td align="left"><code>charge</code></td>
<td align="left">Charge of an atom (default is <code>"  "</code>)</td>
<td align="left"><code>String</code></td>
</tr>
<tr>
<td align="left"><code>residue</code></td>
<td align="left">Residue an atom belongs to</td>
<td align="left"><code>Residue</code></td>
</tr>
<tr>
<td align="left"><code>ishetero</code></td>
<td align="left"><code>true</code> if the residue or atom is a hetero residue/atom</td>
<td align="left"><code>Bool</code></td>
</tr>
<tr>
<td align="left"><code>isdisorderedatom</code></td>
<td align="left"><code>true</code> if the atom is disordered</td>
<td align="left"><code>Bool</code></td>
</tr>
<tr>
<td align="left"><code>pdbline</code></td>
<td align="left">PDB ATOM/HETATM record for an atom</td>
<td align="left"><code>String</code></td>
</tr>
<tr>
<td align="left"><code>resname</code></td>
<td align="left">Residue name of a residue or atom</td>
<td align="left"><code>String</code></td>
</tr>
<tr>
<td align="left"><code>resnumber</code></td>
<td align="left">Residue number of a residue or atom</td>
<td align="left"><code>Int</code></td>
</tr>
<tr>
<td align="left"><code>inscode</code></td>
<td align="left">Insertion code of a residue or atom</td>
<td align="left"><code>Char</code></td>
</tr>
<tr>
<td align="left"><code>resid</code></td>
<td align="left">Residue ID of an atom or residue (<code>full=true</code> includes chain)</td>
<td align="left"><code>String</code></td>
</tr>
<tr>
<td align="left"><code>atomnames</code></td>
<td align="left">Atom names of the atoms in a residue, sorted by serial</td>
<td align="left"><code>Array{String,1}</code></td>
</tr>
<tr>
<td align="left"><code>atoms</code></td>
<td align="left">Dictionary of atoms in a residue</td>
<td align="left"><code>Dict{String, AbstractAtom}</code></td>
</tr>
<tr>
<td align="left"><code>isdisorderedres</code></td>
<td align="left"><code>true</code> if the residue has multiple residue names</td>
<td align="left"><code>Bool</code></td>
</tr>
<tr>
<td align="left"><code>disorderedres</code></td>
<td align="left">Access a particular residue name in a <code>DisorderedResidue</code></td>
<td align="left"><code>Residue</code></td>
</tr>
<tr>
<td align="left"><code>chain</code></td>
<td align="left">Chain a residue or atom belongs to</td>
<td align="left"><code>Chain</code></td>
</tr>
<tr>
<td align="left"><code>chainid</code></td>
<td align="left">Chain ID of a chain, residue or atom</td>
<td align="left"><code>Char</code></td>
</tr>
<tr>
<td align="left"><code>resids</code></td>
<td align="left">Sorted residue IDs in a chain</td>
<td align="left"><code>Array{String,1}</code></td>
</tr>
<tr>
<td align="left"><code>residues</code></td>
<td align="left">Dictionary of residues in a chain</td>
<td align="left"><code>Dict{String, AbstractResidue}</code></td>
</tr>
<tr>
<td align="left"><code>model</code></td>
<td align="left">Model a chain, residue or atom belongs to</td>
<td align="left"><code>Model</code></td>
</tr>
<tr>
<td align="left"><code>modelnumber</code></td>
<td align="left">Model number of a model, chain, residue or atom</td>
<td align="left"><code>Int</code></td>
</tr>
<tr>
<td align="left"><code>chainids</code></td>
<td align="left">Sorted chain IDs in a model or structure</td>
<td align="left"><code>Array{Char,1}</code></td>
</tr>
<tr>
<td align="left"><code>chains</code></td>
<td align="left">Dictionary of chains in a model or structure</td>
<td align="left"><code>Dict{Char, Chain}</code></td>
</tr>
<tr>
<td align="left"><code>structure</code></td>
<td align="left">Structure a model, chain, residue or atom belongs to</td>
<td align="left"><code>ProteinStructure</code></td>
</tr>
<tr>
<td align="left"><code>structurename</code></td>
<td align="left">Name of the structure an element belongs to</td>
<td align="left"><code>String</code></td>
</tr>
<tr>
<td align="left"><code>modelnumbers</code></td>
<td align="left">Sorted model numbers in a structure</td>
<td align="left"><code>Array{Int,1}</code></td>
</tr>
<tr>
<td align="left"><code>models</code></td>
<td align="left">Dictionary of models in a structure</td>
<td align="left"><code>Dict{Int, Model}</code></td>
</tr>
</tbody>
</table>
<p>The <code>strip</code> keyword argument determines whether surrounding whitespace is stripped for <code>atomname</code>, <code>element</code>, <code>charge</code>, <code>resname</code> and <code>atomnames</code> (default <code>true</code>).</p>
<p>The coordinates of an atom can be set using <code>x!</code>, <code>y!</code>, <code>z!</code> and <code>coords!</code>.</p>
<p><a id='Manipulating-structures-1'></a></p>
<h2 id="manipulating-structures">Manipulating structures</h2>
<p>Elements can be looped over to reveal the sub-elements in the correct order:</p>
<div class="codehilite"><pre><span></span><span class="k">for</span> <span class="n">mod</span> <span class="k">in</span> <span class="n">struc</span>
    <span class="k">for</span> <span class="n">ch</span> <span class="k">in</span> <span class="n">mod</span>
        <span class="k">for</span> <span class="n">res</span> <span class="k">in</span> <span class="n">ch</span>
            <span class="k">for</span> <span class="n">at</span> <span class="k">in</span> <span class="n">res</span>
                <span class="c"># Do something</span>
            <span class="k">end</span>
        <span class="k">end</span>
    <span class="k">end</span>
<span class="k">end</span>
</pre></div>


<p>Models are ordered numerically; chains are ordered by character, except the empty chain is last; residues are ordered by residue number and insertion code with hetero residues after standard residues; atoms are ordered by atom serial.</p>
<p><code>collect</code> can be used to get arrays of sub-elements. <code>collectatoms</code>, <code>collectresidues</code>, <code>collectchains</code> and <code>collectmodels</code> return arrays of a particular type from a structural element or element array.</p>
<p>Selectors are functions passed as additional arguments to these functions. Only elements that return <code>true</code> when passed to the selector are retained.</p>
<table>
<thead>
<tr>
<th align="left">Command</th>
<th align="left">Action</th>
<th align="left">Return type</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>collect(struc['A'][50])</code></td>
<td align="left">Collect the sub-elements of an element, e.g. atoms from a residue</td>
<td align="left"><code>Array{AbstractAtom,1}</code></td>
</tr>
<tr>
<td align="left"><code>collectresidues(struc)</code></td>
<td align="left">Collect the residues of an element</td>
<td align="left"><code>Array{AbstractResidue,1}</code></td>
</tr>
<tr>
<td align="left"><code>collectatoms(struc)</code></td>
<td align="left">Collect the atoms of an element</td>
<td align="left"><code>Array{AbstractAtom,1}</code></td>
</tr>
<tr>
<td align="left"><code>collectatoms(struc, calphaselector)</code></td>
<td align="left">Collect the C-alpha atoms of an element</td>
<td align="left"><code>Array{AbstractAtom,1}</code></td>
</tr>
<tr>
<td align="left"><code>collectatoms(struc, calphaselector, disorderselector)</code></td>
<td align="left">Collect the disordered C-alpha atoms of an element</td>
<td align="left"><code>Array{AbstractAtom,1}</code></td>
</tr>
</tbody>
</table>
<p>It is easy to define your own atom, residue, chain or model selectors. The below will collect all atoms with x coordinate less than 0:</p>
<div class="codehilite"><pre><span></span><span class="n">xselector</span><span class="p">(</span><span class="n">at</span><span class="p">::</span><span class="n">AbstractAtom</span><span class="p">)</span> <span class="o">=</span> <span class="n">x</span><span class="p">(</span><span class="n">at</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">0</span>
<span class="n">collectatoms</span><span class="p">(</span><span class="n">struc</span><span class="p">,</span> <span class="n">xselector</span><span class="p">)</span>
</pre></div>


<p>Alternatively, you can use an anonymous function:</p>
<div class="codehilite"><pre><span></span><span class="n">collectatoms</span><span class="p">(</span><span class="n">struc</span><span class="p">,</span> <span class="n">at</span> <span class="o">-&gt;</span> <span class="n">x</span><span class="p">(</span><span class="n">at</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>


<p><code>countatoms</code>, <code>countresidues</code>, <code>countchains</code> and <code>countmodels</code> can be used to count elements. For example:</p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">countatoms</span><span class="p">(</span><span class="n">struc</span><span class="p">)</span>
<span class="mi">754</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">countatoms</span><span class="p">(</span><span class="n">struc</span><span class="p">,</span> <span class="n">calphaselector</span><span class="p">)</span>
<span class="mi">85</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">countresidues</span><span class="p">(</span><span class="n">struc</span><span class="p">,</span> <span class="n">standardselector</span><span class="p">)</span>
<span class="mi">85</span>
</pre></div>


<p>The sequence of a protein can be retrieved by passing a <code>Chain</code> or array of residues to <code>AminoAcidSequence</code>:</p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">AminoAcidSequence</span><span class="p">(</span><span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">],</span> <span class="n">standardselector</span><span class="p">)</span>
<span class="mi">85</span><span class="n">aa</span> <span class="n">Amino</span> <span class="n">Acid</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">RCGSQGGGSTCPGLRCCSIWGWCGDSEPYCGRTCENKCWSGERSDHRCGAAVGNPPCGQDRCCSVHGWCGGGNDYCSGGNCQYRC</span>
</pre></div>


<p><a id='Writing-PDB-files-1'></a></p>
<h2 id="writing-pdb-files">Writing PDB files</h2>
<p>PDB format files can be written:</p>
<div class="codehilite"><pre><span></span><span class="n">writepdb</span><span class="p">(</span><span class="s">&quot;1EN2_out.pdb&quot;</span><span class="p">,</span> <span class="n">struc</span><span class="p">)</span>
</pre></div>


<p>Any element type can be given as input to <code>writepdb</code>. Atom selectors can also be given as additional arguments:</p>
<div class="codehilite"><pre><span></span><span class="n">writepdb</span><span class="p">(</span><span class="s">&quot;1EN2_out.pdb&quot;</span><span class="p">,</span> <span class="n">struc</span><span class="p">,</span> <span class="n">backboneselector</span><span class="p">)</span>
</pre></div>


<p><a id='Spatial-calculations-1'></a></p>
<h2 id="spatial-calculations">Spatial calculations</h2>
<p>Various functions are provided to calculate spatial quantities for proteins:</p>
<table>
<thead>
<tr>
<th align="left">Command</th>
<th align="left">Returns</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>distance</code></td>
<td align="left">Minimum distance between two elements</td>
</tr>
<tr>
<td align="left"><code>sqdistance</code></td>
<td align="left">Minimum square distance between two elements</td>
</tr>
<tr>
<td align="left"><code>bondangle</code></td>
<td align="left">Angle between three atoms</td>
</tr>
<tr>
<td align="left"><code>dihedralangle</code></td>
<td align="left">Dihedral angle defined by four atoms</td>
</tr>
<tr>
<td align="left"><code>omegaangle</code></td>
<td align="left">Omega angle between a residue and the previous residue</td>
</tr>
<tr>
<td align="left"><code>phiangle</code></td>
<td align="left">Phi angle between a residue and the previous residue</td>
</tr>
<tr>
<td align="left"><code>psiangle</code></td>
<td align="left">Psi angle between a residue and the next residue</td>
</tr>
<tr>
<td align="left"><code>ramachandranangles</code></td>
<td align="left"><code>Vector</code>s of phi and psi angles of an element</td>
</tr>
<tr>
<td align="left"><code>contactmap</code></td>
<td align="left">Contact map of two element, or one element with itself</td>
</tr>
<tr>
<td align="left"><code>rmsd</code></td>
<td align="left">RMSD between two elements of the same size - assumes they are superimposed</td>
</tr>
<tr>
<td align="left"><code>displacements</code></td>
<td align="left"><code>Vector</code> of displacements between two elements of the same size - assumes they are superimposed</td>
</tr>
</tbody>
</table>
<p>For example:</p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">distance</span><span class="p">(</span><span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">10</span><span class="p">],</span> <span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">20</span><span class="p">])</span>
<span class="mf">10.782158874733762</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">rad2deg</span><span class="p">(</span><span class="n">bondangle</span><span class="p">(</span><span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">50</span><span class="p">][</span><span class="s">&quot;N&quot;</span><span class="p">],</span> <span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">50</span><span class="p">][</span><span class="s">&quot;CA&quot;</span><span class="p">],</span> <span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">50</span><span class="p">][</span><span class="s">&quot;C&quot;</span><span class="p">]))</span>
<span class="mf">110.77765846083398</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">rad2deg</span><span class="p">(</span><span class="n">dihedralangle</span><span class="p">(</span><span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">50</span><span class="p">][</span><span class="s">&quot;N&quot;</span><span class="p">],</span> <span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">50</span><span class="p">][</span><span class="s">&quot;CA&quot;</span><span class="p">],</span> <span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">50</span><span class="p">][</span><span class="s">&quot;C&quot;</span><span class="p">],</span> <span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">51</span><span class="p">][</span><span class="s">&quot;N&quot;</span><span class="p">]))</span>
<span class="o">-</span><span class="mf">177.38288114072924</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">rad2deg</span><span class="p">(</span><span class="n">psiangle</span><span class="p">(</span><span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">50</span><span class="p">],</span> <span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">51</span><span class="p">]))</span>
<span class="o">-</span><span class="mf">177.38288114072924</span>
</pre></div>


<p><a id='Examples-1'></a></p>
<h2 id="examples">Examples</h2>
<p>A few further examples of <code>Bio.Structure</code> usage are given below.</p>
<p><strong>A)</strong> To plot the temperature factors of a protein, if you have Gadfly installed:</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">Gadfly</span>
<span class="n">calphas</span> <span class="o">=</span> <span class="n">collectatoms</span><span class="p">(</span><span class="n">struc</span><span class="p">,</span> <span class="n">calphaselector</span><span class="p">)</span>
<span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">resnumber</span><span class="o">.</span><span class="p">(</span><span class="n">calphas</span><span class="p">),</span>
     <span class="n">y</span><span class="o">=</span><span class="n">tempfactor</span><span class="o">.</span><span class="p">(</span><span class="n">calphas</span><span class="p">),</span>
     <span class="n">Guide</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">&quot;Residue number&quot;</span><span class="p">),</span>
     <span class="n">Guide</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">&quot;Temperature factor&quot;</span><span class="p">),</span>
     <span class="n">Geom</span><span class="o">.</span><span class="n">line</span><span class="p">)</span>
</pre></div>


<p><strong>B)</strong> To print the PDB records for all C-alpha atoms within 5 Angstroms of residue 38:</p>
<div class="codehilite"><pre><span></span><span class="k">for</span> <span class="n">at</span> <span class="k">in</span> <span class="n">calphas</span>
    <span class="k">if</span> <span class="n">distance</span><span class="p">(</span><span class="n">struc</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">][</span><span class="mi">38</span><span class="p">],</span> <span class="n">at</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">5.0</span> <span class="o">&amp;&amp;</span> <span class="n">resnumber</span><span class="p">(</span><span class="n">at</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">38</span>
        <span class="n">println</span><span class="p">(</span><span class="n">pdbline</span><span class="p">(</span><span class="n">at</span><span class="p">))</span>
    <span class="k">end</span>
<span class="k">end</span>
</pre></div>


<p><strong>D)</strong> To view the contact map of a structure:</p>
<div class="codehilite"><pre><span></span><span class="n">cbetas</span> <span class="o">=</span> <span class="n">collectatoms</span><span class="p">(</span><span class="n">struc</span><span class="p">,</span> <span class="n">cbetaselector</span><span class="p">)</span>
<span class="n">contacts</span> <span class="o">=</span> <span class="n">contactmap</span><span class="p">(</span><span class="n">cbetas</span><span class="p">,</span> <span class="mf">7.0</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="k">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">cbetas</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span> <span class="k">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">cbetas</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">contacts</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">]</span>
            <span class="n">print</span><span class="p">(</span><span class="s">&quot;â–ˆ&quot;</span><span class="p">)</span>
        <span class="k">else</span>
            <span class="n">print</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="p">)</span>
        <span class="k">end</span>
    <span class="k">end</span>
    <span class="n">println</span><span class="p">()</span>
<span class="k">end</span>
</pre></div>


<p><code>cbetaselector</code> selects C-beta atoms, or C-alpha atoms for glycine residues. <code>contactmap</code> can also be given two structural elements as arguments, in which case a non-symmetrical 2D array is returned showing contacts between the elements.</p>
<p><strong>E)</strong> To show the Ramachandran phi/psi angle plot of a structure, if you have Gadfly installed:</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">Gadfly</span>
<span class="n">phi_angles</span><span class="p">,</span> <span class="n">psi_angles</span> <span class="o">=</span> <span class="n">ramachandranangles</span><span class="p">(</span><span class="n">struc</span><span class="p">,</span> <span class="n">standardselector</span><span class="p">)</span>
<span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">rad2deg</span><span class="o">.</span><span class="p">(</span><span class="n">phi_angles</span><span class="p">),</span>
     <span class="n">y</span><span class="o">=</span><span class="n">rad2deg</span><span class="o">.</span><span class="p">(</span><span class="n">psi_angles</span><span class="p">),</span>
     <span class="n">Guide</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">&quot;Phi / degrees&quot;</span><span class="p">),</span>
     <span class="n">Guide</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">&quot;Psi / degrees&quot;</span><span class="p">),</span>
     <span class="n">Guide</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">ticks</span><span class="o">=</span><span class="p">[</span><span class="o">-</span><span class="mi">180</span><span class="p">,</span><span class="o">-</span><span class="mi">90</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">90</span><span class="p">,</span><span class="mi">180</span><span class="p">]),</span>
     <span class="n">Guide</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">ticks</span><span class="o">=</span><span class="p">[</span><span class="o">-</span><span class="mi">180</span><span class="p">,</span><span class="o">-</span><span class="mi">90</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">90</span><span class="p">,</span><span class="mi">180</span><span class="p">]))</span>
</pre></div>


<p><strong>F)</strong> To calculate the RMSD and displacements between the heavy (non-hydrogen) atoms of two models in an NMR structure:</p>
<div class="codehilite"><pre><span></span><span class="n">downloadpdb</span><span class="p">(</span><span class="s">&quot;1SSU&quot;</span><span class="p">)</span>
<span class="n">struc_nmr</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="s">&quot;1SSU.pdb&quot;</span><span class="p">,</span> <span class="n">PDB</span><span class="p">)</span>
<span class="n">rmsd</span><span class="p">(</span><span class="n">struc_nmr</span><span class="p">[</span><span class="mi">5</span><span class="p">],</span> <span class="n">struc_nmr</span><span class="p">[</span><span class="mi">10</span><span class="p">],</span> <span class="n">heavyatomselector</span><span class="p">)</span>
<span class="n">displacements</span><span class="p">(</span><span class="n">struc_nmr</span><span class="p">[</span><span class="mi">5</span><span class="p">],</span> <span class="n">struc_nmr</span><span class="p">[</span><span class="mi">10</span><span class="p">],</span> <span class="n">heavyatomselector</span><span class="p">)</span>
</pre></div></div>
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Copyright (c) 2016 The BioJulia Organisation<br></small>
        
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../../js/jquery-1.10.2.min.js"></script>
    <script src="../../js/bootstrap-3.0.3.min.js"></script>
    <script src="../../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '../..';
    </script>
    <script data-main="../../mkdocs/js/search.js" src="../../mkdocs/js/require.js"></script>
    <script src="../../js/base.js"></script>
    <script>
        ((window.gitter = {}).chat = {}).options = {
            room: 'BioJulia/Bio.jl'
        };
    </script>
    <script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <script src="../../assets/mathjaxhelper.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
