<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0"> 
    <meta name="author" content="The BioJulia Organisation"> 
    <link rel="canonical" href="https://biojulia.github.io/Bio.jl/man/alignments/"> 
    <link rel="shortcut icon" href="../../img/favicon.ico">

    <title>Alignments - Bio.jl</title>

    <link href="../../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../../css/base.css" rel="stylesheet">
    <link href="../../css/biojulia.css" rel="stylesheet">
    <link rel="stylesheet" href="../../css/highlight.css">
    <link rel="stylesheet" href="../../css/friendly.css" type="text/css">
    <link href="../../assets/Documenter.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="../..">Bio.jl</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../..">Home</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Manual <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../install/">Installation</a>
</li>

                        
                            
<li >
    <a href="../reading/">IO API</a>
</li>

                        
                            
<li >
    <a href="../seq/">Biological Sequences</a>
</li>

                        
                            
<li class="active">
    <a href="./">Alignments</a>
</li>

                        
                            
<li >
    <a href="../intervals/">Intervals</a>
</li>

                        
                            
<li >
    <a href="../var/">Biological Variation</a>
</li>

                        
                            
<li >
    <a href="../structure/">Structure</a>
</li>

                        
                            
<li >
    <a href="../util/">Util</a>
</li>

                        
                            
<li >
    <a href="../tools/">Tools</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Reference <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../../lib/seq/">Seq</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li >
                        <a href="../contributing/">Contributing</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../seq/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../intervals/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/BioJulia/Bio.jl">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="first-level active"><a href="#align-sequence-alignments">Align: Sequence Alignments</a></li>
        
            <li class="second-level"><a href="#representing-alignments">Representing alignments</a></li>
            
                <li class="third-level"><a href="#operations">Operations</a></li>
            
        
            <li class="second-level"><a href="#aligned-sequence">Aligned sequence</a></li>
            
        
            <li class="second-level"><a href="#operating-on-alignments">Operating on alignments</a></li>
            
        
            <li class="second-level"><a href="#pairwise-alignment">Pairwise alignment</a></li>
            
                <li class="third-level"><a href="#operations-on-pairwise-alignment">Operations on pairwise alignment</a></li>
            
        
            <li class="second-level"><a href="#substitution-matrices">Substitution matrices</a></li>
            
        
            <li class="second-level"><a href="#alignment-file-formats-for-high-throughput-sequencing">Alignment file formats for high-throughput sequencing</a></li>
            
                <li class="third-level"><a href="#sam-and-bam-file-formats">SAM and BAM file formats</a></li>
            
                <li class="third-level"><a href="#performance-tips">Performance tips</a></li>
            
        
    
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<p><a id='Align:-Sequence-Alignments-1'></a></p>
<h1 id="align-sequence-alignments">Align: Sequence Alignments</h1>
<p>The <code>Align</code> module contains tools for computing and working with sequence alignments.</p>
<p><a id='Representing-alignments-1'></a></p>
<h2 id="representing-alignments">Representing alignments</h2>
<p>The <code>Alignment</code> type can represent a wide variety of global or local sequence alignments while facilitating efficient coordinate transformation.  Alignment are always relative to a possibly unspecified reference sequence and represent a series of <a href="https://en.wikipedia.org/wiki/Edit_distance">edit operations</a> performed on that reference to transform it to the query sequence.</p>
<p>To represent an alignment we use a series of "anchors" stored in the <code>AlignmentAnchor</code> type. Anchors are form of run-length encoding alignment operations, but rather than store an operation along with a length, we store the end-point of that operation in both reference and query coordinates.</p>
<div class="codehilite"><pre><span></span><span class="k">immutable</span> <span class="n">AlignmentAnchor</span>
    <span class="n">seqpos</span><span class="p">::</span><span class="kt">Int</span>
    <span class="n">refpos</span><span class="p">::</span><span class="kt">Int</span>
    <span class="n">op</span><span class="p">::</span><span class="n">Operation</span>
<span class="k">end</span>
</pre></div>


<p>Every alignment starts with a special <code>OP_START</code> operation which is used to give the position in the reference and query prior to the start of the alignment, or 0, if the alignment starts at position 1.</p>
<p>For example, consider the following alignment:</p>
<div class="codehilite"><pre><span></span>              0   4        9  12 15     19
              |   |        |  |  |      |
    query:     TGGC----ATCATTTAACG---CAAG
reference: AGGGTGGCATTTATCAG---ACGTTTCGAGAC
              |   |   |    |     |  |   |
              4   8   12   17    20 23  27
</pre></div>


<p>Using anchors we would represent this as the following series of anchors:</p>
<div class="codehilite"><pre><span></span><span class="p">[</span>
    <span class="n">AlignmentAnchor</span><span class="p">(</span> <span class="mi">0</span><span class="p">,</span>  <span class="mi">4</span><span class="p">,</span> <span class="n">OP_START</span><span class="p">),</span>
    <span class="n">AlignmentAnchor</span><span class="p">(</span> <span class="mi">4</span><span class="p">,</span>  <span class="mi">8</span><span class="p">,</span> <span class="n">OP_MATCH</span><span class="p">),</span>
    <span class="n">AlignmentAnchor</span><span class="p">(</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="n">OP_DELETE</span><span class="p">),</span>
    <span class="n">AlignmentAnchor</span><span class="p">(</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">17</span><span class="p">,</span> <span class="n">OP_MATCH</span><span class="p">),</span>
    <span class="n">AlignmentAnchor</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">17</span><span class="p">,</span> <span class="n">OP_INSERT</span><span class="p">),</span>
    <span class="n">AlignmentAnchor</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="n">OP_MATCH</span><span class="p">),</span>
    <span class="n">AlignmentAnchor</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">23</span><span class="p">,</span> <span class="n">OP_DELETE</span><span class="p">),</span>
    <span class="n">AlignmentAnchor</span><span class="p">(</span><span class="mi">19</span><span class="p">,</span> <span class="mi">27</span><span class="p">,</span> <span class="n">OP_MATCH</span><span class="p">)</span>
<span class="p">]</span>
</pre></div>


<p>An <code>Alignment</code> object can be created from a series of anchors:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">Alignment</span><span class="p">([</span>

<span class="gt">           AlignmentAnchor(0, 4, OP_START),</span>

<span class="gt">           AlignmentAnchor(4, 8, OP_MATCH),</span>

<span class="gt">           AlignmentAnchor(4, 12, OP_DELETE)</span>

<span class="gt">       ])</span>
<span class="go">········</span>
<span class="go">····----</span>
</pre></div>


<p><a id='Operations-1'></a></p>
<h3 id="operations">Operations</h3>
<p>Alignment operations follow closely from those used in the <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAM/BAM format</a> and are stored in the <code>Operation</code> bitstype.</p>
<table>
<thead>
<tr>
<th align="left">Operation</th>
<th align="left">Operation Type</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>OP_MATCH</code></td>
<td align="left">match</td>
<td align="left">non-specific match</td>
</tr>
<tr>
<td align="left"><code>OP_INSERT</code></td>
<td align="left">insert</td>
<td align="left">insertion into reference sequence</td>
</tr>
<tr>
<td align="left"><code>OP_DELETE</code></td>
<td align="left">delete</td>
<td align="left">deletion from reference sequence</td>
</tr>
<tr>
<td align="left"><code>OP_SKIP</code></td>
<td align="left">delete</td>
<td align="left">(typically long) deletion from the reference, e.g. due to RNA splicing</td>
</tr>
<tr>
<td align="left"><code>OP_SOFT_CLIP</code></td>
<td align="left">insert</td>
<td align="left">sequence removed from the beginning or end of the query sequence but stored</td>
</tr>
<tr>
<td align="left"><code>OP_HARD_CLIP</code></td>
<td align="left">insert</td>
<td align="left">sequence removed from the beginning or end of the query sequence and not stored</td>
</tr>
<tr>
<td align="left"><code>OP_PAD</code></td>
<td align="left">special</td>
<td align="left">not currently supported, but present for SAM/BAM compatibility</td>
</tr>
<tr>
<td align="left"><code>OP_SEQ_MATCH</code></td>
<td align="left">match</td>
<td align="left">match operation with matching sequence positions</td>
</tr>
<tr>
<td align="left"><code>OP_SEQ_MISMATCH</code></td>
<td align="left">match</td>
<td align="left">match operation with mismatching sequence positions</td>
</tr>
<tr>
<td align="left"><code>OP_BACK</code></td>
<td align="left">special</td>
<td align="left">not currently supported, but present for SAM/BAM compatibility</td>
</tr>
<tr>
<td align="left"><code>OP_START</code></td>
<td align="left">special</td>
<td align="left">indicate the start of an alignment within the reference and query sequence</td>
</tr>
</tbody>
</table>
<p><a id='Aligned-sequence-1'></a></p>
<h2 id="aligned-sequence">Aligned sequence</h2>
<p>A sequence aligned to another sequence is represented by the <code>AlignedSequence</code> type, which is a pair of the aligned sequence and an <code>Alignment</code> object.</p>
<p>The following example creates an aligned sequence object from a sequence and an alignment:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">AlignedSequence</span><span class="p">(</span>  <span class="c"># pass an Alignment object</span>

<span class="gt">           dna&quot;ACGTAT&quot;,</span>

<span class="gt">           Alignment([</span>

<span class="gt">               AlignmentAnchor(0, 0, OP_START),</span>

<span class="gt">               AlignmentAnchor(3, 3, OP_MATCH),</span>

<span class="gt">               AlignmentAnchor(6, 3, OP_INSERT)</span>

<span class="gt">           ])</span>

<span class="gt">       )</span>
<span class="go">···---</span>
<span class="go">ACGTAT</span>

<span class="gp">julia&gt;</span> <span class="n">AlignedSequence</span><span class="p">(</span>  <span class="c"># or pass a vector of anchors</span>

<span class="gt">           dna&quot;ACGTAT&quot;,</span>

<span class="gt">           [</span>

<span class="gt">               AlignmentAnchor(0, 0, OP_START),</span>

<span class="gt">               AlignmentAnchor(3, 3, OP_MATCH),</span>

<span class="gt">               AlignmentAnchor(6, 3, OP_INSERT)</span>

<span class="gt">           ]</span>

<span class="gt">       )</span>
<span class="go">···---</span>
<span class="go">ACGTAT</span>
</pre></div>


<p>If you already have an aligned sequence with gap symbols, it can be converted to an <code>AlignedSequence</code> object by passing a reference sequence with it:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;ACGT--AAT--&quot;</span>
<span class="go">11nt DNA Sequence:</span>
<span class="go">ACGT--AAT--</span>

<span class="gp">julia&gt;</span> <span class="n">ref</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;ACGTTTAT-GG&quot;</span>
<span class="go">11nt DNA Sequence:</span>
<span class="go">ACGTTTAT-GG</span>

<span class="gp">julia&gt;</span> <span class="n">AlignedSequence</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">ref</span><span class="p">)</span>
<span class="go">········-··</span>
<span class="go">ACGT--AAT--</span>
</pre></div>


<p><a id='Operating-on-alignments-1'></a></p>
<h2 id="operating-on-alignments">Operating on alignments</h2>
<p><a id='Base.first' href='#Base.first'>#</a>
<strong><code>Base.first</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>first(coll)
</pre></div>


<p>Get the first element of an iterable collection. Returns the start point of a <a href="../:obj:Range"><code>Range</code></a> even if it is empty.</p>
<p><a target='_blank' href='https://github.com/JuliaLang/julia/tree/3c9d75391c72d7c32eea75ff187ce77b2d5effc8/base/docs/helpdb/Base.jl#L5773-L5778' class='documenter-source'>source</a><br></p>
<p><a id='Base.last' href='#Base.last'>#</a>
<strong><code>Base.last</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>last(coll)
</pre></div>


<p>Get the last element of an ordered collection, if it can be computed in O(1) time. This is accomplished by calling <a href="../:func:endof"><code>endof</code></a> to get the last index. Returns the end point of a <a href="../:obj:Range"><code>Range</code></a> even if it is empty.</p>
<p><a target='_blank' href='https://github.com/JuliaLang/julia/tree/3c9d75391c72d7c32eea75ff187ce77b2d5effc8/base/docs/helpdb/Base.jl#L889-L895' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.seq2ref' href='#Bio.Align.seq2ref'>#</a>
<strong><code>Bio.Align.seq2ref</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>seq2ref(aln, i)
</pre></div>


<p>Map a position from sequence to reference.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/alignment.jl#L86-L90' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.ref2seq' href='#Bio.Align.ref2seq'>#</a>
<strong><code>Bio.Align.ref2seq</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>ref2seq(aln, i)
</pre></div>


<p>Map a position from reference to sequence.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/alignment.jl#L104-L108' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.cigar' href='#Bio.Align.cigar'>#</a>
<strong><code>Bio.Align.cigar</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>cigar(aln::Alignment)
</pre></div>


<p>Make a CIGAR string encoding of <code>aln</code>.</p>
<p>This is not entirely lossless as it discards the alignments start positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/alignment.jl#L122-L128' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>cigar(rec::BAMRecord)
</pre></div>


<p>Return a CIGAR string of the alignment <code>rec</code>. See also <code>cigar_rle</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/hts/bam/record.jl#L193-L197' class='documenter-source'>source</a><br></p>
<p><a id='Pairwise-alignment-1'></a></p>
<h2 id="pairwise-alignment">Pairwise alignment</h2>
<p>Pairwise alignment is a sequence alignment between two sequences.  The <code>Bio.Align</code> module implements several pairwise alignment methods that maximize alignment score or minimize alignment cost.</p>
<p>A pair of optimization type and score (or cost) model determines the best alignments between two sequences. The next example uses a pair of <code>GlobalAlignment</code> and <code>AffineGapScoreModel</code> to obtain the best alignment:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">problem</span> <span class="o">=</span> <span class="n">GlobalAlignment</span><span class="p">()</span>
<span class="go">Bio.Align.GlobalAlignment()</span>

<span class="gp">julia&gt;</span> <span class="n">scoremodel</span> <span class="o">=</span> <span class="n">AffineGapScoreModel</span><span class="p">(</span>

<span class="gt">                  match=5,</span>

<span class="gt">                  mismatch=-4,</span>

<span class="gt">                  gap_open=-4,</span>

<span class="gt">                  gap_extend=-1</span>

<span class="gt">              )</span>
<span class="go">Bio.Align.AffineGapScoreModel{Int64}:</span>
<span class="go">    mismatch = -4</span>
<span class="go">    gap_open = -4</span>
<span class="go">  gap_extend = -1</span>


<span class="gt">       match = 5</span>
<span class="gp">julia&gt;</span> <span class="n">pairalign</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;CGGATTA&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;GGTTTAC&quot;</span><span class="p">,</span> <span class="n">scoremodel</span><span class="p">)</span>
<span class="go">Bio.Align.PairwiseAlignmentResult{Int64,Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}},Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}}}:</span>
<span class="go">  score: 11</span>
<span class="go">  seq: 1 CGGATTA- 7</span>
<span class="go">  ref: 0 -GGTTTAC 7</span>
</pre></div>


<p><code>pairalign</code> takes an alignment type as its first argument, then two sequences to align, and score (or cost) model. Alignment type is one of the following four types:</p>
<ul>
<li><code>GlobalAlignment</code>: global-to-global alignment</li>
<li><code>SemiGlobalAlignment</code>: local-to-global alignment</li>
<li><code>LocalAlignment</code>: local-to-local alignment</li>
<li><code>OverlapAlignment</code>: end-free alignment</li>
</ul>
<p>For scoring model, <code>AffineGapScoreModel</code> is currently supported. It imposes an <strong>affine gap penalty</strong> for insertions and deletions: <code>gap_open + k * gap_extend</code> for a consecutive insertion/deletion of length <code>k</code>. The affine gap penalty is flexible enough to create a constant and linear scoring model. Setting <code>gap_extend = 0</code> or <code>gap_open = 0</code>, they are equivalent to the constant or linear gap penalty, respectively. The first argument of <code>AffineGapScoreModel</code> can be a substitution matrix like <code>AffineGapScoreModel(BLOSUM62, gap_open=-10, gap_extend=-1)</code>. For details on substitution matrices, see the <a href="./#Substitution-matrices-1">Substitution matrices</a> section.</p>
<p>Alignment type can also be a distance of two sequences:</p>
<ul>
<li><code>EditDistance</code></li>
<li><code>LevenshteinDistance</code></li>
<li><code>HammingDistance</code></li>
</ul>
<p>In this alignment, <code>CostModel</code> is used instead of <code>AffineGapScoreModel</code> to define cost of substitution, insertion, and deletion:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">costmodel</span> <span class="o">=</span> <span class="n">CostModel</span><span class="p">(</span><span class="n">match</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">mismatch</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">insertion</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">deletion</span><span class="o">=</span><span class="mi">1</span><span class="p">);</span>

<span class="gp">julia&gt;</span> <span class="n">pairalign</span><span class="p">(</span><span class="n">EditDistance</span><span class="p">(),</span> <span class="s">&quot;abcd&quot;</span><span class="p">,</span> <span class="s">&quot;adcde&quot;</span><span class="p">,</span> <span class="n">costmodel</span><span class="p">)</span>
<span class="go">Bio.Align.PairwiseAlignmentResult{Int64,String,String}:</span>
<span class="go">  distance: 2</span>
<span class="go">  seq: 1 abcd- 4</span>
<span class="go">  ref: 1 adcde 5</span>
</pre></div>


<p><a id='Operations-on-pairwise-alignment-1'></a></p>
<h3 id="operations-on-pairwise-alignment">Operations on pairwise alignment</h3>
<p><code>pairalign</code> returns a <code>PairwiseAlignmentResult</code> object and some accessors are provided for it.</p>
<p><a id='Bio.Align.score' href='#Bio.Align.score'>#</a>
<strong><code>Bio.Align.score</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>score(alignment_result)
</pre></div>


<p>Return score of alignment.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/result.jl#L33-L37' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.distance' href='#Bio.Align.distance'>#</a>
<strong><code>Bio.Align.distance</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>distance(alignment_result)
</pre></div>


<p>Retrun distance of alignment.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/result.jl#L40-L44' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.hasalignment' href='#Bio.Align.hasalignment'>#</a>
<strong><code>Bio.Align.hasalignment</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>hasalignment(alignment_result)
</pre></div>


<p>Check if alignment is stored or not.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/result.jl#L48-L52' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.alignment' href='#Bio.Align.alignment'>#</a>
<strong><code>Bio.Align.alignment</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alignment(alignment_result)
</pre></div>


<p>Return alignment if any.</p>
<p>See also: <code>hasalignment</code></p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/result.jl#L55-L61' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>alignment(rec::BAMRecord)
</pre></div>


<p>Make an alignment object from <code>rec</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/hts/bam/record.jl#L206-L210' class='documenter-source'>source</a><br></p>
<p>Pairwise alignment also implements some useful operations on it.</p>
<p><a id='Bio.Align.count_matches' href='#Bio.Align.count_matches'>#</a>
<strong><code>Bio.Align.count_matches</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_matches(aln)
</pre></div>


<p>Count the number of matching positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/alignment.jl#L84-L88' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.count_mismatches' href='#Bio.Align.count_mismatches'>#</a>
<strong><code>Bio.Align.count_mismatches</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_mismatches(aln)
</pre></div>


<p>Count the number of mismatching positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/alignment.jl#L91-L95' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.count_insertions' href='#Bio.Align.count_insertions'>#</a>
<strong><code>Bio.Align.count_insertions</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_insertions(aln)
</pre></div>


<p>Count the number of inserting positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/alignment.jl#L98-L102' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.count_deletions' href='#Bio.Align.count_deletions'>#</a>
<strong><code>Bio.Align.count_deletions</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_deletions(aln)
</pre></div>


<p>Count the number of deleting positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/alignment.jl#L105-L109' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.count_aligned' href='#Bio.Align.count_aligned'>#</a>
<strong><code>Bio.Align.count_aligned</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_aligned(aln)
</pre></div>


<p>Count the number of aligned positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/8e19330a756461c4e040ffb0097e8f033a94c152/src/align/pairwise/alignment.jl#L112-L116' class='documenter-source'>source</a><br></p>
<p>The example below shows a use case of these operations:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">s1</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;CCTAGGAGGG&quot;</span><span class="p">;</span>

<span class="gp">julia&gt;</span> <span class="n">s2</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;ACCTGGTATGATAGCG&quot;</span><span class="p">;</span>

<span class="gp">julia&gt;</span> <span class="n">scoremodel</span> <span class="o">=</span> <span class="n">AffineGapScoreModel</span><span class="p">(</span><span class="n">EDNAFULL</span><span class="p">,</span> <span class="n">gap_open</span><span class="o">=-</span><span class="mi">5</span><span class="p">,</span> <span class="n">gap_extend</span><span class="o">=-</span><span class="mi">1</span><span class="p">);</span>

<span class="gp">julia&gt;</span> <span class="n">res</span> <span class="o">=</span> <span class="n">pairalign</span><span class="p">(</span><span class="n">GlobalAlignment</span><span class="p">(),</span> <span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">scoremodel</span><span class="p">)</span>  <span class="c"># run pairwise alignment</span>
<span class="go">Bio.Align.PairwiseAlignmentResult{Int64,Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}},Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}}}:</span>
<span class="go">  score: 13</span>
<span class="go">  seq:  0 -CCTAGG------AGGG 10</span>
<span class="go">  ref:  1 ACCT-GGTATGATAGCG 16</span>



<span class="gt">           ||| ||      || |</span>
<span class="gp">julia&gt;</span> <span class="n">score</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>  <span class="c"># get the achieved score of this alignment</span>
<span class="go">13</span>

<span class="gp">julia&gt;</span> <span class="n">aln</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="go">PairwiseAlignment{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}},Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}}}:</span>
<span class="go">  seq:  0 -CCTAGG------AGGG 10</span>
<span class="go">  ref:  1 ACCT-GGTATGATAGCG 16</span>



<span class="gt">           ||| ||      || |</span>
<span class="gp">julia&gt;</span> <span class="n">count_matches</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="go">8</span>

<span class="gp">julia&gt;</span> <span class="n">count_mismatches</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="go">1</span>

<span class="gp">julia&gt;</span> <span class="n">count_insertions</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="go">1</span>

<span class="gp">julia&gt;</span> <span class="n">count_deletions</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="go">7</span>

<span class="gp">julia&gt;</span> <span class="n">count_aligned</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="go">17</span>

<span class="gp">julia&gt;</span> <span class="n">collect</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>  <span class="c"># pairwise alignment is iterable</span>
<span class="go">17-element Array{Tuple{Bio.Seq.DNANucleotide,Bio.Seq.DNANucleotide},1}:</span>
<span class="go"> (-,A)</span>
<span class="go"> (C,C)</span>
<span class="go"> (C,C)</span>
<span class="go"> (T,T)</span>
<span class="go"> (A,-)</span>
<span class="go"> (G,G)</span>
<span class="go"> (G,G)</span>
<span class="go"> (-,T)</span>
<span class="go"> (-,A)</span>
<span class="go"> (-,T)</span>
<span class="go"> (-,G)</span>
<span class="go"> (-,A)</span>
<span class="go"> (-,T)</span>
<span class="go"> (A,A)</span>
<span class="go"> (G,G)</span>
<span class="go"> (G,C)</span>
<span class="go"> (G,G)</span>

<span class="gp">julia&gt;</span> <span class="n">DNASequence</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="k">in</span> <span class="n">aln</span><span class="p">])</span>  <span class="c"># create aligned `s1` with gaps</span>
<span class="go">17nt DNA Sequence:</span>
<span class="go">-CCTAGG------AGGG</span>

<span class="gp">julia&gt;</span> <span class="n">DNASequence</span><span class="p">([</span><span class="n">y</span> <span class="k">for</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="k">in</span> <span class="n">aln</span><span class="p">])</span>  <span class="c"># create aligned `s2` with gaps</span>
<span class="go">17nt DNA Sequence:</span>
<span class="go">ACCT-GGTATGATAGCG</span>
</pre></div>


<p><a id='Substitution-matrices-1'></a></p>
<h2 id="substitution-matrices">Substitution matrices</h2>
<p>A substitution matrix is a function of substitution score (or cost) from one symbol to other. Substitution value of <code>submat</code> from <code>x</code> to <code>y</code> can be obtained by writing <code>submat[x,y]</code>. In <code>Bio.Align</code>, <code>SubstitutionMatrix</code> and <code>DichotomousSubstitutionMatrix</code> are two distinct types representing substitution matrices.</p>
<p><code>SubstitutionMatrix</code> is a general substitution matrix type that is a thin wrapper of regular matrix.</p>
<p>Some common substitution matrices are provided. For DNA and RNA, <code>EDNAFULL</code> is defined:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">EDNAFULL</span>
<span class="go">Bio.Align.SubstitutionMatrix{Bio.Seq.DNANucleotide,Int64}:</span>
<span class="go">     A  C  G  T  M  R  W  S  Y  K  V  H  D  B  N</span>
<span class="go">  A  5 -4 -4 -4  1  1  1 -4 -4 -4 -1 -1 -1 -4 -2</span>
<span class="go">  C -4  5 -4 -4  1 -4 -4  1  1 -4 -1 -1 -4 -1 -2</span>
<span class="go">  G -4 -4  5 -4 -4  1 -4  1 -4  1 -1 -4 -1 -1 -2</span>
<span class="go">  T -4 -4 -4  5 -4 -4  1 -4  1  1 -4 -1 -1 -1 -2</span>
<span class="go">  M  1  1 -4 -4 -1 -2 -2 -2 -2 -4 -1 -1 -3 -3 -1</span>
<span class="go">  R  1 -4  1 -4 -2 -1 -2 -2 -4 -2 -1 -3 -1 -3 -1</span>
<span class="go">  W  1 -4 -4  1 -2 -2 -1 -4 -2 -2 -3 -1 -1 -3 -1</span>
<span class="go">  S -4  1  1 -4 -2 -2 -4 -1 -2 -2 -1 -3 -3 -1 -1</span>
<span class="go">  Y -4  1 -4  1 -2 -4 -2 -2 -1 -2 -3 -1 -3 -1 -1</span>
<span class="go">  K -4 -4  1  1 -4 -2 -2 -2 -2 -1 -3 -3 -1 -1 -1</span>
<span class="go">  V -1 -1 -1 -4 -1 -1 -3 -1 -3 -3 -1 -2 -2 -2 -1</span>
<span class="go">  H -1 -1 -4 -1 -1 -3 -1 -3 -1 -3 -2 -1 -2 -2 -1</span>
<span class="go">  D -1 -4 -1 -1 -3 -1 -1 -3 -3 -1 -2 -2 -1 -2 -1</span>
<span class="go">  B -4 -1 -1 -1 -3 -3 -3 -1 -1 -1 -2 -2 -2 -1 -1</span>
<span class="go">  N -2 -2 -2 -2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1</span>
<span class="go">(underlined values are default ones)</span>
</pre></div>


<p>For amino acids, PAM (Point Accepted Mutation) and BLOSUM (BLOcks SUbstitution Matrix) matrices are defined:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">BLOSUM62</span>
<span class="go">Bio.Align.SubstitutionMatrix{Bio.Seq.AminoAcid,Int64}:</span>
<span class="go">     A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  O  U  B  J  Z  X  *</span>
<span class="go">  A  4 -1 -2 -2  0 -1 -1  0 -2 -1 -1 -1 -1 -2 -1  1  0 -3 -2  0  0̲  0̲ -2  0̲ -1  0 -4</span>
<span class="go">  R -1  5  0 -2 -3  1  0 -2  0 -3 -2  2 -1 -3 -2 -1 -1 -3 -2 -3  0̲  0̲ -1  0̲  0 -1 -4</span>
<span class="go">  N -2  0  6  1 -3  0  0  0  1 -3 -3  0 -2 -3 -2  1  0 -4 -2 -3  0̲  0̲  3  0̲  0 -1 -4</span>
<span class="go">  D -2 -2  1  6 -3  0  2 -1 -1 -3 -4 -1 -3 -3 -1  0 -1 -4 -3 -3  0̲  0̲  4  0̲  1 -1 -4</span>
<span class="go">  C  0 -3 -3 -3  9 -3 -4 -3 -3 -1 -1 -3 -1 -2 -3 -1 -1 -2 -2 -1  0̲  0̲ -3  0̲ -3 -2 -4</span>
<span class="go">  Q -1  1  0  0 -3  5  2 -2  0 -3 -2  1  0 -3 -1  0 -1 -2 -1 -2  0̲  0̲  0  0̲  3 -1 -4</span>
<span class="go">  E -1  0  0  2 -4  2  5 -2  0 -3 -3  1 -2 -3 -1  0 -1 -3 -2 -2  0̲  0̲  1  0̲  4 -1 -4</span>
<span class="go">  G  0 -2  0 -1 -3 -2 -2  6 -2 -4 -4 -2 -3 -3 -2  0 -2 -2 -3 -3  0̲  0̲ -1  0̲ -2 -1 -4</span>
<span class="go">  H -2  0  1 -1 -3  0  0 -2  8 -3 -3 -1 -2 -1 -2 -1 -2 -2  2 -3  0̲  0̲  0  0̲  0 -1 -4</span>
<span class="go">  I -1 -3 -3 -3 -1 -3 -3 -4 -3  4  2 -3  1  0 -3 -2 -1 -3 -1  3  0̲  0̲ -3  0̲ -3 -1 -4</span>
<span class="go">  L -1 -2 -3 -4 -1 -2 -3 -4 -3  2  4 -2  2  0 -3 -2 -1 -2 -1  1  0̲  0̲ -4  0̲ -3 -1 -4</span>
<span class="go">  K -1  2  0 -1 -3  1  1 -2 -1 -3 -2  5 -1 -3 -1  0 -1 -3 -2 -2  0̲  0̲  0  0̲  1 -1 -4</span>
<span class="go">  M -1 -1 -2 -3 -1  0 -2 -3 -2  1  2 -1  5  0 -2 -1 -1 -1 -1  1  0̲  0̲ -3  0̲ -1 -1 -4</span>
<span class="go">  F -2 -3 -3 -3 -2 -3 -3 -3 -1  0  0 -3  0  6 -4 -2 -2  1  3 -1  0̲  0̲ -3  0̲ -3 -1 -4</span>
<span class="go">  P -1 -2 -2 -1 -3 -1 -1 -2 -2 -3 -3 -1 -2 -4  7 -1 -1 -4 -3 -2  0̲  0̲ -2  0̲ -1 -2 -4</span>
<span class="go">  S  1 -1  1  0 -1  0  0  0 -1 -2 -2  0 -1 -2 -1  4  1 -3 -2 -2  0̲  0̲  0  0̲  0  0 -4</span>
<span class="go">  T  0 -1  0 -1 -1 -1 -1 -2 -2 -1 -1 -1 -1 -2 -1  1  5 -2 -2  0  0̲  0̲ -1  0̲ -1  0 -4</span>
<span class="go">  W -3 -3 -4 -4 -2 -2 -3 -2 -2 -3 -2 -3 -1  1 -4 -3 -2 11  2 -3  0̲  0̲ -4  0̲ -3 -2 -4</span>
<span class="go">  Y -2 -2 -2 -3 -2 -1 -2 -3  2 -1 -1 -2 -1  3 -3 -2 -2  2  7 -1  0̲  0̲ -3  0̲ -2 -1 -4</span>
<span class="go">  V  0 -3 -3 -3 -1 -2 -2 -3 -3  3  1 -2  1 -1 -2 -2  0 -3 -1  4  0̲  0̲ -3  0̲ -2 -1 -4</span>
<span class="go">  O  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲</span>
<span class="go">  U  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲</span>
<span class="go">  B -2 -1  3  4 -3  0  1 -1  0 -3 -4  0 -3 -3 -2  0 -1 -4 -3 -3  0̲  0̲  4  0̲  1 -1 -4</span>
<span class="go">  J  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲  0̲</span>
<span class="go">  Z -1  0  0  1 -3  3  4 -2  0 -3 -3  1 -1 -3 -1  0 -1 -3 -2 -2  0̲  0̲  1  0̲  4 -1 -4</span>
<span class="go">  X  0 -1 -1 -1 -2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -2  0  0 -2 -1 -1  0̲  0̲ -1  0̲ -1 -1 -4</span>
<span class="go">  * -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4  0̲  0̲ -4  0̲ -4 -4  1</span>
<span class="go">(underlined values are default ones)</span>
</pre></div>


<table>
<thead>
<tr>
<th align="left">Matrix</th>
<th align="left">Constants</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">PAM</td>
<td align="left"><code>PAM30</code>, <code>PAM70</code>, <code>PAM250</code></td>
</tr>
<tr>
<td align="left">BLOSUM</td>
<td align="left"><code>BLOSUM45</code>, <code>BLOSUM50</code>, <code>BLOSUM62</code>, <code>BLOSUM80</code>, <code>BLOSUM90</code></td>
</tr>
</tbody>
</table>
<p>These matrices are downloaded from: <a href="ftp://ftp.ncbi.nih.gov/blast/matrices/">ftp://ftp.ncbi.nih.gov/blast/matrices/</a>.</p>
<p><code>SubstitutionMatrix</code> can be modified like a regular matrix:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">mysubmat</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="n">BLOSUM62</span><span class="p">);</span>  <span class="c"># create a copy</span>

<span class="gp">julia&gt;</span> <span class="n">mysubmat</span><span class="p">[</span><span class="n">AA_A</span><span class="p">,</span><span class="n">AA_R</span><span class="p">]</span>  <span class="c"># score of AA_A =&gt; AA_R substitution is -1</span>
<span class="go">-1</span>

<span class="gp">julia&gt;</span> <span class="n">mysubmat</span><span class="p">[</span><span class="n">AA_A</span><span class="p">,</span><span class="n">AA_R</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">3</span>  <span class="c"># set the score to -3</span>
<span class="go">-3</span>

<span class="gp">julia&gt;</span> <span class="n">mysubmat</span><span class="p">[</span><span class="n">AA_A</span><span class="p">,</span><span class="n">AA_R</span><span class="p">]</span>  <span class="c"># the score is modified</span>
<span class="go">-3</span>
</pre></div>


<p>Make sure to create a copy of the original matrix when you create a matrix from a predefined matrix. In the above case, <code>BLOSUM62</code> is shared in the whole program and modification on it will affect any result that uses <code>BLOSUM62</code>.</p>
<p><code>DichotomousSubstitutionMatrix</code> is a specialized matrix for matching or mismatching substitution.  This is a preferable choice when performance is more important than flexibility because looking up score is faster than <code>SubstitutionMatrix</code>.</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">submat</span> <span class="o">=</span> <span class="n">DichotomousSubstitutionMatrix</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Bio.Align.DichotomousSubstitutionMatrix{Int64}:</span>
<span class="go">     match =  1</span>
<span class="go">  mismatch = -1</span>

<span class="gp">julia&gt;</span> <span class="n">submat</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">,</span><span class="sc">&#39;A&#39;</span><span class="p">]</span>  <span class="c"># match</span>
<span class="go">1</span>

<span class="gp">julia&gt;</span> <span class="n">submat</span><span class="p">[</span><span class="sc">&#39;A&#39;</span><span class="p">,</span><span class="sc">&#39;B&#39;</span><span class="p">]</span>  <span class="c"># mismatch</span>
<span class="go">-1</span>
</pre></div>


<p><a id='Alignment-file-formats-for-high-throughput-sequencing-1'></a></p>
<h2 id="alignment-file-formats-for-high-throughput-sequencing">Alignment file formats for high-throughput sequencing</h2>
<p>High-throughput sequencing (HTS) technologies generate a large amount of data in the form of a large number of nucleotide sequencing reads. One of the most common tasks in bioinformatics is to align these reads against known reference genomes, chromosomes, or contigs. The <code>Bio.Align</code> module provides several data formats commonly used for this kind of task.</p>
<p><a id='SAM-and-BAM-file-formats-1'></a></p>
<h3 id="sam-and-bam-file-formats">SAM and BAM file formats</h3>
<p>SAM and BAM are the most popular file formats and have the same reading and writing interface as all other formats in Bio.jl (see <a href="../reading/">Reading and writing data</a>]):</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">BAMReader</span><span class="p">,</span> <span class="s">&quot;data.bam&quot;</span><span class="p">)</span>  <span class="c"># same for SAM</span>
<span class="k">for</span> <span class="n">record</span> <span class="k">in</span> <span class="n">reader</span>
    <span class="c"># do something</span>
<span class="k">end</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p><code>SAMRecord</code> and <code>BAMRecord</code> supports the following accessors:</p>
<table>
<thead>
<tr>
<th align="left">Accessor</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>refname</code></td>
<td align="left">reference sequence name</td>
</tr>
<tr>
<td align="left"><code>leftposition</code></td>
<td align="left">1-based leftmost mapping position</td>
</tr>
<tr>
<td align="left"><code>rightposition</code></td>
<td align="left">1-based rightmost mapping position</td>
</tr>
<tr>
<td align="left"><code>refindex</code></td>
<td align="left">1-based reference sequence index (BAM only)</td>
</tr>
<tr>
<td align="left"><code>nextrefname</code></td>
<td align="left"><code>refname</code> of the mate/next read</td>
</tr>
<tr>
<td align="left"><code>nextleftposition</code></td>
<td align="left"><code>leftposition</code> of the mate/next read</td>
</tr>
<tr>
<td align="left"><code>nextrefindex</code></td>
<td align="left"><code>refindex</code> of the mate/next read (BAM only)</td>
</tr>
<tr>
<td align="left"><code>mappingquality</code></td>
<td align="left">mapping quality</td>
</tr>
<tr>
<td align="left"><code>flag</code></td>
<td align="left">bitwise flag</td>
</tr>
<tr>
<td align="left"><code>templatelength</code></td>
<td align="left">observed template length</td>
</tr>
<tr>
<td align="left"><code>seqname</code></td>
<td align="left">template name</td>
</tr>
<tr>
<td align="left"><code>cigar</code></td>
<td align="left">CIGAR string</td>
</tr>
<tr>
<td align="left"><code>cigar_rle</code></td>
<td align="left">run-length encoded CIGAR operations (BAM only)</td>
</tr>
<tr>
<td align="left"><code>alignment</code></td>
<td align="left">sequence alignment</td>
</tr>
<tr>
<td align="left"><code>sequence</code></td>
<td align="left">DNA sequence</td>
</tr>
<tr>
<td align="left"><code>seqlength</code></td>
<td align="left">DNA sequence length</td>
</tr>
<tr>
<td align="left"><code>qualities</code></td>
<td align="left">base qualities</td>
</tr>
<tr>
<td align="left"><code>[&lt;tag&gt;]</code></td>
<td align="left">value of an optional field with <code>tag</code></td>
</tr>
</tbody>
</table>
<p>16-bit flags are defined in the SAM specification as follows:</p>
<table>
<thead>
<tr>
<th align="left">Flag</th>
<th align="left">Bit</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>SAM_FLAG_PAIRED</code></td>
<td align="left"><code>0x0001</code></td>
<td align="left">template having multiple segments in sequencing</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_PROPER_PAIR</code></td>
<td align="left"><code>0x0002</code></td>
<td align="left">each segment properly aligned according to the aligner</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_UNMAP</code></td>
<td align="left"><code>0x0004</code></td>
<td align="left">segment unmapped</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_MUNMAP</code></td>
<td align="left"><code>0x0008</code></td>
<td align="left">next segment in the template unmapped</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_REVERSE</code></td>
<td align="left"><code>0x0010</code></td>
<td align="left">SEQ being reverse complemented</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_MREVERSE</code></td>
<td align="left"><code>0x0020</code></td>
<td align="left">SEQ of the next segment in the template being reverse complemented</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_READ1</code></td>
<td align="left"><code>0x0040</code></td>
<td align="left">the first segment in the template</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_READ2</code></td>
<td align="left"><code>0x0080</code></td>
<td align="left">the last segment in the template</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_SECONDARY</code></td>
<td align="left"><code>0x0100</code></td>
<td align="left">secondary alignment</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_QCFAIL</code></td>
<td align="left"><code>0x0200</code></td>
<td align="left">not passing filters, such as platform/vendor quality controls</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_DUP</code></td>
<td align="left"><code>0x0400</code></td>
<td align="left">PCR or optical duplicate</td>
</tr>
<tr>
<td align="left"><code>SAM_FLAG_SUPPLEMENTARY</code></td>
<td align="left"><code>0x0800</code></td>
<td align="left">supplementary alignment</td>
</tr>
</tbody>
</table>
<p><a id='Performance-tips-1'></a></p>
<h3 id="performance-tips">Performance tips</h3>
<p>The size of a BAM file is often extremely large. The iterator interface mentioned above allocates an object for each record and that may be a bottleneck of reading data from a BAM file. In-place reading reuses a preallocated object for every record and no memory allocation happens in reading:</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">BAMReader</span><span class="p">,</span> <span class="s">&quot;data.bam&quot;</span><span class="p">)</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">BAMRecord</span><span class="p">()</span>
<span class="k">while</span> <span class="o">!</span><span class="n">eof</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
    <span class="n">read!</span><span class="p">(</span><span class="n">reader</span><span class="p">,</span> <span class="n">record</span><span class="p">)</span>
    <span class="c"># do something</span>
<span class="k">end</span>
</pre></div>


<p>Accessing optional fields will results in type instability in Julia, which has a significant negative impact on performance. If the user knows the type of a value in advance, specifying it as a type annotation will alleviate the problem:</p>
<div class="codehilite"><pre><span></span><span class="k">for</span> <span class="n">record</span> <span class="k">in</span> <span class="n">open</span><span class="p">(</span><span class="n">BAMReader</span><span class="p">,</span> <span class="s">&quot;data.bam&quot;</span><span class="p">)</span>
    <span class="n">nm</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;NM&quot;</span><span class="p">]::</span><span class="n">UInt8</span>
    <span class="c"># do something</span>
<span class="k">end</span>
</pre></div></div>
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Copyright (c) 2016 The BioJulia Organisation<br></small>
        
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../../js/jquery-1.10.2.min.js"></script>
    <script src="../../js/bootstrap-3.0.3.min.js"></script>
    <script src="../../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '../..';
    </script>
    <script data-main="../../mkdocs/js/search.js" src="../../mkdocs/js/require.js"></script>
    <script src="../../js/base.js"></script>
    <script>
        ((window.gitter = {}).chat = {}).options = {
            room: 'BioJulia/Bio.jl'
        };
    </script>
    <script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <script src="../../assets/mathjaxhelper.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
