<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0"> 
    <meta name="author" content="The BioJulia Organisation"> 
    <link rel="canonical" href="https://biojulia.github.io/Bio.jl/man/reading/"> 
    <link rel="shortcut icon" href="../../img/favicon.ico">

    <title>IO API - Bio.jl</title>

    <link href="../../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../../css/base.css" rel="stylesheet">
    <link href="../../css/biojulia.css" rel="stylesheet">
    <link rel="stylesheet" href="../../css/highlight.css">
    <link rel="stylesheet" href="../../css/friendly.css" type="text/css">
    <link href="../../assets/Documenter.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="../..">Bio.jl</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../..">Home</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Manual <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../frontpage/">Front Page</a>
</li>

                        
                            
<li >
    <a href="../install/">Installation</a>
</li>

                        
                            
<li class="active">
    <a href="./">IO API</a>
</li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Biological Sequences</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../seq/module/">Overview</a>
</li>

        
            
<li >
    <a href="../seq/symbols/">Biological Symbols</a>
</li>

        
            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Sequence types</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../seq/sequences/sequences/">Overview</a>
</li>

        
            
<li >
    <a href="../seq/sequences/bioseq/">BioSequence</a>
</li>

        
            
<li >
    <a href="../seq/sequences/refseq/">Reference Sequences</a>
</li>

        
            
<li >
    <a href="../seq/sequences/kmer/">Nucleic acid k-mers</a>
</li>

        
    </ul>
  </li>

        
            
<li >
    <a href="../seq/search/">Searching</a>
</li>

        
            
<li >
    <a href="../seq/seqrecords/">Sequence Records</a>
</li>

        
            
<li >
    <a href="../seq/demultiplexer/">Demultiplexing</a>
</li>

        
    </ul>
  </li>

                        
                            
<li >
    <a href="../alignments/">Alignments</a>
</li>

                        
                            
<li >
    <a href="../intervals/">Intervals</a>
</li>

                        
                            
<li >
    <a href="../var/">Genetic Variation</a>
</li>

                        
                            
<li >
    <a href="../phylo/">Phylogenetic Trees</a>
</li>

                        
                            
<li >
    <a href="../structure/">Structure</a>
</li>

                        
                            
<li >
    <a href="../services/">Services</a>
</li>

                        
                            
<li >
    <a href="../util/">Util</a>
</li>

                        
                            
<li >
    <a href="../tools/">Tools</a>
</li>

                        
                            
<li >
    <a href="../">Index</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li >
                        <a href="../contributing/">Contributing</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../install/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../seq/module/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/BioJulia/Bio.jl">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
                <li><a href="https://biojulia.github.io">BioJulia Home</a></li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="first-level active"><a href="#reading-and-writing-data">Reading and writing data</a></li>
        
            <li class="second-level"><a href="#reading-by-iteration">Reading by iteration</a></li>
            
        
            <li class="second-level"><a href="#in-place-reading">In-place reading</a></li>
            
        
            <li class="second-level"><a href="#writing-data">Writing data</a></li>
            
        
            <li class="second-level"><a href="#supported-file-formats">Supported file formats</a></li>
            
                <li class="third-level"><a href="#fasta">FASTA</a></li>
            
                <li class="third-level"><a href="#fastq">FASTQ</a></li>
            
                <li class="third-level"><a href="#2bit">.2bit</a></li>
            
                <li class="third-level"><a href="#abif">ABIF</a></li>
            
                <li class="third-level"><a href="#bed">BED</a></li>
            
                <li class="third-level"><a href="#gff3">GFF3</a></li>
            
                <li class="third-level"><a href="#bigbed">bigBed</a></li>
            
                <li class="third-level"><a href="#pdb">PDB</a></li>
            
                <li class="third-level"><a href="#sam">SAM</a></li>
            
                <li class="third-level"><a href="#bam">BAM</a></li>
            
                <li class="third-level"><a href="#vcf">VCF</a></li>
            
                <li class="third-level"><a href="#bcf">BCF</a></li>
            
        
    
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<p><a id='Reading-and-writing-data-1'></a></p>
<h1 id="reading-and-writing-data">Reading and writing data</h1>
<p>Bio.jl has a unified interface for reading and writing files in a variety of formats. Reader and writer type names have a prefix of the file format. For example, files of a format <code>X</code> can be read using <code>XReader</code> and can be written using <code>XWriter</code>.  To initialize a reader/writer of <code>X</code>, you can use one of the following syntaxes:</p>
<div class="codehilite"><pre><span></span><span class="c"># reader</span>
<span class="n">open</span><span class="p">(</span><span class="o">::</span><span class="kt">Type</span><span class="p">{</span><span class="n">XReader</span><span class="p">},</span> <span class="n">filepath</span><span class="o">::</span><span class="kt">AbstractString</span><span class="p">,</span> <span class="n">args</span><span class="o">...</span><span class="p">)</span>
<span class="n">XReader</span><span class="p">(</span><span class="n">stream</span><span class="o">::</span><span class="kt">IO</span><span class="p">,</span> <span class="n">args</span><span class="o">...</span><span class="p">)</span>

<span class="c"># writer</span>
<span class="n">open</span><span class="p">(</span><span class="o">::</span><span class="kt">Type</span><span class="p">{</span><span class="n">XWriter</span><span class="p">},</span> <span class="n">filepath</span><span class="o">::</span><span class="kt">AbstractString</span><span class="p">,</span> <span class="n">args</span><span class="o">...</span><span class="p">)</span>
<span class="n">XWriter</span><span class="p">(</span><span class="n">stream</span><span class="o">::</span><span class="kt">IO</span><span class="p">,</span> <span class="n">args</span><span class="o">...</span><span class="p">)</span>
</pre></div>


<p>For example, when reading a FASTA file, a reader for the FASTA file format can be initialized as:</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span>  <span class="c"># import FASTA</span>
<span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">FASTA</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;hg38.fa&quot;</span><span class="p">)</span>
<span class="c"># do something</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p><a id='Reading-by-iteration-1'></a></p>
<h2 id="reading-by-iteration">Reading by iteration</h2>
<p>Readers in Bio.jl all read and return entries one at a time. The most convenient way to do this by iteration:</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">BED</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;input.bed&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">record</span> <span class="kp">in</span> <span class="n">reader</span>
    <span class="c"># perform some operation on entry</span>
<span class="k">end</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p><a id='In-place-reading-1'></a></p>
<h2 id="in-place-reading">In-place reading</h2>
<p>Iterating through entries in a file is convenient, but for each entry in the file, the reader must allocate, and ultimately the garbage collector must spend time to deallocate it. For performance critical applications, a separate lower level parsing interface can be used that avoid unnecessary allocation by overwriting one entry. For files with a large number of small entries, this can greatly speed up reading.</p>
<p>Instead of looping over a reader stream <code>read!</code> is called with a preallocated entry.  Some care is necessary when using this interface because <code>record</code> is completely overwritten on each iteration:</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">BED</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;input.bed&quot;</span><span class="p">)</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">BED</span><span class="o">.</span><span class="n">Record</span><span class="p">()</span>
<span class="k">while</span> <span class="o">!</span><span class="n">eof</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
    <span class="n">read!</span><span class="p">(</span><span class="n">reader</span><span class="p">,</span> <span class="n">record</span><span class="p">)</span>
    <span class="c"># perform some operation on `record`</span>
<span class="k">end</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p>Empty record types that correspond to the file format be found using <code>eltype</code>, making it easy to allocate an empty record for any reader stream:</p>
<div class="codehilite"><pre><span></span><span class="n">record</span> <span class="o">=</span> <span class="n">eltype</span><span class="p">(</span><span class="n">stream</span><span class="p">)()</span>
</pre></div>


<p><a id='Writing-data-1'></a></p>
<h2 id="writing-data">Writing data</h2>
<p>A FASTA file will be created as follows:</p>
<div class="codehilite"><pre><span></span><span class="n">writer</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">FASTAWriter</span><span class="p">,</span> <span class="s">&quot;out.fa&quot;</span><span class="p">)</span>
<span class="n">write</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">FASTASeqRecord</span><span class="p">(</span><span class="s">&quot;seq1&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;ACGTN&quot;</span><span class="p">))</span>
<span class="n">write</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">FASTASeqRecord</span><span class="p">(</span><span class="s">&quot;seq2&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TTATA&quot;</span><span class="p">,</span> <span class="s">&quot;AT rich&quot;</span><span class="p">))</span>
<span class="n">close</span><span class="p">(</span><span class="n">writer</span><span class="p">)</span>
</pre></div>


<p>Another way is using Julia's do-block syntax, which closes the data file after finished writing:</p>
<div class="codehilite"><pre><span></span><span class="n">open</span><span class="p">(</span><span class="n">FASTAWriter</span><span class="p">,</span> <span class="s">&quot;out.fa&quot;</span><span class="p">)</span> <span class="k">do</span> <span class="n">writer</span>
    <span class="n">write</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">FASTASeqRecord</span><span class="p">(</span><span class="s">&quot;seq1&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;ACGTN&quot;</span><span class="p">))</span>
    <span class="n">write</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">FASTASeqRecord</span><span class="p">(</span><span class="s">&quot;seq2&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TTATA&quot;</span><span class="p">,</span> <span class="s">&quot;AT rich&quot;</span><span class="p">))</span>
<span class="k">end</span>
</pre></div>


<p><a id='Supported-file-formats-1'></a></p>
<h2 id="supported-file-formats">Supported file formats</h2>
<p>The following table summarizes supported file formats.</p>
<table>
<thead>
<tr>
<th align="left">File format</th>
<th align="left">Prefix</th>
<th align="left">Module</th>
<th align="left">Specification</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">FASTA</td>
<td align="left"><code>FASTA</code></td>
<td align="left"><code>Bio.Seq</code></td>
<td align="left"><a href="https://en.wikipedia.org/wiki/FASTA_format">https://en.wikipedia.org/wiki/FASTA_format</a></td>
</tr>
<tr>
<td align="left">FASTQ</td>
<td align="left"><code>FASTQ</code></td>
<td align="left"><code>Bio.Seq</code></td>
<td align="left"><a href="https://en.wikipedia.org/wiki/FASTQ_format">https://en.wikipedia.org/wiki/FASTQ_format</a></td>
</tr>
<tr>
<td align="left">.2bit</td>
<td align="left"><code>TwoBit</code></td>
<td align="left"><code>Bio.Seq</code></td>
<td align="left"><a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format7">http://genome.ucsc.edu/FAQ/FAQformat.html#format7</a></td>
</tr>
<tr>
<td align="left">BED</td>
<td align="left"><code>BED</code></td>
<td align="left"><code>Bio.Intervals</code></td>
<td align="left"><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">https://genome.ucsc.edu/FAQ/FAQformat.html#format1</a></td>
</tr>
<tr>
<td align="left">GFF3</td>
<td align="left"><code>GFF3</code></td>
<td align="left"><code>Bio.Intervals</code></td>
<td align="left"><a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md</a></td>
</tr>
<tr>
<td align="left">bigBed</td>
<td align="left"><code>BigBed</code></td>
<td align="left"><code>Bio.Intervals</code></td>
<td align="left"><a href="https://doi.org/10.1093/bioinformatics/btq351">https://doi.org/10.1093/bioinformatics/btq351</a></td>
</tr>
<tr>
<td align="left">PDB</td>
<td align="left"><code>PDB</code></td>
<td align="left"><code>Bio.Structure</code></td>
<td align="left"><a href="http://www.wwpdb.org/documentation/file-format-content/format33/v3.3.html">http://www.wwpdb.org/documentation/file-format-content/format33/v3.3.html</a></td>
</tr>
<tr>
<td align="left">SAM</td>
<td align="left"><code>SAM</code></td>
<td align="left"><code>Bio.Align</code></td>
<td align="left"><a href="https://samtools.github.io/hts-specs/SAMv1.pdf">https://samtools.github.io/hts-specs/SAMv1.pdf</a></td>
</tr>
<tr>
<td align="left">BAM</td>
<td align="left"><code>BAM</code></td>
<td align="left"><code>Bio.Align</code></td>
<td align="left"><a href="https://samtools.github.io/hts-specs/SAMv1.pdf">https://samtools.github.io/hts-specs/SAMv1.pdf</a></td>
</tr>
<tr>
<td align="left">VCF</td>
<td align="left"><code>VCF</code></td>
<td align="left"><code>Bio.Var</code></td>
<td align="left"><a href="https://samtools.github.io/hts-specs/VCFv4.3.pdf">https://samtools.github.io/hts-specs/VCFv4.3.pdf</a></td>
</tr>
<tr>
<td align="left">BCF</td>
<td align="left"><code>BCF</code></td>
<td align="left"><code>Bio.Var</code></td>
<td align="left"><a href="https://samtools.github.io/hts-specs/VCFv4.3.pdf">https://samtools.github.io/hts-specs/VCFv4.3.pdf</a></td>
</tr>
</tbody>
</table>
<p><a id='FASTA-1'></a></p>
<h3 id="fasta">FASTA</h3>
<ul>
<li>Reader type: <code>FASTA.Reader</code></li>
<li>Writer type: <code>FASTA.Writer</code></li>
<li>Element type: <code>FASTA.Record</code></li>
</ul>
<p>FASTA is a text-based file format for representing biological sequences. A FASTA file stores a list of records with identifier, description, and sequence. The template of a sequence record is:</p>
<div class="codehilite"><pre><span></span>&gt;{identifier} {description}?
{sequence}
</pre></div>


<p>Here is an example of a chromosomal sequence:</p>
<div class="codehilite"><pre><span></span>&gt;chrI chromosome 1
CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACC
CACACACACACATCCTAACACTACCCTAACACAGCCCTAATCTAACCCTG
</pre></div>


<p>Usually sequence records will be read sequentially from a file by iteration. But if the FASTA file has an auxiliary index file formatted in <a href="http://www.htslib.org/doc/faidx.html">fai</a>, the reader supports random access to FASTA records, which would be useful when accessing specific parts of a huge genome sequence:</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">FASTAReader</span><span class="p">,</span> <span class="s">&quot;sacCer.fa&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="s">&quot;sacCer.fa.fai&quot;</span><span class="p">)</span>
<span class="n">chrIV</span> <span class="o">=</span> <span class="n">reader</span><span class="p">[</span><span class="s">&quot;chrIV&quot;</span><span class="p">]</span>  <span class="c"># directly read chromosome 4</span>
</pre></div>


<p><a id='Bio.Seq.FASTA.Reader' href='#Bio.Seq.FASTA.Reader'>#</a>
<strong><code>Bio.Seq.FASTA.Reader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">FASTA</span><span class="p">.</span><span class="nc">Reader</span><span class="o">(</span><span class="nt">input</span><span class="p">::</span><span class="nd">IO</span><span class="o">;</span> <span class="nt">index</span><span class="o">=</span><span class="nt">nothing</span><span class="o">)</span>
</pre></div>


<p>Create a data reader of the FASTA file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
<li><code>index=nothing</code>: filepath to a random access index (currently <em>fai</em> is supported)</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/reader.jl#L13-L21' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.FASTA.Writer' href='#Bio.Seq.FASTA.Writer'>#</a>
<strong><code>Bio.Seq.FASTA.Writer</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">FASTA</span><span class="p">.</span><span class="nc">Writer</span><span class="o">(</span><span class="nt">output</span><span class="p">::</span><span class="nd">IO</span><span class="o">;</span> <span class="nt">width</span><span class="o">=</span><span class="nt">70</span><span class="o">)</span>
</pre></div>


<p>Create a data writer of the FASTA file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>width=70</code>: wrapping width of sequence characters</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/writer.jl#L4-L12' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.FASTA.Record' href='#Bio.Seq.FASTA.Record'>#</a>
<strong><code>Bio.Seq.FASTA.Record</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>FASTA.Record()
</pre></div>


<p>Create an unfilled FASTA record.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L14-L18' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>FASTA.Record(data::Vector{UInt8})
</pre></div>


<p>Create a FASTA record object from <code>data</code>.</p>
<p>This function verifies and indexes fields for accessors. Note that the ownership of <code>data</code> is transferred to a new record object.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L23-L30' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>FASTA.Record(str::AbstractString)
</pre></div>


<p>Create a FASTA record object from <code>str</code>.</p>
<p>This function verifies and indexes fields for accessors.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L41-L47' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>FASTA.Record(identifier, sequence)
</pre></div>


<p>Create a FASTA record object from <code>identifier</code> and <code>sequence</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L60-L64' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>FASTA.Record(identifier, description, sequence)
</pre></div>


<p>Create a FASTA record object from <code>identifier</code>, <code>description</code> and <code>sequence</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L69-L73' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.FASTA.identifier' href='#Bio.Seq.FASTA.identifier'>#</a>
<strong><code>Bio.Seq.FASTA.identifier</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>identifier(record::Record)::String
</pre></div>


<p>Get the sequence identifier of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L153-L157' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.FASTA.description' href='#Bio.Seq.FASTA.description'>#</a>
<strong><code>Bio.Seq.FASTA.description</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>description(record::Record)::String
</pre></div>


<p>Get the description of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L178-L182' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.FASTA.sequence' href='#Bio.Seq.FASTA.sequence'>#</a>
<strong><code>Bio.Seq.FASTA.sequence</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>sequence(::Type{S}, record::Record, [part::UnitRange{Int}])::S
</pre></div>


<p>Get the sequence of <code>record</code>.</p>
<p><code>S</code> can be either a subtype of <code>Bio.Seq.Sequence</code> or <code>String</code>. If <code>part</code> argument is given, it returns the specified part of the sequence.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L195-L202' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>sequence(record::Record, [part::UnitRange{Int}])
</pre></div>


<p>Get the sequence of <code>record</code>.</p>
<p>This function infers the sequence type from the data. When it is wrong or unreliable, use <code>sequence(::Type{S}, record::Record)</code>.  If <code>part</code> argument is given, it returns the specified part of the sequence.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fasta/record.jl#L220-L228' class='documenter-source'>source</a><br></p>
<p><a id='FASTQ-1'></a></p>
<h3 id="fastq">FASTQ</h3>
<ul>
<li>Reader type: <code>FASTQReader{S&lt;:Sequence}</code></li>
<li>Writer type: <code>FASTQWriter{T&lt;:IO}</code></li>
<li>Element type: <code>SeqRecord{S,FASTQMetadata}</code> (alias: <code>FASTQSeqRecord{S}</code>)</li>
</ul>
<p>FASTQ is a text-based file format for representing DNA sequences along with qualities for each base. A FASTQ file stores a list of sequence records in the following format:</p>
<div class="codehilite"><pre><span></span>@{name} {description}?
{sequence}
+
{qualities}
</pre></div>


<p>Here is an example of one record from a FASTQ file:</p>
<div class="codehilite"><pre><span></span>@FSRRS4401BE7HA
tcagTTAAGATGGGAT
+
###EEEEEEEEE##E#
</pre></div>


<p>To read a file containing such records, one could use:</p>
<div class="codehilite"><pre><span></span><span class="c"># The default base quality encoding is Sanger.</span>
<span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">FASTQReader</span><span class="p">,</span> <span class="s">&quot;reads.fastq&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">record</span> <span class="kp">in</span> <span class="n">reader</span>
    <span class="c"># do something</span>
<span class="k">end</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>

<span class="c"># If performance is important, in-place reading will be much faster.</span>
<span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">FASTQReader</span><span class="p">,</span> <span class="s">&quot;reads.fastq&quot;</span><span class="p">)</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">FASTQSeqRecord</span><span class="p">{</span><span class="n">DNASequence</span><span class="p">}()</span>
<span class="k">while</span> <span class="o">!</span><span class="n">eof</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
    <span class="n">read!</span><span class="p">(</span><span class="n">reader</span><span class="p">,</span> <span class="n">record</span><span class="p">)</span>
    <span class="c"># do something</span>
<span class="k">end</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p><a id='Bio.Seq.FASTQReader' href='#Bio.Seq.FASTQReader'>#</a>
<strong><code>Bio.Seq.FASTQReader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>FASTQReader(input::IO, quality_encoding=:sanger, fill_ambiguous=nothing)
</pre></div>


<p>Create a data reader of the FASTQ file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
<li><code>quality_encoding=:sanger</code>: encoding of base qualities; see the following table for available values</li>
<li><code>fill_ambiguous=nothing</code>: fill ambiguous nucleotides with the given nucleotide</li>
</ul>
<table>
<thead>
<tr>
<th align="left">Quality encoding</th>
<th align="left">Symbol</th>
<th align="center">ASCII offset</th>
<th align="center">Quality range</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Sanger</td>
<td align="left"><code>:sanger</code></td>
<td align="center">+33</td>
<td align="center">0-93</td>
</tr>
<tr>
<td align="left">Solexa</td>
<td align="left"><code>:solexa</code></td>
<td align="center">+64</td>
<td align="center">-5-62</td>
</tr>
<tr>
<td align="left">Illumina 1.3+</td>
<td align="left"><code>:illumina13</code></td>
<td align="center">+64</td>
<td align="center">0-62</td>
</tr>
<tr>
<td align="left">Illumina 1.5+</td>
<td align="left"><code>:illumina15</code></td>
<td align="center">+64</td>
<td align="center">2-62</td>
</tr>
<tr>
<td align="left">Illumina 1.8+</td>
<td align="left"><code>:illumina18</code></td>
<td align="center">+33</td>
<td align="center">0-93</td>
</tr>
</tbody>
</table>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fastq/reader.jl#L4-L21' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.FASTQWriter' href='#Bio.Seq.FASTQWriter'>#</a>
<strong><code>Bio.Seq.FASTQWriter</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">FASTQWriter</span><span class="o">(</span><span class="nt">output</span><span class="p">::</span><span class="nd">IO</span><span class="o">;</span> <span class="nt">quality_header</span><span class="o">=</span><span class="nt">false</span><span class="o">,</span> <span class="nt">quality_encoding</span><span class="o">=</span><span class="p">:</span><span class="nd">sanger</span><span class="o">)</span>
</pre></div>


<p>Create a data writer of the FASTQ file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>quality_header=false</code>: output the title line at the third line just after '+'</li>
<li><code>quality_encoding=:sanger</code>: encoding of base qualities; see <code>FASTQReader</code> for available values</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/fastq/writer.jl#L4-L13' class='documenter-source'>source</a><br></p>
<p><a id='.2bit-1'></a></p>
<h3 id="2bit">.2bit</h3>
<ul>
<li>Reader type: <code>TwoBit.Reader{T&lt;:IO}</code></li>
<li>Writer type: <code>TwoBit.Writer{T&lt;:IO}</code></li>
<li>Element type: <code>TwoBit.Record</code></li>
</ul>
<p>.2bit is a binary file format designed for storing a genome consists of multiple chromosomal sequences. The reading speed is often an order of magnitude faster than that of FASTA and the file size is smaller. However, since the .2bit file format is specialized for genomic sequences, it cannot store either RNA or amino acid sequences.</p>
<p>Like FASTA, the .2bit reader supports random access using an index included in the header section of a .2bit file:</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">TwoBit</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;sacCer.2bit&quot;</span><span class="p">)</span>  <span class="c"># load a random access index in the header</span>
<span class="n">chrIV</span> <span class="o">=</span> <span class="n">reader</span><span class="p">[</span><span class="s">&quot;chrIV&quot;</span><span class="p">]</span>                      <span class="c"># directly read chromosome 4</span>
</pre></div>


<p><a id='Bio.Seq.TwoBit.Reader' href='#Bio.Seq.TwoBit.Reader'>#</a>
<strong><code>Bio.Seq.TwoBit.Reader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>TwoBit.Reader(input::IO)
</pre></div>


<p>Create a data reader of the 2bit file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/twobit/reader.jl#L15-L22' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.TwoBit.Writer' href='#Bio.Seq.TwoBit.Writer'>#</a>
<strong><code>Bio.Seq.TwoBit.Writer</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>TwoBitWriter(output::IO, names::AbstractVector)
</pre></div>


<p>Create a data writer of the 2bit file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>names</code>: a vector of sequence names written to <code>output</code></li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/twobit/writer.jl#L13-L21' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.TwoBit.Record' href='#Bio.Seq.TwoBit.Record'>#</a>
<strong><code>Bio.Seq.TwoBit.Record</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>TwoBit.Record()
</pre></div>


<p>Create an unfilled 2bit record.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/twobit/record.jl#L17-L21' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.TwoBit.seqnames' href='#Bio.Seq.TwoBit.seqnames'>#</a>
<strong><code>Bio.Seq.TwoBit.seqnames</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>seqnames(reader::Reader)::Vector{String}
</pre></div>


<p>Get the sequence names.</p>
<p>Sequences are stored in this order.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/twobit/reader.jl#L54-L60' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.TwoBit.sequence' href='#Bio.Seq.TwoBit.sequence'>#</a>
<strong><code>Bio.Seq.TwoBit.sequence</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>sequence([::Type{S},] record::Record)::S
</pre></div>


<p>Get the sequence of <code>record</code> as <code>S</code>.</p>
<p><code>S</code> can be either <code>Bio.Seq.ReferenceSequence</code> or <code>Bio.Seq.DNASequence</code>. If <code>S</code> is omitted, the default type is <code>Bio.Seq.ReferenceSequence</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/twobit/record.jl#L96-L103' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.TwoBit.maskedblocks' href='#Bio.Seq.TwoBit.maskedblocks'>#</a>
<strong><code>Bio.Seq.TwoBit.maskedblocks</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>maskedblocks(record::Record)::Vector{UnitRange{Int}}
</pre></div>


<p>Get the masked blocks.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/seq/twobit/record.jl#L136-L140' class='documenter-source'>source</a><br></p>
<p><a id='ABIF-1'></a></p>
<h3 id="abif">ABIF</h3>
<ul>
<li>Reader type: <code>AbifReader{T&lt;:IO}</code></li>
</ul>
<p>ABIF is a binary file format for storing data produced by sequencers, those developed by Applied Biosystems, Inc. When the file is opened, we save all the existing tags, so we can read only the tags that are needed.</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">AbifReader</span><span class="p">,</span> <span class="s">&quot;3100.ab1&quot;</span><span class="p">)</span>       <span class="c"># load a random access</span>
<span class="n">data</span>   <span class="o">=</span> <span class="n">reader</span><span class="p">[</span><span class="s">&quot;DATA&quot;</span><span class="p">]</span>                     <span class="c"># directly read all existing `DATA` Tags</span>
<span class="n">data</span>   <span class="o">=</span> <span class="n">reader</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>                          <span class="c"># directly read Tag at index</span>
<span class="n">data</span>  <span class="o">=</span> <span class="n">tags</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>                       <span class="c"># return all existing tags</span>

<span class="c"># iterator by all tags</span>
<span class="k">for</span> <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="kp">in</span> <span class="n">getindex</span><span class="p">(</span><span class="n">reader</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
<span class="k">end</span>
</pre></div>


<div class="codehilite"><pre><span></span>Bio.Seq.AbifReader
Bio.Seq.tags
Bio.Seq.elements
</pre></div>


<p><a id='BED-1'></a></p>
<h3 id="bed">BED</h3>
<ul>
<li>Reader type: <code>BED.Reader</code></li>
<li>Writer type: <code>BED.Writer</code></li>
<li>Element type: <code>BED.Record</code></li>
</ul>
<p>BED is a text-based file format for representing genomic annotations like genes, transcripts, and so on. A BED file has tab-delimited and variable-length fields; the first three fields denoting a genomic interval are mandatory.</p>
<p>This is an example of RNA transcripts:</p>
<div class="codehilite"><pre><span></span>chr9    68331023    68424451    NM_015110   0   +
chr9    68456943    68486659    NM_001206   0   -
</pre></div>


<p><a id='Bio.Intervals.BED.Reader' href='#Bio.Intervals.BED.Reader'>#</a>
<strong><code>Bio.Intervals.BED.Reader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BED.Reader(input::IO)
</pre></div>


<p>Create a data reader of the BED file format.</p>
<p><strong>Arguments:</strong></p>
<ul>
<li><code>input</code>: data source</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/reader.jl#L12-L19' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.Writer' href='#Bio.Intervals.BED.Writer'>#</a>
<strong><code>Bio.Intervals.BED.Writer</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BED.Writer(output::IO)
</pre></div>


<p>Create a data writer of the BED file format.</p>
<p><strong>Arguments:</strong></p>
<ul>
<li><code>output</code>: data sink</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/writer.jl#L4-L11' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.Record' href='#Bio.Intervals.BED.Record'>#</a>
<strong><code>Bio.Intervals.BED.Record</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BED.Record()
</pre></div>


<p>Create an unfilled BED record.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L25-L29' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>BED.Record(data::Vector{UInt8})
</pre></div>


<p>Create a BED record object from <code>data</code>.</p>
<p>This function verifies and indexes fields for accessors. Note that the ownership of <code>data</code> is transferred to a new record object.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L41-L48' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>BED.Record(str::AbstractString)
</pre></div>


<p>Create a BED record object from <code>str</code>.</p>
<p>This function verifies and indexes fields for accessors.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L66-L72' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.chrom' href='#Bio.Intervals.BED.chrom'>#</a>
<strong><code>Bio.Intervals.BED.chrom</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>chrom(record::Record)::String
</pre></div>


<p>Get the chromosome name of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L202-L206' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.chromstart' href='#Bio.Intervals.BED.chromstart'>#</a>
<strong><code>Bio.Intervals.BED.chromstart</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>chromstart(record::Record)::Int
</pre></div>


<p>Get the starting position of <code>record</code>.</p>
<p>Note that the first base is numbered 1.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L224-L230' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.chromend' href='#Bio.Intervals.BED.chromend'>#</a>
<strong><code>Bio.Intervals.BED.chromend</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>chromend(record::Record)::Int
</pre></div>


<p>Get the end position of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L248-L252' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.name' href='#Bio.Intervals.BED.name'>#</a>
<strong><code>Bio.Intervals.BED.name</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>name(record::Record)::String
</pre></div>


<p>Get the name of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L270-L274' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.score' href='#Bio.Intervals.BED.score'>#</a>
<strong><code>Bio.Intervals.BED.score</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>score(record::Record)::Int
</pre></div>


<p>Get the score between 0 and 1000.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L287-L291' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.strand' href='#Bio.Intervals.BED.strand'>#</a>
<strong><code>Bio.Intervals.BED.strand</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>strand(record::Record)::Bio.Intervals.Strand
</pre></div>


<p>Get the strand of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L304-L308' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.thickstart' href='#Bio.Intervals.BED.thickstart'>#</a>
<strong><code>Bio.Intervals.BED.thickstart</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>thickstart(record::Record)::Int
</pre></div>


<p>Get the starting position at which <code>record</code> is drawn thickly.</p>
<p>Note that the first base is numbered 1.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L325-L331' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.thickend' href='#Bio.Intervals.BED.thickend'>#</a>
<strong><code>Bio.Intervals.BED.thickend</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>thickend(record::Record)::Int
</pre></div>


<p>Get the end position at which <code>record</code> is drawn thickly.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L344-L348' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.itemrgb' href='#Bio.Intervals.BED.itemrgb'>#</a>
<strong><code>Bio.Intervals.BED.itemrgb</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>itemrgb(record::Record)::ColorTypes.RGB
</pre></div>


<p>Get the RGB value of <code>record</code>.</p>
<p>The return type is defined in <a href="https://github.com/JuliaGraphics/ColorTypes.jl">ColorTypes.jl</a>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L361-L367' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.blockcount' href='#Bio.Intervals.BED.blockcount'>#</a>
<strong><code>Bio.Intervals.BED.blockcount</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>blockcount(record::Record)::Int
</pre></div>


<p>Get the number of blocks (exons) in <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L417-L421' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.blocksizes' href='#Bio.Intervals.BED.blocksizes'>#</a>
<strong><code>Bio.Intervals.BED.blocksizes</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>blocksizes(record::Record)::Vector{Int}
</pre></div>


<p>Get the block (exon) sizes of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L434-L438' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BED.blockstarts' href='#Bio.Intervals.BED.blockstarts'>#</a>
<strong><code>Bio.Intervals.BED.blockstarts</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>blockstarts(record::Record)::Vector{Int}
</pre></div>


<p>Get the block (exon) starts of <code>record</code>.</p>
<p>Note that the first base is numbered 1.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bed/record.jl#L451-L457' class='documenter-source'>source</a><br></p>
<p><a id='GFF3-1'></a></p>
<h3 id="gff3">GFF3</h3>
<ul>
<li>Reader type: <code>Bio.Intervals.GFF3.Reader</code></li>
<li>Writer type: <code>Bio.Intervals.GFF3.Writer</code></li>
<li>Element type: <code>Bio.Intervals.GFF3.Record</code></li>
</ul>
<p>GFF3 is a text-based file format for representing genomic annotations. The major difference from BED is that is GFF3 is more structured and can include sequences in the FASTA file format.</p>
<p><a id='Bio.Intervals.GFF3.Reader' href='#Bio.Intervals.GFF3.Reader'>#</a>
<strong><code>Bio.Intervals.GFF3.Reader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">GFF3Reader</span><span class="o">(</span><span class="nt">input</span><span class="p">::</span><span class="nd">IO</span><span class="o">;</span> <span class="nt">save_directives</span><span class="p">::</span><span class="nd">Bool</span><span class="o">=</span><span class="nt">false</span><span class="o">)</span>
</pre></div>


<p>Create a reader for data in GFF3 format.</p>
<p><strong>Arguments:</strong></p>
<ul>
<li><code>input</code>: data source</li>
<li><code>save_directives=false</code>: if true, store directive lines, which can be accessed with the <code>directives</code> function</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/gff3/reader.jl#L4-L13' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.GFF3.Writer' href='#Bio.Intervals.GFF3.Writer'>#</a>
<strong><code>Bio.Intervals.GFF3.Writer</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>GFF3.Writer(output::IO)
</pre></div>


<p>Create a data writer of the GFF3 file format.</p>
<p><strong>Arguments:</strong></p>
<ul>
<li><code>output</code>: data sink</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/gff3/writer.jl#L4-L11' class='documenter-source'>source</a><br></p>
<p><a id='bigBed-1'></a></p>
<h3 id="bigbed">bigBed</h3>
<ul>
<li>Reader type: <code>BigBedReader</code></li>
<li>Writer type: <code>BigBedWriter{T&lt;:IO}</code></li>
<li>Element type: <code>Interval{BEDMetadata}</code> (alias: <code>BEDInterval</code>)</li>
</ul>
<p>BigBed is a binary file format for representing genomic annotations and often created from BED files. The bigBed files are indexed to quickly fetch specific regions.</p>
<p><a id='Bio.Intervals.BigBedReader' href='#Bio.Intervals.BigBedReader'>#</a>
<strong><code>Bio.Intervals.BigBedReader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BigBedReader(input::IO)
</pre></div>


<p>Create a data reader of the BigBed file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bbi/reader.jl#L4-L11' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Intervals.BigBedWriter' href='#Bio.Intervals.BigBedWriter'>#</a>
<strong><code>Bio.Intervals.BigBedWriter</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BigBedWriter(output::IO)
</pre></div>


<p>Create a data writer of the BigBed file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/intervals/bbi/writer.jl#L4-L11' class='documenter-source'>source</a><br></p>
<p><a id='PDB-1'></a></p>
<h3 id="pdb">PDB</h3>
<p>PDB is a text-based file format for representing 3D macromolecular structures. This has different reader interfaces from other file formats. Please consult the <a href="../structure/">Bio.Structure</a> chapter for details.</p>
<p><a id='SAM-1'></a></p>
<h3 id="sam">SAM</h3>
<ul>
<li>Reader type: <code>SAMReader</code></li>
<li>Writer type: <code>SAMWriter{T&lt;:IO}</code></li>
<li>Element type: <code>SAMRecord</code></li>
</ul>
<p>SAM is a text-based file format for representing sequence alignments.</p>
<p><a id='Bio.Align.SAMReader' href='#Bio.Align.SAMReader'>#</a>
<strong><code>Bio.Align.SAMReader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>SAMReader(input::IO)
</pre></div>


<p>Create a data reader of the SAM file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/align/hts/sam/reader.jl#L4-L11' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.SAMWriter' href='#Bio.Align.SAMWriter'>#</a>
<strong><code>Bio.Align.SAMWriter</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>SAMWriter(output::IO, header::SAMHeader=SAMHeader())
</pre></div>


<p>Create a data writer of the SAM file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>header=SAMHeader()</code>: SAM header object</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/align/hts/sam/writer.jl#L4-L12' class='documenter-source'>source</a><br></p>
<p><a id='BAM-1'></a></p>
<h3 id="bam">BAM</h3>
<ul>
<li>Reader type: <code>BAMReader</code></li>
<li>Writer type: <code>BAMWriter</code></li>
<li>Element type: <code>BAMRecord</code></li>
</ul>
<p>BAM is a binary counterpart of the SAM file format.</p>
<p>When writing data in the BAM file format, the underlying output stream needs to be wrapped with a <code>BGZFStream</code> object provided from <a href="https://github.com/BioJulia/BGZFStreams.jl">BGZFStreams.jl</a>.</p>
<p><a id='Bio.Align.BAMReader' href='#Bio.Align.BAMReader'>#</a>
<strong><code>Bio.Align.BAMReader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">BAMReader</span><span class="o">(</span><span class="nt">input</span><span class="p">::</span><span class="nd">IO</span><span class="o">;</span> <span class="nt">index</span><span class="o">=</span><span class="nt">nothing</span><span class="o">)</span>
</pre></div>


<p>Create a data reader of the BAM file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
<li><code>index=nothing</code>: filepath to a random access index (currently <em>bai</em> is Supported)</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/align/hts/bam/reader.jl#L4-L12' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Align.BAMWriter' href='#Bio.Align.BAMWriter'>#</a>
<strong><code>Bio.Align.BAMWriter</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BAMWriter(output::BGZFStream, header::SAMHeader)
</pre></div>


<p>Create a data writer of the BAM file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>header</code>: SAM header object</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/align/hts/bam/writer.jl#L4-L12' class='documenter-source'>source</a><br></p>
<p><a id='VCF-1'></a></p>
<h3 id="vcf">VCF</h3>
<ul>
<li>Reader type: <code>VCF.Reader</code></li>
<li>Writer type: <code>VCF.Writer{T&lt;:IO}</code></li>
<li>Element type: <code>VCF.Record</code></li>
</ul>
<p>VCF is a text-based file format for representing genetic variations.</p>
<p><a id='Bio.Var.VCF.Reader' href='#Bio.Var.VCF.Reader'>#</a>
<strong><code>Bio.Var.VCF.Reader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>VCF.Reader(input::IO)
</pre></div>


<p>Create a data reader of the VCF file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/reader.jl#L16-L23' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.Writer' href='#Bio.Var.VCF.Writer'>#</a>
<strong><code>Bio.Var.VCF.Writer</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>VCF.Writer(output::IO, header::VCF.Header)
</pre></div>


<p>Create a data writer of the VCF file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>header</code>: VCF header object</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/writer.jl#L8-L16' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.Record' href='#Bio.Var.VCF.Record'>#</a>
<strong><code>Bio.Var.VCF.Record</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>VCF.Record()
</pre></div>


<p>Create an unfilled VCF record.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L21-L25' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>VCF.Record(data::Vector{UInt8})
</pre></div>


<p>Create a VCF object from <code>data</code> containing a VCF record. This function verifies the format and indexes fields for accessors. Note that the ownership of <code>data</code> is transferred to a new record.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L36-L42' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>VCF.Record(str::AbstractString)
</pre></div>


<p>Create a VCF object from <code>str</code> containing a VCF record. This function verifies the format and indexes fields for accessors.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L59-L64' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.chrom' href='#Bio.Var.VCF.chrom'>#</a>
<strong><code>Bio.Var.VCF.chrom</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>chrom(record::Record)::String
</pre></div>


<p>Get the chromosome name of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L339-L343' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.pos' href='#Bio.Var.VCF.pos'>#</a>
<strong><code>Bio.Var.VCF.pos</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>pos(record::Record)::Int
</pre></div>


<p>Get the reference position of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L356-L360' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.id' href='#Bio.Var.VCF.id'>#</a>
<strong><code>Bio.Var.VCF.id</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>id(record::Record)::Vector{String}
</pre></div>


<p>Get the identifiers of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L374-L378' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.ref' href='#Bio.Var.VCF.ref'>#</a>
<strong><code>Bio.Var.VCF.ref</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>ref(record::Record)::String
</pre></div>


<p>Get the reference bases of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L391-L395' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.alt' href='#Bio.Var.VCF.alt'>#</a>
<strong><code>Bio.Var.VCF.alt</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alt(record::Record)::Vector{String}
</pre></div>


<p>Get the alternate bases of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L408-L412' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.qual' href='#Bio.Var.VCF.qual'>#</a>
<strong><code>Bio.Var.VCF.qual</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>qual(record::Record)::Float64
</pre></div>


<p>Get the quality score of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L425-L429' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.filter' href='#Bio.Var.VCF.filter'>#</a>
<strong><code>Bio.Var.VCF.filter</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>filter(record::Record)::Vector{String}
</pre></div>


<p>Get the filter status of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L443-L447' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.info' href='#Bio.Var.VCF.info'>#</a>
<strong><code>Bio.Var.VCF.info</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>info(record::Record)::Vector{Pair{String,String}}
</pre></div>


<p>Get the additional information of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L460-L464' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>info(record::Record, key::String)::String
</pre></div>


<p>Get the additional information of <code>record</code> with <code>key</code>. Keys without corresponding values return an empty string.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L482-L487' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.VCF.format' href='#Bio.Var.VCF.format'>#</a>
<strong><code>Bio.Var.VCF.format</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>format(record::Record)::Vector{String}
</pre></div>


<p>Get the genotype format of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/vcf/record.jl#L543-L547' class='documenter-source'>source</a><br></p>
<p><a id='BCF-1'></a></p>
<h3 id="bcf">BCF</h3>
<ul>
<li>Reader type: <code>BCF.Reader{T&lt;:IO}</code></li>
<li>Writer type: <code>BCF.Writer{T&lt;:IO}</code></li>
<li>Element type: <code>BCF.Record</code></li>
</ul>
<p>BCF is a binary counterpart of the VCF file format.</p>
<p><a id='Bio.Var.BCF.Reader' href='#Bio.Var.BCF.Reader'>#</a>
<strong><code>Bio.Var.BCF.Reader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BCF.Reader(input::IO)
</pre></div>


<p>Create a data reader of the BCF file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/reader.jl#L10-L17' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.Writer' href='#Bio.Var.BCF.Writer'>#</a>
<strong><code>Bio.Var.BCF.Writer</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BCF.Writer(output::IO, header::VCF.Header)
</pre></div>


<p>Create a data writer of the BCF file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>header</code>: VCF header object</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/writer.jl#L8-L16' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.Record' href='#Bio.Var.BCF.Record'>#</a>
<strong><code>Bio.Var.BCF.Record</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BCF.Record()
</pre></div>


<p>Create an unfilled BCF record.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L13-L17' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.chrom' href='#Bio.Var.BCF.chrom'>#</a>
<strong><code>Bio.Var.BCF.chrom</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>chrom(record::Record)::Int
</pre></div>


<p>Get the chromosome index of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L201-L205' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.pos' href='#Bio.Var.BCF.pos'>#</a>
<strong><code>Bio.Var.BCF.pos</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>pos(record::Record)::Int
</pre></div>


<p>Get the reference position of <code>record</code>.</p>
<p>Note that the position of the first base is 1 (i.e. 1-based coordinate).</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L211-L217' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.rlen' href='#Bio.Var.BCF.rlen'>#</a>
<strong><code>Bio.Var.BCF.rlen</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>rlen(record::Record)::Int
</pre></div>


<p>Get the length of <code>record</code> projected onto the reference sequence.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L223-L227' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.qual' href='#Bio.Var.BCF.qual'>#</a>
<strong><code>Bio.Var.BCF.qual</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>qual(record::Record)::Float32
</pre></div>


<p>Get the quality score of <code>record</code>.</p>
<p>Note that <code>0x7F800001</code> (signaling NaN) is interpreted as a missing value.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L233-L239' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.ref' href='#Bio.Var.BCF.ref'>#</a>
<strong><code>Bio.Var.BCF.ref</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>ref(record::Record)::String
</pre></div>


<p>Get the reference bases of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L272-L276' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.alt' href='#Bio.Var.BCF.alt'>#</a>
<strong><code>Bio.Var.BCF.alt</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alt(record::Record)::Vector{String}
</pre></div>


<p>Get the alternate bases of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L286-L290' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.filter' href='#Bio.Var.BCF.filter'>#</a>
<strong><code>Bio.Var.BCF.filter</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>filter(record::Record)::Vector{Int}
</pre></div>


<p>Get the filter indexes of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L308-L312' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.info' href='#Bio.Var.BCF.info'>#</a>
<strong><code>Bio.Var.BCF.info</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>info(record::Record, [simplify::Bool=true])::Vector{Tuple{Int,Any}}
</pre></div>


<p>Get the additional information of <code>record</code>.</p>
<p>When <code>simplify</code> is <code>true</code>, a vector with a single element is converted to the element itself and an empty vector of the void type is converted to <code>nothing</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L325-L332' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>info(record::Record, key::Integer, [simplify::Bool=true])
</pre></div>


<p>Get the additional information of <code>record</code> with <code>key</code>.</p>
<p>When <code>simplify</code> is <code>true</code>, a vector with a single element is converted to the element itself and an empty vector of the void type is converted to <code>nothing</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L360-L367' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.BCF.genotype' href='#Bio.Var.BCF.genotype'>#</a>
<strong><code>Bio.Var.BCF.genotype</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>genotype(record::Record)::Vector{Tuple{Int,Vector{Any}}}
</pre></div>


<p>Get the genotypes of <code>record</code>.</p>
<p>BCF genotypes are encoded by field, not by sample like VCF.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/4b9dbb1fae2906db7eed646f6028ac68cdf8f421/src/var/bcf/record.jl#L398-L404' class='documenter-source'>source</a><br></p></div>
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Copyright (c) 2016 The BioJulia Organisation<br></small>
        
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../../js/jquery-1.10.2.min.js"></script>
    <script src="../../js/bootstrap-3.0.3.min.js"></script>
    <script src="../../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '../..';
    </script>
    <script data-main="../../mkdocs/js/search.js" src="../../mkdocs/js/require.js"></script>
    <script src="../../js/base.js"></script>
    <script>
        ((window.gitter = {}).chat = {}).options = {
            room: 'BioJulia/Bio.jl'
        };
    </script>
    <script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <script src="../../assets/mathjaxhelper.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
