<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0"> 
    <meta name="author" content="The BioJulia Organisation"> 
    <link rel="canonical" href="https://biojulia.github.io/Bio.jl/man/seq/"> 
    <link rel="shortcut icon" href="../../img/favicon.ico">

    <title>Biological Sequences - Bio.jl</title>

    <link href="../../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../../css/base.css" rel="stylesheet">
    <link href="../../css/biojulia.css" rel="stylesheet">
    <link rel="stylesheet" href="../../css/highlight.css">
    <link rel="stylesheet" href="../../css/friendly.css" type="text/css">
    <link href="../../assets/Documenter.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="../..">Bio.jl</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../..">Home</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Manual <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../install/">Installation</a>
</li>

                        
                            
<li >
    <a href="../reading/">IO API</a>
</li>

                        
                            
<li class="active">
    <a href="./">Biological Sequences</a>
</li>

                        
                            
<li >
    <a href="../alignments/">Alignments</a>
</li>

                        
                            
<li >
    <a href="../intervals/">Intervals</a>
</li>

                        
                            
<li >
    <a href="../var/">Biological Variation</a>
</li>

                        
                            
<li >
    <a href="../structure/">Structure</a>
</li>

                        
                            
<li >
    <a href="../util/">Util</a>
</li>

                        
                            
<li >
    <a href="../tools/">Tools</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Reference <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../../lib/seq/">Seq</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li >
                        <a href="../contributing/">Contributing</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../reading/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../alignments/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/BioJulia/Bio.jl">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="first-level active"><a href="#bioseq-biological-sequences">Bio.Seq: Biological Sequences</a></li>
        
            <li class="second-level"><a href="#biological-symbols">Biological symbols</a></li>
            
                <li class="third-level"><a href="#dna-and-rna-nucleotides">DNA and RNA nucleotides</a></li>
            
                <li class="third-level"><a href="#amino-acids">Amino acids</a></li>
            
                <li class="third-level"><a href="#other-functions">Other functions</a></li>
            
        
            <li class="second-level"><a href="#overview-of-sequences-types">Overview of sequences types</a></li>
            
        
            <li class="second-level"><a href="#general-purpose-sequences">General-purpose sequences</a></li>
            
                <li class="third-level"><a href="#constructing-sequences">Constructing sequences</a></li>
            
                <li class="third-level"><a href="#indexing-and-modifying">Indexing and modifying</a></li>
            
                <li class="third-level"><a href="#other-operations-on-sequences">Other operations on sequences</a></li>
            
                <li class="third-level"><a href="#compact-representation">Compact representation</a></li>
            
                <li class="third-level"><a href="#defining-a-new-alphabet">Defining a new alphabet</a></li>
            
        
            <li class="second-level"><a href="#nucleotide-k-mers">Nucleotide k-mers</a></li>
            
        
            <li class="second-level"><a href="#reference-sequences">Reference sequences</a></li>
            
        
            <li class="second-level"><a href="#sequence-search">Sequence search</a></li>
            
                <li class="third-level"><a href="#exact-search">Exact search</a></li>
            
                <li class="third-level"><a href="#approximate-search">Approximate search</a></li>
            
                <li class="third-level"><a href="#regular-expression-search">Regular expression search</a></li>
            
        
            <li class="second-level"><a href="#sequence-composition">Sequence composition</a></li>
            
        
            <li class="second-level"><a href="#sequence-records">Sequence records</a></li>
            
        
            <li class="second-level"><a href="#sequence-demultiplexing">Sequence demultiplexing</a></li>
            
        
    
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<p><a id='Bio.Seq:-Biological-Sequences-1'></a></p>
<h1 id="bioseq-biological-sequences">Bio.Seq: Biological Sequences</h1>
<p>The <code>Bio.Seq</code> module provides several data types for handling biological symbols and sequences.</p>
<p><a id='Biological-symbols-1'></a></p>
<h2 id="biological-symbols">Biological symbols</h2>
<p>The <code>Bio.Seq</code> module provides three biological symbol (character) types:</p>
<table>
<thead>
<tr>
<th align="left">Type</th>
<th align="left">Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>DNANucleotide</code></td>
<td align="left">DNA nucleotide</td>
</tr>
<tr>
<td align="left"><code>RNANucleotide</code></td>
<td align="left">RNA nucleotide</td>
</tr>
<tr>
<td align="left"><code>AminoAcid</code></td>
<td align="left">Amino acid</td>
</tr>
</tbody>
</table>
<p>These symbols can be elements of sequences like characters can be elements of strings.  See sections beginning from <a href="./#Overview-of-sequences-types-1">Overview of sequences types</a> section for details.</p>
<p><a id='DNA-and-RNA-nucleotides-1'></a></p>
<h3 id="dna-and-rna-nucleotides">DNA and RNA nucleotides</h3>
<p>Set of nucleotide symbols in Bio.jl covers IUPAC nucleotide base plus a gap symbol:</p>
<table>
<thead>
<tr>
<th align="left">Symbol</th>
<th align="left">Constant</th>
<th align="left">Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">'A'</td>
<td align="left"><code>DNA_A</code> / <code>RNA_A</code></td>
<td align="left">A; Adenine</td>
</tr>
<tr>
<td align="left">'C'</td>
<td align="left"><code>DNA_C</code> / <code>RNA_C</code></td>
<td align="left">C; Cytosine</td>
</tr>
<tr>
<td align="left">'G'</td>
<td align="left"><code>DNA_G</code> / <code>RNA_G</code></td>
<td align="left">G; Guanine</td>
</tr>
<tr>
<td align="left">'T'</td>
<td align="left"><code>DNA_T</code></td>
<td align="left">T; Thymine (DNA only)</td>
</tr>
<tr>
<td align="left">'U'</td>
<td align="left"><code>RNA_U</code></td>
<td align="left">U; Uracil (RNA only)</td>
</tr>
<tr>
<td align="left">'M'</td>
<td align="left"><code>DNA_M</code> / <code>RNA_M</code></td>
<td align="left">A or C</td>
</tr>
<tr>
<td align="left">'R'</td>
<td align="left"><code>DNA_R</code> / <code>RNA_R</code></td>
<td align="left">A or G</td>
</tr>
<tr>
<td align="left">'W'</td>
<td align="left"><code>DNA_W</code> / <code>RNA_W</code></td>
<td align="left">A or T/U</td>
</tr>
<tr>
<td align="left">'S'</td>
<td align="left"><code>DNA_S</code> / <code>RNA_S</code></td>
<td align="left">C or G</td>
</tr>
<tr>
<td align="left">'Y'</td>
<td align="left"><code>DNA_Y</code> / <code>RNA_Y</code></td>
<td align="left">C or T/U</td>
</tr>
<tr>
<td align="left">'K'</td>
<td align="left"><code>DNA_K</code> / <code>RNA_K</code></td>
<td align="left">G or T/U</td>
</tr>
<tr>
<td align="left">'V'</td>
<td align="left"><code>DNA_V</code> / <code>RNA_V</code></td>
<td align="left">A or C or G; not T/U</td>
</tr>
<tr>
<td align="left">'H'</td>
<td align="left"><code>DNA_H</code> / <code>RNA_H</code></td>
<td align="left">A or C or T; not G</td>
</tr>
<tr>
<td align="left">'D'</td>
<td align="left"><code>DNA_D</code> / <code>RNA_D</code></td>
<td align="left">A or G or T/U; not C</td>
</tr>
<tr>
<td align="left">'B'</td>
<td align="left"><code>DNA_B</code> / <code>RNA_B</code></td>
<td align="left">C or G or T/U; not A</td>
</tr>
<tr>
<td align="left">'N'</td>
<td align="left"><code>DNA_N</code> / <code>RNA_N</code></td>
<td align="left">A or C or G or T/U</td>
</tr>
<tr>
<td align="left">'-'</td>
<td align="left"><code>DNA_Gap</code> / <code>RNA_Gap</code></td>
<td align="left">Gap (none of the above)</td>
</tr>
</tbody>
</table>
<p><a href="http://www.insdc.org/documents/feature_table.html#7.4.1">http://www.insdc.org/documents/feature_table.html#7.4.1</a></p>
<p>Symbols are accessible as constants with <code>DNA_</code> or <code>RNA_</code> prefix:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">DNA_A</span>
<span class="go">DNA_A</span>

<span class="gp">julia&gt;</span> <span class="n">DNA_T</span>
<span class="go">DNA_T</span>

<span class="gp">julia&gt;</span> <span class="n">RNA_U</span>
<span class="go">RNA_U</span>

<span class="gp">julia&gt;</span> <span class="n">DNA_Gap</span>
<span class="go">DNA_Gap</span>

<span class="gp">julia&gt;</span> <span class="nb">typeof</span><span class="p">(</span><span class="n">DNA_A</span><span class="p">)</span>
<span class="go">Bio.Seq.DNANucleotide</span>

<span class="gp">julia&gt;</span> <span class="nb">typeof</span><span class="p">(</span><span class="n">RNA_A</span><span class="p">)</span>
<span class="go">Bio.Seq.RNANucleotide</span>
</pre></div>


<p>Symbols can be constructed by converting regular characters:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="nb">convert</span><span class="p">(</span><span class="n">DNANucleotide</span><span class="p">,</span> <span class="sc">&#39;C&#39;</span><span class="p">)</span>
<span class="go">DNA_C</span>

<span class="gp">julia&gt;</span> <span class="nb">convert</span><span class="p">(</span><span class="n">DNANucleotide</span><span class="p">,</span> <span class="sc">&#39;C&#39;</span><span class="p">)</span> <span class="o">===</span> <span class="n">DNA_C</span>
<span class="go">true</span>
</pre></div>


<p>Every nucleotide is encoded using the lower 4 bits of a byte. An unambiguous nucleotide has only one set bit and the other bits are unset. The table below summarizes all unambiguous nucleotides and their corresponding bits. An ambiguous nucleotide is the bitwise OR of unambiguous nucleotides that the ambiguous nucleotide can take. For example, <code>DNA_R</code> (meaning the nucleotide is either <code>DNA_A</code> or <code>DNA_G</code>) is encoded as <code>0101</code> because <code>0101</code> is the bitwise OR of <code>0001</code> (<code>DNA_A</code>) and <code>0100</code> (<code>DNA_G</code>). The gap symbol is always <code>0000</code>.</p>
<table>
<thead>
<tr>
<th align="left">Nucleotide</th>
<th align="left">Bits</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>DNA_A</code>, <code>RNA_A</code></td>
<td align="left"><code>0001</code></td>
</tr>
<tr>
<td align="left"><code>DNA_C</code>, <code>RNA_C</code></td>
<td align="left"><code>0010</code></td>
</tr>
<tr>
<td align="left"><code>DNA_G</code>, <code>RNA_G</code></td>
<td align="left"><code>0100</code></td>
</tr>
<tr>
<td align="left"><code>DNA_T</code>, <code>RNA_U</code></td>
<td align="left"><code>1000</code></td>
</tr>
</tbody>
</table>
<p>The next examples demonstrate bit operations of DNA:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">bits</span><span class="p">(</span><span class="n">reinterpret</span><span class="p">(</span><span class="n">UInt8</span><span class="p">,</span> <span class="n">DNA_A</span><span class="p">))</span>
<span class="go">&quot;00000001&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">bits</span><span class="p">(</span><span class="n">reinterpret</span><span class="p">(</span><span class="n">UInt8</span><span class="p">,</span> <span class="n">DNA_G</span><span class="p">))</span>
<span class="go">&quot;00000100&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">bits</span><span class="p">(</span><span class="n">reinterpret</span><span class="p">(</span><span class="n">UInt8</span><span class="p">,</span> <span class="n">DNA_R</span><span class="p">))</span>
<span class="go">&quot;00000101&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">bits</span><span class="p">(</span><span class="n">reinterpret</span><span class="p">(</span><span class="n">UInt8</span><span class="p">,</span> <span class="n">DNA_B</span><span class="p">))</span>
<span class="go">&quot;00001110&quot;</span>

<span class="gp">julia&gt;</span> <span class="o">~</span><span class="n">DNA_A</span>
<span class="go">DNA_B</span>

<span class="gp">julia&gt;</span> <span class="n">DNA_A</span> <span class="o">|</span> <span class="n">DNA_G</span>
<span class="go">DNA_R</span>

<span class="gp">julia&gt;</span> <span class="n">DNA_R</span> <span class="o">&amp;</span> <span class="n">DNA_B</span>
<span class="go">DNA_G</span>
</pre></div>


<p><a id='Amino-acids-1'></a></p>
<h3 id="amino-acids">Amino acids</h3>
<p>Set of amino acid symbols also covers IUPAC amino acid symbols plus a gap symbol:</p>
<table>
<thead>
<tr>
<th align="left">Symbol</th>
<th align="left">Constant</th>
<th align="left">Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">'A'</td>
<td align="left"><code>AA_A</code></td>
<td align="left">Alanine</td>
</tr>
<tr>
<td align="left">'R'</td>
<td align="left"><code>AA_R</code></td>
<td align="left">Arginine</td>
</tr>
<tr>
<td align="left">'N'</td>
<td align="left"><code>AA_N</code></td>
<td align="left">Asparagine</td>
</tr>
<tr>
<td align="left">'D'</td>
<td align="left"><code>AA_D</code></td>
<td align="left">Aspartic acid (Aspartate)</td>
</tr>
<tr>
<td align="left">'C'</td>
<td align="left"><code>AA_C</code></td>
<td align="left">Cysteine</td>
</tr>
<tr>
<td align="left">'Q'</td>
<td align="left"><code>AA_Q</code></td>
<td align="left">Glutamine</td>
</tr>
<tr>
<td align="left">'E'</td>
<td align="left"><code>AA_E</code></td>
<td align="left">Glutamic acid (Glutamate)</td>
</tr>
<tr>
<td align="left">'G'</td>
<td align="left"><code>AA_G</code></td>
<td align="left">Glycine</td>
</tr>
<tr>
<td align="left">'H'</td>
<td align="left"><code>AA_H</code></td>
<td align="left">Histidine</td>
</tr>
<tr>
<td align="left">'I'</td>
<td align="left"><code>AA_I</code></td>
<td align="left">Isoleucine</td>
</tr>
<tr>
<td align="left">'L'</td>
<td align="left"><code>AA_L</code></td>
<td align="left">Leucine</td>
</tr>
<tr>
<td align="left">'K'</td>
<td align="left"><code>AA_K</code></td>
<td align="left">Lysine</td>
</tr>
<tr>
<td align="left">'M'</td>
<td align="left"><code>AA_M</code></td>
<td align="left">Methionine</td>
</tr>
<tr>
<td align="left">'F'</td>
<td align="left"><code>AA_F</code></td>
<td align="left">Phenylalanine</td>
</tr>
<tr>
<td align="left">'P'</td>
<td align="left"><code>AA_P</code></td>
<td align="left">Proline</td>
</tr>
<tr>
<td align="left">'S'</td>
<td align="left"><code>AA_S</code></td>
<td align="left">Serine</td>
</tr>
<tr>
<td align="left">'T'</td>
<td align="left"><code>AA_T</code></td>
<td align="left">Threonine</td>
</tr>
<tr>
<td align="left">'W'</td>
<td align="left"><code>AA_W</code></td>
<td align="left">Tryptophan</td>
</tr>
<tr>
<td align="left">'Y'</td>
<td align="left"><code>AA_Y</code></td>
<td align="left">Tyrosine</td>
</tr>
<tr>
<td align="left">'V'</td>
<td align="left"><code>AA_V</code></td>
<td align="left">Valine</td>
</tr>
<tr>
<td align="left">'O'</td>
<td align="left"><code>AA_O</code></td>
<td align="left">Pyrrolysine</td>
</tr>
<tr>
<td align="left">'U'</td>
<td align="left"><code>AA_U</code></td>
<td align="left">Selenocysteine</td>
</tr>
<tr>
<td align="left">'B'</td>
<td align="left"><code>AA_B</code></td>
<td align="left">Aspartic acid or Asparagine</td>
</tr>
<tr>
<td align="left">'J'</td>
<td align="left"><code>AA_J</code></td>
<td align="left">Leucine or Isoleucine</td>
</tr>
<tr>
<td align="left">'Z'</td>
<td align="left"><code>AA_Z</code></td>
<td align="left">Glutamine or Glutamic acid</td>
</tr>
<tr>
<td align="left">'X'</td>
<td align="left"><code>AA_X</code></td>
<td align="left">Any amino acid</td>
</tr>
<tr>
<td align="left">'*'</td>
<td align="left"><code>AA_Term</code></td>
<td align="left">Termination codon</td>
</tr>
<tr>
<td align="left">'-'</td>
<td align="left"><code>AA_Gap</code></td>
<td align="left">Gap (none of the above)</td>
</tr>
</tbody>
</table>
<p><a href="http://www.insdc.org/documents/feature_table.html#7.4.3">http://www.insdc.org/documents/feature_table.html#7.4.3</a></p>
<p>Symbols are accessible as constants with <code>AA_</code> prefix:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">AA_A</span>
<span class="go">AA_A</span>

<span class="gp">julia&gt;</span> <span class="n">AA_Q</span>
<span class="go">AA_Q</span>

<span class="gp">julia&gt;</span> <span class="n">AA_Term</span>
<span class="go">AA_Term</span>

<span class="gp">julia&gt;</span> <span class="nb">typeof</span><span class="p">(</span><span class="n">AA_A</span><span class="p">)</span>
<span class="go">Bio.Seq.AminoAcid</span>
</pre></div>


<p>Symbols can be constructed by converting regular characters:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="nb">convert</span><span class="p">(</span><span class="n">AminoAcid</span><span class="p">,</span> <span class="sc">&#39;A&#39;</span><span class="p">)</span>
<span class="go">AA_A</span>

<span class="gp">julia&gt;</span> <span class="nb">convert</span><span class="p">(</span><span class="n">AminoAcid</span><span class="p">,</span> <span class="sc">&#39;P&#39;</span><span class="p">)</span> <span class="o">===</span> <span class="n">AA_P</span>
<span class="go">true</span>
</pre></div>


<p><a id='Other-functions-1'></a></p>
<h3 id="other-functions">Other functions</h3>
<p><a id='Bio.Seq.alphabet' href='#Bio.Seq.alphabet'>#</a>
<strong><code>Bio.Seq.alphabet</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alphabet(typ)
</pre></div>


<p>Return an iterator of symbols of <code>typ</code>.</p>
<p><code>typ</code> is one of <code>DNANucleotide</code>, <code>RNANucleotide</code>, or <code>AminoAcid</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/Seq.jl#L172-L178' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.gap' href='#Bio.Seq.gap'>#</a>
<strong><code>Bio.Seq.gap</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>gap(typ)
</pre></div>


<p>Return the gap symbol of <code>typ</code>.</p>
<p><code>typ</code> is one of <code>DNANucleotide</code>, <code>RNANucleotide</code>, <code>AminoAcid</code>, or <code>Char</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/Seq.jl#L181-L187' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.iscompatible' href='#Bio.Seq.iscompatible'>#</a>
<strong><code>Bio.Seq.iscompatible</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>iscompatible(x, y)
</pre></div>


<p>Return <code>true</code> if and only if <code>x</code> and <code>y</code> are compatible with each other (i.e. <code>x</code> and <code>y</code> can be the same symbol).</p>
<p><code>x</code> and <code>y</code> must be the same type (<code>DNANucleotide</code>, <code>RNANucleotide</code> or <code>AminoAcid</code>).</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">iscompatible</span><span class="p">(</span><span class="n">DNA_A</span><span class="p">,</span> <span class="n">DNA_A</span><span class="p">)</span>
<span class="n">true</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">iscompatible</span><span class="p">(</span><span class="n">DNA_C</span><span class="p">,</span> <span class="n">DNA_N</span><span class="p">)</span>  <span class="c"># DNA_N can be DNA_C</span>
<span class="n">true</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">iscompatible</span><span class="p">(</span><span class="n">DNA_C</span><span class="p">,</span> <span class="n">DNA_R</span><span class="p">)</span>  <span class="c"># DNA_R (A or G) cannot be DNA_C</span>
<span class="n">false</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">iscompatible</span><span class="p">(</span><span class="n">AA_A</span><span class="p">,</span> <span class="n">AA_X</span><span class="p">)</span>    <span class="c"># AA_X can be AA_A</span>
<span class="n">true</span>
</pre></div>


<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/nucleotide.jl#L145-L169' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.isambiguous' href='#Bio.Seq.isambiguous'>#</a>
<strong><code>Bio.Seq.isambiguous</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>isambiguous(nt::Nucleotide)
</pre></div>


<p>Test if <code>nt</code> is ambiguous nucleotide.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/nucleotide.jl#L115-L119' class='documenter-source'>source</a><br></p>
<p><a id='Overview-of-sequences-types-1'></a></p>
<h2 id="overview-of-sequences-types">Overview of sequences types</h2>
<p>The <code>Bio.Seq</code> module provides representations and tools for manipulating nucleotide and amino acid sequences. Sequences in Bio.jl are more strictly typed than in many other libraries; elements in a sequence are typed as biological symbol instead of character or byte.  They are special purpose types rather than simply strings and hence offer additional functionality that naive string types don't have. Though this strictness sacrifices some convenience, it also means you can always rely on a DNA sequence type to store DNA and nothing but DNA, without having to check, or deal with lowercase versus uppercase and so on. Strict separation of sequence types also means we are free to choose the most efficient representation. DNA and RNA sequences are encoded using four bits per base by default making them memory efficient, and also allowing us to speed up many common operations like nucleotide composition, reverse complement, and <em>k</em>-mer enumeration.</p>
<p>The <code>Bio.Seq</code> provides three different sequence types: <code>BioSequence</code>, <code>Kmer</code> and <code>ReferenceSequence</code>. Each of these types is a subtype of an abstract type called <code>Sequence</code> and supports various string-like operations such as random access and iteration. Different sequence types have different features. In most situations, <code>BioSequence</code> type will do and is used as the default representation. But sometimes other types are much more preferable in terms of memory efficiency and computation performance.  Here is the summary table of these three types:</p>
<table>
<thead>
<tr>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Element type</th>
<th align="left">Mutability</th>
<th align="left">Allocation</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>BioSequence{A&lt;:Alphabet}</code></td>
<td align="left">general-purpose biological sequences</td>
<td align="left">DNA, RNA, Amino acids</td>
<td align="left">mutable</td>
<td align="left">heap</td>
</tr>
<tr>
<td align="left"><code>Kmer{T&lt;:Nucleotide,k}</code></td>
<td align="left">specialized for short nucleotide sequences</td>
<td align="left">DNA, RNA</td>
<td align="left">immutable</td>
<td align="left">stack / register</td>
</tr>
<tr>
<td align="left"><code>ReferenceSequence</code></td>
<td align="left">specialized for long reference genomes</td>
<td align="left">DNA</td>
<td align="left">immutable</td>
<td align="left">heap</td>
</tr>
</tbody>
</table>
<p>Details of these different representations are explained in the following sections:</p>
<ul>
<li><code>BioSequence</code>: <a href="./#General-purpose-sequences-1">General-purpose sequences</a></li>
<li><code>Kmer</code>: <a href="./#Nucleotide-k-mers-1">Nucleotide k-mers</a></li>
<li><code>ReferenceSequence</code>: <a href="./#Reference-sequences-1">Reference sequences</a></li>
</ul>
<p><a id='General-purpose-sequences-1'></a></p>
<h2 id="general-purpose-sequences">General-purpose sequences</h2>
<p><code>BioSequence{A}</code> is a generic sequence type parameterized by an alphabet type <code>A</code> that defines the domain (or set) of biological symbols, and each alphabet has an associated symbol type. For example, <code>AminoAcidAlphabet</code> is associated with <code>AminoAcid</code> and hence an object of the <code>BioSequence{AminoAcidAlphabet}</code> type represents a sequence of amino acids.  Symbols from multiple alphabets can't be intermixed in one sequence type.</p>
<p>The following table summarizes common sequence types that are defined in the <code>Bio.Seq</code> module:</p>
<table>
<thead>
<tr>
<th align="left">Type</th>
<th align="left">Symbol type</th>
<th align="left">Type alias</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>BioSequence{DNAAlphabet{4}}</code></td>
<td align="left"><code>DNANucleotide</code></td>
<td align="left"><code>DNASequence</code></td>
</tr>
<tr>
<td align="left"><code>BioSequence{RNAAlphabet{4}}</code></td>
<td align="left"><code>RNANucleotide</code></td>
<td align="left"><code>RNASequence</code></td>
</tr>
<tr>
<td align="left"><code>BioSequence{AminoAcidAlphabet}</code></td>
<td align="left"><code>AminoAcid</code></td>
<td align="left"><code>AminoAcidSequence</code></td>
</tr>
<tr>
<td align="left"><code>BioSequence{CharAlphabet}</code></td>
<td align="left"><code>Char</code></td>
<td align="left"><code>CharSequence</code></td>
</tr>
</tbody>
</table>
<p>Parameterized definition of the <code>BioSequence{A}</code> type is for the purpose of unifying the data structure and operations of any symbol type. In most cases, users don't have to care about it and can use type aliases listed above. However, the alphabet type fixes the internal memory encoding and plays an important role when optimizing performance of a program (see <a href="./#Compact-representation-1">Compact representation</a> section for low-memory encodings).  It also enables a user to define their own alphabet only by defining few numbers of methods. This is described in <a href="./#Defining-a-new-alphabet-1">Defining a new alphabet</a> section.</p>
<p><a id='Constructing-sequences-1'></a></p>
<h3 id="constructing-sequences">Constructing sequences</h3>
<p>Sequence types corresponding to these alphabets can be constructed a number of different ways. Most immediately, sequence literals can be constructed using the string macros <code>dna</code>, <code>rna</code>, <code>aa</code>, and <code>char</code>:</p>
<div class="codehilite"><pre><span></span><span class="go"># String decorators are provided for common sequence types</span>
<span class="gp">julia&gt;</span> <span class="n">dna</span><span class="s">&quot;TACGTANNATC&quot;</span>
<span class="go">11nt DNA Sequence:</span>
<span class="go">TACGTANNATC</span>

<span class="gp">julia&gt;</span> <span class="n">rna</span><span class="s">&quot;AUUUGNCCANU&quot;</span>
<span class="go">11nt RNA Sequence:</span>
<span class="go">AUUUGNCCANU</span>

<span class="gp">julia&gt;</span> <span class="n">aa</span><span class="s">&quot;ARNDCQEGHILKMFPSTWYVX&quot;</span>
<span class="go">21aa Amino Acid Sequence:</span>
<span class="go">ARNDCQEGHILKMFPSTWYVX</span>

<span class="gp">julia&gt;</span> <span class="n">char</span><span class="s">&quot;αβγδϵ&quot;</span>
<span class="go">5char Char Sequence:</span>
<span class="go">αβγδϵ</span>
</pre></div>


<p>Sequence can also be constructed from strings or arrays of nucleotide or amino acid symbols using constructors or the <code>convert</code> function:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">DNASequence</span><span class="p">(</span><span class="s">&quot;TTANC&quot;</span><span class="p">)</span>
<span class="go">5nt DNA Sequence:</span>
<span class="go">TTANC</span>

<span class="gp">julia&gt;</span> <span class="n">DNASequence</span><span class="p">([</span><span class="n">DNA_T</span><span class="p">,</span> <span class="n">DNA_T</span><span class="p">,</span> <span class="n">DNA_A</span><span class="p">,</span> <span class="n">DNA_N</span><span class="p">,</span> <span class="n">DNA_C</span><span class="p">])</span>
<span class="go">5nt DNA Sequence:</span>
<span class="go">TTANC</span>

<span class="gp">julia&gt;</span> <span class="nb">convert</span><span class="p">(</span><span class="n">DNASequence</span><span class="p">,</span> <span class="p">[</span><span class="n">DNA_T</span><span class="p">,</span> <span class="n">DNA_T</span><span class="p">,</span> <span class="n">DNA_A</span><span class="p">,</span> <span class="n">DNA_N</span><span class="p">,</span> <span class="n">DNA_C</span><span class="p">])</span>
<span class="go">5nt DNA Sequence:</span>
<span class="go">TTANC</span>
</pre></div>


<p>Using <code>convert</code>, these operations are reversible: sequences can be converted to strings or arrays:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="nb">convert</span><span class="p">(</span><span class="n">String</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TTANGTA&quot;</span><span class="p">)</span>
<span class="go">&quot;TTANGTA&quot;</span>

<span class="gp">julia&gt;</span> <span class="nb">convert</span><span class="p">(</span><span class="n">Vector</span><span class="p">{</span><span class="n">DNANucleotide</span><span class="p">},</span> <span class="n">dna</span><span class="s">&quot;TTANGTA&quot;</span><span class="p">)</span>
<span class="go">7-element Array{Bio.Seq.DNANucleotide,1}:</span>
<span class="go"> DNA_T</span>
<span class="go"> DNA_T</span>
<span class="go"> DNA_A</span>
<span class="go"> DNA_N</span>
<span class="go"> DNA_G</span>
<span class="go"> DNA_T</span>
<span class="go"> DNA_A</span>
</pre></div>


<p>Sequences can also be concatenated into longer sequences:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">DNASequence</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ACGT&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;NNNN&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TGCA&quot;</span><span class="p">)</span>
<span class="go">12nt DNA Sequence:</span>
<span class="go">ACGTNNNNTGCA</span>

<span class="gp">julia&gt;</span> <span class="n">dna</span><span class="s">&quot;ACGT&quot;</span> <span class="o">*</span> <span class="n">dna</span><span class="s">&quot;TGCA&quot;</span>
<span class="go">8nt DNA Sequence:</span>
<span class="go">ACGTTGCA</span>

<span class="gp">julia&gt;</span> <span class="n">repeat</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;TA&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="go">20nt DNA Sequence:</span>
<span class="go">TATATATATATATATATATA</span>

<span class="gp">julia&gt;</span> <span class="n">dna</span><span class="s">&quot;TA&quot;</span> <span class="o">^</span> <span class="mi">10</span>
<span class="go">20nt DNA Sequence:</span>
<span class="go">TATATATATATATATATATA</span>
</pre></div>


<p>Despite being separate types, <code>DNASequence</code> and <code>RNASequence</code> can freely be converted between efficiently without copying the underlying data:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">dna</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;TTANGTAGACCG&quot;</span>
<span class="go">12nt DNA Sequence:</span>
<span class="go">TTANGTAGACCG</span>

<span class="gp">julia&gt;</span> <span class="n">rna</span> <span class="o">=</span> <span class="nb">convert</span><span class="p">(</span><span class="n">RNASequence</span><span class="p">,</span> <span class="n">dna</span><span class="p">)</span>
<span class="go">12nt RNA Sequence:</span>
<span class="go">UUANGUAGACCG</span>

<span class="gp">julia&gt;</span> <span class="n">dna</span><span class="o">.</span><span class="n">data</span> <span class="o">===</span> <span class="n">rna</span><span class="o">.</span><span class="n">data</span>  <span class="c"># underlying data are same</span>
<span class="go">true</span>
</pre></div>


<p>A random sequence can be obtained by the <code>randdnaseq</code>, <code>randrnaseq</code> and <code>randaaseq</code> functions, which generate <code>DNASequence</code>, <code>RNASequence</code> and <code>AminoAcidSequence</code>, respectively. Generated sequences are composed of the standard symbols without ambiguity and gap. For example, <code>randdnaseq(6)</code> may generate <code>dna"TCATAG"</code> but never generates <code>dna"TNANAG"</code> or <code>dna"T-ATAG"</code>.</p>
<p>A translatable <code>RNASequence</code> can also be converted to an <code>AminoAcidSequence</code> using the <a href="./#Bio.Seq.translate"><code>translate</code></a> function described below.</p>
<p><a id='Indexing-and-modifying-1'></a></p>
<h3 id="indexing-and-modifying">Indexing and modifying</h3>
<p>Sequences for the most part behave like other vector or string types. They can be indexed using integers or ranges:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;ACGTTTANAGTNNAGTACC&quot;</span>
<span class="go">19nt DNA Sequence:</span>
<span class="go">ACGTTTANAGTNNAGTACC</span>

<span class="gp">julia&gt;</span> <span class="n">seq</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
<span class="go">DNA_T</span>

<span class="gp">julia&gt;</span> <span class="n">seq</span><span class="p">[</span><span class="mi">6</span><span class="p">:</span><span class="k">end</span><span class="p">]</span>
<span class="go">14nt DNA Sequence:</span>
<span class="go">TANAGTNNAGTACC</span>
</pre></div>


<p>Indexing by range creates a subsequence of the original sequence. Unlike <code>Vector</code> in the standard library, creating a subsequences is copy-free: a subsequence is just a reference to the original sequence with its range.  You may think that this is unsafe because modifying subsequences propagates to the original sequence, but this doesn't happen actually:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;AAAA&quot;</span>    <span class="c"># create a sequence</span>
<span class="go">4nt DNA Sequence:</span>
<span class="go">AAAA</span>

<span class="gp">julia&gt;</span> <span class="n">subseq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>  <span class="c"># create a subsequence from `seq`</span>
<span class="go">2nt DNA Sequence:</span>
<span class="go">AA</span>

<span class="gp">julia&gt;</span> <span class="n">subseq</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">DNA_T</span>  <span class="c"># modify the second element of it</span>
<span class="go">DNA_T</span>

<span class="gp">julia&gt;</span> <span class="n">subseq</span>             <span class="c"># the subsequence is modified</span>
<span class="go">2nt DNA Sequence:</span>
<span class="go">AT</span>

<span class="gp">julia&gt;</span> <span class="n">seq</span>                <span class="c"># but the original sequence is not</span>
<span class="go">4nt DNA Sequence:</span>
<span class="go">AAAA</span>
</pre></div>


<p>This is because modifying a sequence checks whether its underlying data are shared with other sequences under the hood. If and only if the data are shared, the subsequence creates a copy of itself. Any modifying operation does this check. This is called <em>copy-on-write</em> strategy and users don't need to care about it because it is transparent from outward.</p>
<p>The following modifying operations are currently supported:</p>
<div class="codehilite"><pre><span></span><span class="n">setindex!</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">item</span><span class="p">,</span> <span class="n">index</span><span class="p">)</span>
<span class="n">push!</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">item</span><span class="p">)</span>
<span class="n">pop!</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">shift!</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">unshift!</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">item</span><span class="p">)</span>
<span class="n">insert!</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">index</span><span class="p">,</span> <span class="n">item</span><span class="p">)</span>
<span class="n">deleteat!</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">index</span><span class="p">)</span>
<span class="n">append!</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">other_seq</span><span class="p">)</span>
<span class="n">copy!</span><span class="p">(</span><span class="n">dst_seq</span><span class="p">,</span> <span class="n">dest_offset</span><span class="p">,</span> <span class="n">src_seq</span><span class="p">,</span> <span class="n">src_offset</span><span class="p">,</span> <span class="n">len</span><span class="p">)</span>
<span class="n">reverse!</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>


<p>And these two are functions for nucleotide sequences:</p>
<div class="codehilite"><pre><span></span><span class="n">complement!</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">reverse_complement!</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;ACG&quot;</span>
<span class="go">3nt DNA Sequence:</span>
<span class="go">ACG</span>

<span class="gp">julia&gt;</span> <span class="n">push!</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">DNA_T</span><span class="p">)</span>
<span class="go">4nt DNA Sequence:</span>
<span class="go">ACGT</span>

<span class="gp">julia&gt;</span> <span class="n">append!</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;AT&quot;</span><span class="p">)</span>
<span class="go">6nt DNA Sequence:</span>
<span class="go">ACGTAT</span>

<span class="gp">julia&gt;</span> <span class="n">reverse!</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">6nt DNA Sequence:</span>
<span class="go">TATGCA</span>

<span class="gp">julia&gt;</span> <span class="n">complement!</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">6nt DNA Sequence:</span>
<span class="go">ATACGT</span>

<span class="gp">julia&gt;</span> <span class="n">reverse_complement!</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">6nt DNA Sequence:</span>
<span class="go">ACGTAT</span>
</pre></div>


<p>Sequences also work as iterators over symbols:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
<span class="go">0</span>

<span class="gp">julia&gt;</span> <span class="k">for</span> <span class="n">nt</span> <span class="k">in</span> <span class="n">dna</span><span class="s">&quot;ATNGNNT&quot;</span>

<span class="gt">           if nt == DNA_N</span>

<span class="gt">               n += 1</span>

<span class="gt">           end</span>

<span class="gt">       end</span>

<span class="gp">julia&gt;</span> <span class="n">n</span>
<span class="go">3</span>
</pre></div>


<p><a id='Other-operations-on-sequences-1'></a></p>
<h3 id="other-operations-on-sequences">Other operations on sequences</h3>
<p>A number of common sequence operations are provided in the <code>Bio.Seq</code> module:</p>
<p><a id='Bio.Seq.complement' href='#Bio.Seq.complement'>#</a>
<strong><code>Bio.Seq.complement</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>complement(nt::Nucleotide)
</pre></div>


<p>Return the complementary nucleotide of <code>nt</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/nucleotide.jl#L124-L128' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>complement(seq)
</pre></div>


<p>Make a complement sequence of <code>seq</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/bioseq.jl#L693-L697' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>complement(kmer::Kmer)
</pre></div>


<p>Return the complement of <code>kmer</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/kmer.jl#L144-L148' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.reverse_complement' href='#Bio.Seq.reverse_complement'>#</a>
<strong><code>Bio.Seq.reverse_complement</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>reverse_complement(seq)
</pre></div>


<p>Make a reversed complement sequence of <code>seq</code>.</p>
<p>Ambiguous nucleotides are left as-is.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/bioseq.jl#L713-L719' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>reverse_complement(kmer::Kmer)
</pre></div>


<p>Return the reverse complement of <code>kmer</code></p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/kmer.jl#L158-L162' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.mismatches' href='#Bio.Seq.mismatches'>#</a>
<strong><code>Bio.Seq.mismatches</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>mismatches(seq1::BioSequence, seq2::BioSequence[, compatible=false])
</pre></div>


<p>Return the number of mismatches between <code>seq1</code> and <code>seq2</code>.</p>
<p>If <code>seq1</code> and <code>seq2</code> are of differing lengths, only the first <code>min(length(seq1), length(seq2))</code> nucleotides are compared.  When <code>compatible</code> is <code>true</code>, sequence symbols are comapred using <code>iscompatible</code>; otherwise using <code>==</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/bioseq.jl#L764-L772' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>mismatches(a::Kmer, b::Kmer)
</pre></div>


<p>Return the number of mismatches between <code>a</code> and <code>b</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/kmer.jl#L165-L169' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.composition' href='#Bio.Seq.composition'>#</a>
<strong><code>Bio.Seq.composition</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>composition(seq | kmer_iter)
</pre></div>


<p>Calculate composition of biological symbols in <code>seq</code> or k-mers in <code>kmer_iter</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/composition.jl#L56-L60' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.seqmatrix' href='#Bio.Seq.seqmatrix'>#</a>
<strong><code>Bio.Seq.seqmatrix</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>seqmatrix{A&lt;:Alphabet}(vseq::Vector{BioSequence{A}}, major::Symbol)
</pre></div>


<p>Construct a matrix of nucleotides or amino acids from a vector of <code>BioSequence</code>s.</p>
<p>If parameter <code>major</code> is set to <code>:site</code>, the matrix is created such that one nucleotide from each sequence is placed in each column i.e. the matrix is laid out in site-major order. This means that iteration over one position of many sequences is efficient, as julia arrays are laid out in column major order.</p>
<p>If the parameter <code>major</code> is set to <code>:seq</code>, the matrix is created such that each sequence is placed in one column i.e. the matrix is laid out in sequence-major order. This means that iteration across each sequence in turn is efficient, as julia arrays are laid out in column major order.</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">dna</span><span class="s">&quot;AAA&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TTT&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;CCC&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;GGG&quot;</span><span class="p">]</span>
<span class="mi">4</span><span class="o">-</span><span class="n">element</span> <span class="n">Array</span><span class="p">{</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">BioSequence</span><span class="p">{</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">DNAAlphabet</span><span class="p">{</span><span class="mi">4</span><span class="p">}},</span><span class="mi">1</span><span class="p">}:</span>
 <span class="mi">3</span><span class="n">nt</span> <span class="n">DNA</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">AAA</span>
 <span class="mi">3</span><span class="n">nt</span> <span class="n">DNA</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">TTT</span>
 <span class="mi">3</span><span class="n">nt</span> <span class="n">DNA</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">CCC</span>
 <span class="mi">3</span><span class="n">nt</span> <span class="n">DNA</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">GGG</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">seqmatrix</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="p">:</span><span class="n">site</span><span class="p">)</span>
<span class="mi">4</span><span class="n">x3</span> <span class="n">Array</span><span class="p">{</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">DNANucleotide</span><span class="p">,</span><span class="mi">2</span><span class="p">}:</span>
 <span class="n">DNA_A</span>  <span class="n">DNA_A</span>  <span class="n">DNA_A</span>
 <span class="n">DNA_T</span>  <span class="n">DNA_T</span>  <span class="n">DNA_T</span>
 <span class="n">DNA_C</span>  <span class="n">DNA_C</span>  <span class="n">DNA_C</span>
 <span class="n">DNA_G</span>  <span class="n">DNA_G</span>  <span class="n">DNA_G</span>

 <span class="n">julia</span><span class="o">&gt;</span> <span class="n">seqmatrix</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="p">:</span><span class="n">seq</span><span class="p">)</span>
 <span class="mi">3</span><span class="n">x4</span> <span class="n">Array</span><span class="p">{</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">DNANucleotide</span><span class="p">,</span><span class="mi">2</span><span class="p">}:</span>
  <span class="n">DNA_A</span>  <span class="n">DNA_T</span>  <span class="n">DNA_C</span>  <span class="n">DNA_G</span>
  <span class="n">DNA_A</span>  <span class="n">DNA_T</span>  <span class="n">DNA_C</span>  <span class="n">DNA_G</span>
  <span class="n">DNA_A</span>  <span class="n">DNA_T</span>  <span class="n">DNA_C</span>  <span class="n">DNA_G</span>
</pre></div>


<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/bioseq.jl#L1026-L1069' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>seqmatrix{A&lt;:Alphabet,T}(::Type{T}, vseq::Vector{BioSequence{A}}, major::Symbol)
</pre></div>


<p>Construct a matrix of <code>T</code> from a vector of <code>BioSequence</code>s.</p>
<p>If parameter <code>major</code> is set to <code>:site</code>, the matrix is created such that one nucleotide from each sequence is placed in each column i.e. the matrix is laid out in site-major order. This means that iteration over one position of many sequences is efficient, as julia arrays are laid out in column major order.</p>
<p>If the parameter <code>major</code> is set to <code>:seq</code>, the matrix is created such that each sequence is placed in one column i.e. the matrix is laid out in sequence-major order. This means that iteration across each sequence in turn is efficient, as julia arrays are laid out in column major order.</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">dna</span><span class="s">&quot;AAA&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TTT&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;CCC&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;GGG&quot;</span><span class="p">]</span>
<span class="mi">4</span><span class="o">-</span><span class="n">element</span> <span class="n">Array</span><span class="p">{</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">BioSequence</span><span class="p">{</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">DNAAlphabet</span><span class="p">{</span><span class="mi">4</span><span class="p">}},</span><span class="mi">1</span><span class="p">}:</span>
 <span class="mi">3</span><span class="n">nt</span> <span class="n">DNA</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">AAA</span>
 <span class="mi">3</span><span class="n">nt</span> <span class="n">DNA</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">TTT</span>
 <span class="mi">3</span><span class="n">nt</span> <span class="n">DNA</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">CCC</span>
 <span class="mi">3</span><span class="n">nt</span> <span class="n">DNA</span> <span class="n">Sequence</span><span class="p">:</span>
<span class="n">GGG</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">seqmatrix</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="p">:</span><span class="n">site</span><span class="p">,</span> <span class="n">UInt8</span><span class="p">)</span>
<span class="mi">4</span><span class="n">×3</span> <span class="n">Array</span><span class="p">{</span><span class="n">UInt8</span><span class="p">,</span><span class="mi">2</span><span class="p">}:</span>
 <span class="mh">0x01</span>  <span class="mh">0x01</span>  <span class="mh">0x01</span>
 <span class="mh">0x08</span>  <span class="mh">0x08</span>  <span class="mh">0x08</span>
 <span class="mh">0x02</span>  <span class="mh">0x02</span>  <span class="mh">0x02</span>
 <span class="mh">0x04</span>  <span class="mh">0x04</span>  <span class="mh">0x04</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">seqmatrix</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="p">:</span><span class="n">seq</span><span class="p">,</span> <span class="n">UInt8</span><span class="p">)</span>
<span class="mi">3</span><span class="n">×4</span> <span class="n">Array</span><span class="p">{</span><span class="n">UInt8</span><span class="p">,</span><span class="mi">2</span><span class="p">}:</span>
 <span class="mh">0x01</span>  <span class="mh">0x08</span>  <span class="mh">0x02</span>  <span class="mh">0x04</span>
 <span class="mh">0x01</span>  <span class="mh">0x08</span>  <span class="mh">0x02</span>  <span class="mh">0x04</span>
 <span class="mh">0x01</span>  <span class="mh">0x08</span>  <span class="mh">0x02</span>  <span class="mh">0x04</span>
</pre></div>


<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/bioseq.jl#L1094-L1137' class='documenter-source'>source</a><br></p>
<p><a id='Translation-1'></a></p>
<h4 id="translation">Translation</h4>
<p>The <a href="./#Bio.Seq.translate"><code>translate</code></a> funtion translates a sequence of codons in a RNA sequence to a amino acid sequence besed on a genetic code mapping.  The <code>Bio.Seq</code> module contains all NCBI defined genetic codes and they are registered in <a href="./#Bio.Seq.ncbi_trans_table"><code>ncbi_trans_table</code></a>.</p>
<p><a id='Bio.Seq.translate' href='#Bio.Seq.translate'>#</a>
<strong><code>Bio.Seq.translate</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>translate(rna_seq, code=standard_genetic_code, allow_ambiguous_codons=true)
</pre></div>


<p>Translate an <code>RNASequence</code> to an <code>AminoAcidSequence</code>.</p>
<p>Translation uses genetic code <code>code</code> to map codons to amino acids. See <code>ncbi_trans_table</code> for available genetic codes. If codons in the given RNA sequence cannot determine a unique amino acid, they will be translated to <code>AA_X</code> if <code>allow_ambiguous_codons</code> is <code>true</code> and otherwise result in an error.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/geneticcode.jl#L311-L321' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.ncbi_trans_table' href='#Bio.Seq.ncbi_trans_table'>#</a>
<strong><code>Bio.Seq.ncbi_trans_table</code></strong> &mdash; <em>Constant</em>.</p>
<p>Genetic code list of NCBI.</p>
<p>The standard genetic code is <code>ncbi_trans_table[1]</code> and others can be shown by <code>show(ncbi_trans_table)</code>. For more details, consult the next link: http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/geneticcode.jl#L90-L97' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">ncbi_trans_table</span>
<span class="go">Translation Tables:</span>
<span class="go">  1. The Standard Code (standard_genetic_code)</span>
<span class="go">  2. The Vertebrate Mitochondrial Code (vertebrate_mitochondrial_genetic_code)</span>
<span class="go">  3. The Yeast Mitochondrial Code (yeast_mitochondrial_genetic_code)</span>
<span class="go">  4. The Mold, Protozoan, and Coelenterate Mitochondrial Code and the Mycoplasma/Spiroplasma Code (mold_mitochondrial_genetic_code)</span>
<span class="go">  5. The Invertebrate Mitochondrial Code (invertebrate_mitochondrial_genetic_code)</span>
<span class="go">  6. The Ciliate, Dasycladacean and Hexamita Nuclear Code (ciliate_nuclear_genetic_code)</span>
<span class="go">  9. The Echinoderm and Flatworm Mitochondrial Code (echinoderm_mitochondrial_genetic_code)</span>
<span class="go"> 10. The Euplotid Nuclear Code (euplotid_nuclear_genetic_code)</span>
<span class="go"> 11. The Bacterial, Archaeal and Plant Plastid Code (bacterial_plastid_genetic_code)</span>
<span class="go"> 12. The Alternative Yeast Nuclear Code (alternative_yeast_nuclear_genetic_code)</span>
<span class="go"> 13. The Ascidian Mitochondrial Code (ascidian_mitochondrial_genetic_code)</span>
<span class="go"> 14. The Alternative Flatworm Mitochondrial Code (alternative_flatworm_mitochondrial_genetic_code)</span>
<span class="go"> 16. Chlorophycean Mitochondrial Code (chlorophycean_mitochondrial_genetic_code)</span>
<span class="go"> 21. Trematode Mitochondrial Code (trematode_mitochondrial_genetic_code)</span>
<span class="go"> 22. Scenedesmus obliquus Mitochondrial Code (scenedesmus_obliquus_mitochondrial_genetic_code)</span>
<span class="go"> 23. Thraustochytrium Mitochondrial Code (thraustochytrium_mitochondrial_genetic_code)</span>
<span class="go"> 24. Pterobranchia Mitochondrial Code (pterobrachia_mitochondrial_genetic_code)</span>
<span class="go"> 25. Candidate Division SR1 and Gracilibacteria Code (candidate_division_sr1_genetic_code)</span>
</pre></div>


<p><a href="http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes">http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes</a></p>
<p><a id='Compact-representation-1'></a></p>
<h3 id="compact-representation">Compact representation</h3>
<p>As we saw above, DNA and RNA sequences can store any ambiguous nucleotides like 'N'.  If you are sure that nucleotide sequences store unambiguous nucleotides only, you can save the memory space of sequences. <code>DNAAlphabet{2}</code> is an alphabet that uses two bits per base and limits to only unambiguous nucleotide symbols (ACGT in DNA and ACGU in RNA). To create a sequence of this alphabet, you need to explicitly pass <code>DNAAlphabet{2}</code> to <code>BioSequence</code> as its parametric type:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">BioSequence</span><span class="p">{</span><span class="n">DNAAlphabet</span><span class="p">{</span><span class="mi">2</span><span class="p">}}(</span><span class="s">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="go">4nt DNA Sequence:</span>
<span class="go">ACGT</span>
</pre></div>


<p>Recall that <code>DNASequence</code> is a type alias of <code>BioSequence{DNAAlphabet{4}}</code>, which uses four bits per base. That is, <code>BioSequence{DNAAlphabet{2}}</code> saves half of memory footprint compared to <code>BioSequence{DNAAlphabet{4}}</code>. If you need to handle reference genomes that are composed of five nucleotides, ACGTN, consider to use the <code>ReferenceSequence</code> type described in the <a href="./#Reference-sequences-1">Reference sequences</a> section.</p>
<p><a id='Defining-a-new-alphabet-1'></a></p>
<h3 id="defining-a-new-alphabet">Defining a new alphabet</h3>
<p>The alphabet type parameter <code>A</code> of <code>BioSequence{A}</code> enables a user to extend functionality of <code>BioSequence</code> with mimimum effort. As an example, definition of a new alphabet type representing a sequence of boolean values is shown below:</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span>

<span class="k">immutable</span> <span class="n">BoolAlphabet</span> <span class="o">&lt;:</span> <span class="n">Alphabet</span> <span class="k">end</span>

<span class="n">Seq</span><span class="o">.</span><span class="n">bitsof</span><span class="p">(::</span><span class="n">Type</span><span class="p">{</span><span class="n">BoolAlphabet</span><span class="p">})</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">Seq</span><span class="o">.</span><span class="n">eltype</span><span class="p">(::</span><span class="n">Type</span><span class="p">{</span><span class="n">BoolAlphabet</span><span class="p">})</span> <span class="o">=</span> <span class="kt">Bool</span>
<span class="n">Seq</span><span class="o">.</span><span class="n">alphabet</span><span class="p">(::</span><span class="n">Type</span><span class="p">{</span><span class="n">BoolAlphabet</span><span class="p">})</span> <span class="o">=</span> <span class="n">false</span><span class="p">:</span><span class="n">true</span>

<span class="k">function</span><span class="nf"> Seq</span><span class="o">.</span><span class="n">encode</span><span class="p">(::</span><span class="n">Type</span><span class="p">{</span><span class="n">BoolAlphabet</span><span class="p">},</span> <span class="n">x</span><span class="p">::</span><span class="kt">Bool</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">UInt64</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mh">0x01</span><span class="p">,</span> <span class="mh">0x00</span><span class="p">))</span>
<span class="k">end</span>

<span class="k">function</span><span class="nf"> Seq</span><span class="o">.</span><span class="n">decode</span><span class="p">(::</span><span class="n">Type</span><span class="p">{</span><span class="n">BoolAlphabet</span><span class="p">},</span> <span class="n">x</span><span class="p">::</span><span class="n">UInt64</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mh">0x01</span>
        <span class="nb">throw</span><span class="p">(</span><span class="n">Seq</span><span class="o">.</span><span class="n">DecodeError</span><span class="p">(</span><span class="n">BoolAlphabet</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span>
    <span class="k">end</span>
    <span class="k">return</span> <span class="n">ifelse</span><span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="mh">0x00</span><span class="p">,</span> <span class="n">false</span><span class="p">,</span> <span class="n">true</span><span class="p">)</span>
<span class="k">end</span>
</pre></div>


<p><a id='Nucleotide-k-mers-1'></a></p>
<h2 id="nucleotide-k-mers">Nucleotide k-mers</h2>
<p>A common strategy to simplify the analysis of sequence data is to operate or short k-mers, for size fixed size <code>k</code>. These can be packed into machine integers allowing extremely efficient code. The <code>Bio.Seq</code> module has built in support for representing short sequences in 64-bit integers. Besides being fixed length, <code>Kmer</code> types, unlike other sequence types cannot contain ambiguous symbols like 'N'.</p>
<p>The <code>Kmer{T,k}</code> type parameterized on symbol type (<code>T</code>, either <code>DNANucleotide</code>, or <code>RNANucleotide</code>) and size <code>k</code>. For ease of writing code, two type aliases for each nucleotide type are defined and named as <code>DNAKmer{k}</code> and <code>RNAKmer{k}</code>:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">DNAKmer</span><span class="p">(</span><span class="s">&quot;ACGT&quot;</span><span class="p">)</span>  <span class="c"># create a DNA 4-mer from a string</span>
<span class="go">DNA 4-mer:</span>
<span class="go">ACGT</span>

<span class="gp">julia&gt;</span> <span class="n">RNAKmer</span><span class="p">(</span><span class="s">&quot;ACGU&quot;</span><span class="p">)</span>  <span class="c"># create an RNA 4-mer from a string</span>
<span class="go">RNA 4-mer:</span>
<span class="go">ACGU</span>

<span class="gp">julia&gt;</span> <span class="nb">typeof</span><span class="p">(</span><span class="n">DNAKmer</span><span class="p">(</span><span class="s">&quot;ACGT&quot;</span><span class="p">))</span>
<span class="go">Bio.Seq.Kmer{Bio.Seq.DNANucleotide,4}</span>
</pre></div>


<p><a id='Bio.Seq.each' href='#Bio.Seq.each'>#</a>
<strong><code>Bio.Seq.each</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>each(::Type{Kmer{T,k}}, seq::Sequence[, step=1])
</pre></div>


<p>Initialize an iterator over all k-mers in a sequence <code>seq</code> skipping ambiguous nucleotides without changing the reading frame.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>Kmer{T,k}</code>: k-mer type to enumerate.</li>
<li><code>seq</code>: a nucleotide sequence.</li>
<li><code>step=1</code>: the number of positions between iterated k-mers</li>
</ul>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span># iterate over DNA codons
for (pos, codon) in each(DNAKmer{3}, dna&quot;ATCCTANAGNTACT&quot;, 3)
    @show pos, codon
end
</pre></div>


<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/eachkmer.jl#L15-L33' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.canonical' href='#Bio.Seq.canonical'>#</a>
<strong><code>Bio.Seq.canonical</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>canonical(kmer::Kmer)
</pre></div>


<p>Return the canonical k-mer of <code>x</code>.</p>
<p>A canonical k-mer is the numerical lesser of a k-mer and its reverse complement. This is useful in hashing/counting k-mers in data that is not strand specific, and thus observing k-mer is equivalent to observing its reverse complement.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/kmer.jl#L172-L180' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.neighbors' href='#Bio.Seq.neighbors'>#</a>
<strong><code>Bio.Seq.neighbors</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>neighbors(kmer::Kmer)
</pre></div>


<p>Return an iterator through k-mers neighboring <code>kmer</code> on a de Bruijn graph.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/kmer.jl#L203-L207' class='documenter-source'>source</a><br></p>
<p><a id='Reference-sequences-1'></a></p>
<h2 id="reference-sequences">Reference sequences</h2>
<p><code>DNASequence</code> (alias of <code>BioSequence{DNAAlphabet{4}}</code>) is a flexible data sturcture but always consumes 4 bits per base, which will waste a large part of the memory space when storing reference genome sequences.  In such a case, <code>ReferenceSequence</code> is helpful because it compresses positions of 'N' symbols so that long DNA sequences are stored with almost 2 bits per base. An important limitation is that the <code>ReferenceSequence</code> type is immutable due to the compression. Other sequence-like operations are supported:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">ReferenceSequence</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;NNCGTATTTTCN&quot;</span><span class="p">)</span>
<span class="go">12nt Reference Sequence:</span>
<span class="go">NNCGTATTTTCN</span>

<span class="gp">julia&gt;</span> <span class="n">seq</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">DNA_N</span>

<span class="gp">julia&gt;</span> <span class="n">seq</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
<span class="go">DNA_T</span>

<span class="gp">julia&gt;</span> <span class="n">seq</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">6</span><span class="p">]</span>
<span class="go">5nt Reference Sequence:</span>
<span class="go">NCGTA</span>

<span class="gp">julia&gt;</span> <span class="n">ReferenceSequence</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ATGM&quot;</span><span class="p">)</span>  <span class="c"># DNA_M is not accepted</span>
<span class="go">ERROR: ArgumentError: invalid symbol M ∉ {A,C,G,T,N} at 4</span>
<span class="go"> in convert at /Users/kenta/.julia/v0.4/Bio/src/seq/refseq.jl:58</span>
<span class="go"> in call at essentials.jl:56</span>
</pre></div>


<p>When reading reference sequences from a FASTA file, the following snippet will avoid allocating temporary sequences and conversion:</p>
<div class="codehilite"><pre><span></span><span class="k">for</span> <span class="n">record</span> <span class="k">in</span> <span class="n">open</span><span class="p">(</span><span class="n">FASTAReader</span><span class="p">{</span><span class="n">ReferenceSequence</span><span class="p">},</span> <span class="s">&quot;hg38.fa&quot;</span><span class="p">)</span>
    <span class="c"># do something</span>
<span class="k">end</span>
</pre></div>


<p><a id='Sequence-search-1'></a></p>
<h2 id="sequence-search">Sequence search</h2>
<p>Three kinds of on-line search functions are provided:</p>
<ol>
<li>Exact search</li>
<li>Approximate search</li>
<li>Regular expression search</li>
</ol>
<p>These are all specialized for biological sequences and ambiguities of symbols are considered.</p>
<p><a id='Exact-search-1'></a></p>
<h3 id="exact-search">Exact search</h3>
<p>Exact search functions search for an occurrence of the query symbol or sequence. Four functions, <code>search</code>, <code>searchindex</code>, <code>rsearch</code>, and <code>rsearchindex</code> are available:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;ACAGCGTAGCT&quot;</span><span class="p">;</span>

<span class="gp">julia&gt;</span> <span class="n">search</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">DNA_G</span><span class="p">)</span>  <span class="c"># search a query symbol</span>
<span class="go">4:4</span>

<span class="gp">julia&gt;</span> <span class="n">query</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;AGC&quot;</span><span class="p">;</span>

<span class="gp">julia&gt;</span> <span class="n">search</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>  <span class="c"># search a query sequence</span>
<span class="go">3:5</span>

<span class="gp">julia&gt;</span> <span class="n">searchindex</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="go">3</span>

<span class="gp">julia&gt;</span> <span class="n">rsearch</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>  <span class="c"># similar to `search` but in the reverse direction</span>
<span class="go">8:10</span>

<span class="gp">julia&gt;</span> <span class="n">rsearchindex</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>  <span class="c"># similar to `searchindex` but in the reverse direction</span>
<span class="go">8</span>
</pre></div>


<p>These search functions take ambiguous symbols into account. That is, if two symbols are compatible (e.g. <code>DNA_A</code> and <code>DNA_N</code>), they match when searching an occurrence. In the following example, 'N' is a wild card that matches any symbols:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">search</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ACNT&quot;</span><span class="p">,</span> <span class="n">DNA_N</span><span class="p">)</span>  <span class="c"># &#39;A&#39; matches &#39;N&#39;</span>
<span class="go">1:1</span>

<span class="gp">julia&gt;</span> <span class="n">search</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ACNT&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;CGT&quot;</span><span class="p">)</span>  <span class="c"># &#39;N&#39; matches &#39;G&#39;</span>
<span class="go">2:4</span>

<span class="gp">julia&gt;</span> <span class="n">search</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ACGT&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;CNT&quot;</span><span class="p">)</span>  <span class="c"># &#39;G&#39; matches &#39;N&#39;</span>
<span class="go">2:4</span>
</pre></div>


<p>The exact sequence search needs preprocessing phase of query sequence before searching phase. This would be enough fast for most search applications. But when searching a query sequence to large amounts of target sequences, caching the result of preprocessing may save time. The <code>ExactSearchQuery</code> creates such a preprocessed query object and is applicable to the search functions:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">query</span> <span class="o">=</span> <span class="n">ExactSearchQuery</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ATT&quot;</span><span class="p">);</span>

<span class="gp">julia&gt;</span> <span class="n">search</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ATTTATT&quot;</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="go">1:3</span>

<span class="gp">julia&gt;</span> <span class="n">rsearch</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ATTTATT&quot;</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="go">5:7</span>
</pre></div>


<p><a id='Approximate-search-1'></a></p>
<h3 id="approximate-search">Approximate search</h3>
<p>The approximate search is similar to the exact search but allows a specific number of errors. That is, it tries to find a subsequence of the target sequence within a specific <a href="https://en.wikipedia.org/wiki/Levenshtein_distance">Levenshtein distance</a> of the query sequence:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;ACAGCGTAGCT&quot;</span><span class="p">;</span>

<span class="gp">julia&gt;</span> <span class="n">approxsearch</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;AGGG&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>  <span class="c"># nothing matches with no errors</span>
<span class="go">0:-1</span>

<span class="gp">julia&gt;</span> <span class="n">approxsearch</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;AGGG&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>  <span class="c"># seq[3:5] matches with one error</span>
<span class="go">3:6</span>

<span class="gp">julia&gt;</span> <span class="n">approxsearch</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;AGGG&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>  <span class="c"># seq[1:4] matches with two errors</span>
<span class="go">1:4</span>
</pre></div>


<p>Like the exact search functions, four kinds of functions (<code>approxsearch</code>, <code>approxsearchindex</code>, <code>approxrsearch</code>, and <code>approxrsearchindex</code>) are available:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;ACAGCGTAGCT&quot;</span><span class="p">;</span> <span class="n">pat</span> <span class="o">=</span> <span class="n">dna</span><span class="s">&quot;AGGG&quot;</span><span class="p">;</span>

<span class="gp">julia&gt;</span> <span class="n">approxsearch</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">pat</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>        <span class="c"># return the range (forward)</span>
<span class="go">1:4</span>

<span class="gp">julia&gt;</span> <span class="n">approxsearchindex</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">pat</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>   <span class="c"># return the starting index (forward)</span>
<span class="go">1</span>

<span class="gp">julia&gt;</span> <span class="n">approxrsearch</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">pat</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>       <span class="c"># return the range (backward)</span>
<span class="go">8:11</span>

<span class="gp">julia&gt;</span> <span class="n">approxrsearchindex</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">pat</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>  <span class="c"># return the starting index (backward)</span>
<span class="go">8</span>
</pre></div>


<p>Preprocessing can be cached in an <code>ApproximateSearchQuery</code> object:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">query</span> <span class="o">=</span> <span class="n">ApproximateSearchQuery</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;AGGG&quot;</span><span class="p">);</span>

<span class="gp">julia&gt;</span> <span class="n">approxsearch</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;AAGAGG&quot;</span><span class="p">,</span> <span class="n">query</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">2:5</span>

<span class="gp">julia&gt;</span> <span class="n">approxsearch</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ACTACGT&quot;</span><span class="p">,</span> <span class="n">query</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">4:6</span>
</pre></div>


<p><a id='Regular-expression-search-1'></a></p>
<h3 id="regular-expression-search">Regular expression search</h3>
<p>Query patterns can be described in regular expressions. The syntax supports a subset of Perl and PROSITE's notation.</p>
<p>The Perl-like syntax starts with <code>biore</code> (<strong>bio</strong>logical <strong>re</strong>gular expression) and ends with a symbol option: "dna", "rna" or "aa". For example, <code>biore"A+"dna</code> is a regular expression for DNA sequences and <code>biore"A+"aa</code> is for amino acid sequences. The symbol options can be abbreviated to its first character: "d", "r" or "a", respectively.</p>
<p>Here are examples of using the regular expression for <code>BioSequence</code>s:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">match</span><span class="p">(</span><span class="n">biore</span><span class="s">&quot;A+C*&quot;</span><span class="n">dna</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;AAAACC&quot;</span><span class="p">)</span>
<span class="go">Nullable(RegexMatch(&quot;AAAACC&quot;))</span>

<span class="gp">julia&gt;</span> <span class="n">match</span><span class="p">(</span><span class="n">biore</span><span class="s">&quot;A+C*&quot;</span><span class="n">d</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;AAAACC&quot;</span><span class="p">)</span>
<span class="go">Nullable(RegexMatch(&quot;AAAACC&quot;))</span>

<span class="gp">julia&gt;</span> <span class="n">ismatch</span><span class="p">(</span><span class="n">biore</span><span class="s">&quot;A+C*&quot;</span><span class="n">dna</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;AAC&quot;</span><span class="p">)</span>
<span class="go">true</span>

<span class="gp">julia&gt;</span> <span class="n">ismatch</span><span class="p">(</span><span class="n">biore</span><span class="s">&quot;A+C*&quot;</span><span class="n">dna</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;C&quot;</span><span class="p">)</span>
<span class="go">false</span>
</pre></div>


<p><code>match</code> always returns a <code>Nullable</code> object and it should be null if no match is found.</p>
<p>The table below summarizes available syntax elements.</p>
<table>
<thead>
<tr>
<th align="center">Syntax</th>
<th align="left">Description</th>
<th align="left">Example</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><code>\|</code></td>
<td align="left">alternation</td>
<td align="left"><code>"A\|T"</code> matches <code>"A"</code> and <code>"T"</code></td>
</tr>
<tr>
<td align="center"><code>*</code></td>
<td align="left">zero or more times repeat</td>
<td align="left"><code>"TA*"</code> matches <code>"T"</code>, <code>"TA"</code> and <code>"TAA"</code></td>
</tr>
<tr>
<td align="center"><code>+</code></td>
<td align="left">one or more times repeat</td>
<td align="left"><code>"TA+"</code> matches <code>"TA"</code> and <code>"TAA"</code></td>
</tr>
<tr>
<td align="center"><code>?</code></td>
<td align="left">zero or one time</td>
<td align="left"><code>"TA?"</code> matches <code>"T"</code> and <code>"TA"</code></td>
</tr>
<tr>
<td align="center"><code>{n,}</code></td>
<td align="left"><code>n</code> or more times repeat</td>
<td align="left"><code>"A{3,}"</code> matches <code>"AAA"</code> and <code>"AAAA"</code></td>
</tr>
<tr>
<td align="center"><code>{n,m}</code></td>
<td align="left"><code>n</code>-<code>m</code> times repeat</td>
<td align="left"><code>"A{3,5}"</code> matches <code>"AAA"</code>, <code>"AAAA"</code> and <code>"AAAAA"</code></td>
</tr>
<tr>
<td align="center"><code>^</code></td>
<td align="left">the start of the sequence</td>
<td align="left"><code>"^TAN*"</code> matches <code>"TATGT"</code></td>
</tr>
<tr>
<td align="center"><code>$</code></td>
<td align="left">the end of the sequence</td>
<td align="left"><code>"N*TA$"</code> matches <code>"GCTA"</code></td>
</tr>
<tr>
<td align="center"><code>(...)</code></td>
<td align="left">pattern grouping</td>
<td align="left"><code>"(TA)+"</code> matches <code>"TA"</code> and <code>"TATA"</code></td>
</tr>
<tr>
<td align="center"><code>[...]</code></td>
<td align="left">one of symbols</td>
<td align="left"><code>"[ACG]+"</code> matches <code>"AGGC"</code></td>
</tr>
</tbody>
</table>
<p><code>eachmatch</code>, <code>matchall</code>, and <code>search</code> are also defined like usual strings:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">matchall</span><span class="p">(</span><span class="n">biore</span><span class="s">&quot;TATA*?&quot;</span><span class="n">d</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TATTATAATTA&quot;</span><span class="p">)</span>  <span class="c"># overlap (default)</span>
<span class="go">4-element Array{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}},1}:</span>
<span class="go"> 3nt DNA Sequence:</span>
<span class="go">TAT</span>
<span class="go"> 3nt DNA Sequence:</span>
<span class="go">TAT</span>
<span class="go"> 4nt DNA Sequence:</span>
<span class="go">TATA</span>
<span class="go"> 5nt DNA Sequence:</span>
<span class="go">TATAA</span>

<span class="gp">julia&gt;</span> <span class="n">matchall</span><span class="p">(</span><span class="n">biore</span><span class="s">&quot;TATA*&quot;</span><span class="n">d</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TATTATAATTA&quot;</span><span class="p">,</span> <span class="n">false</span><span class="p">)</span>  <span class="c"># no overlap</span>
<span class="go">2-element Array{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}},1}:</span>
<span class="go"> 3nt DNA Sequence:</span>
<span class="go">TAT</span>
<span class="go"> 5nt DNA Sequence:</span>
<span class="go">TATAA</span>

<span class="gp">julia&gt;</span> <span class="n">search</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;TATTATAATTA&quot;</span><span class="p">,</span> <span class="n">biore</span><span class="s">&quot;TATA*&quot;</span><span class="n">d</span><span class="p">)</span>
<span class="go">1:3</span>

<span class="gp">julia&gt;</span> <span class="n">search</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;TATTATAATTA&quot;</span><span class="p">,</span> <span class="n">biore</span><span class="s">&quot;TATA*&quot;</span><span class="n">d</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">4:8</span>
</pre></div>


<p>Notewothy differences from strings are:</p>
<ul>
<li>Ambiguous characters match any compatible characters (e.g. <code>biore"N"d</code> is equivalent to <code>biore"[ACGT]"d</code>).</li>
<li>Whitespaces are ignored (e.g. <code>biore"A C G"d</code> is equivalent to <code>biore"ACG"d</code>).</li>
</ul>
<p>The PROSITE notation is described in <a href="http://prosite.expasy.org/scanprosite/scanprosite_doc.html">ScanProsite - user manual</a>. The syntax supports almost all notations including the extended syntax. The PROSITE notation starts with <code>prosite</code> prefix and no symbol option is needed because it always descirbe patterns of amino acid sequences:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">match</span><span class="p">(</span><span class="n">prosite</span><span class="s">&quot;[AC]-x-V-x(4)-{ED}&quot;</span><span class="p">,</span> <span class="n">aa</span><span class="s">&quot;CPVPQARG&quot;</span><span class="p">)</span>
<span class="go">Nullable(RegexMatch(&quot;CPVPQARG&quot;))</span>

<span class="gp">julia&gt;</span> <span class="n">match</span><span class="p">(</span><span class="n">prosite</span><span class="s">&quot;[AC]xVx(4){ED}&quot;</span><span class="p">,</span> <span class="n">aa</span><span class="s">&quot;CPVPQARG&quot;</span><span class="p">)</span>
<span class="go">Nullable(RegexMatch(&quot;CPVPQARG&quot;))</span>
</pre></div>


<p><a id='Sequence-composition-1'></a></p>
<h2 id="sequence-composition">Sequence composition</h2>
<p>Sequence composition can be easily calculated using the <code>composition</code> function:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">comp</span> <span class="o">=</span> <span class="n">composition</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;ACGAG&quot;</span><span class="p">)</span>
<span class="go">DNA Nucleotide Composition:</span>
<span class="go">  DNA_A   =&gt; 2</span>
<span class="go">  DNA_C   =&gt; 1</span>
<span class="go">  DNA_G   =&gt; 2</span>
<span class="go">  DNA_T   =&gt; 0</span>
<span class="go">  DNA_M   =&gt; 0</span>
<span class="go">  DNA_R   =&gt; 0</span>
<span class="go">  DNA_W   =&gt; 0</span>
<span class="go">  DNA_S   =&gt; 0</span>
<span class="go">  DNA_Y   =&gt; 0</span>
<span class="go">  DNA_K   =&gt; 0</span>
<span class="go">  DNA_V   =&gt; 0</span>
<span class="go">  DNA_H   =&gt; 0</span>
<span class="go">  DNA_D   =&gt; 0</span>
<span class="go">  DNA_B   =&gt; 0</span>
<span class="go">  DNA_N   =&gt; 0</span>
<span class="go">  DNA_Gap =&gt; 0</span>

<span class="gp">julia&gt;</span> <span class="n">comp</span><span class="p">[</span><span class="n">DNA_A</span><span class="p">]</span>
<span class="go">2</span>

<span class="gp">julia&gt;</span> <span class="n">comp</span><span class="p">[</span><span class="n">DNA_T</span><span class="p">]</span>
<span class="go">0</span>
</pre></div>


<p>To accumulate composition statistics of multiple sequences, <code>merge!</code> can be used as follows:</p>
<div class="codehilite"><pre><span></span><span class="c"># initiaize an empty composition counter</span>
<span class="n">comp</span> <span class="o">=</span> <span class="n">composition</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;&quot;</span><span class="p">)</span>

<span class="c"># iterate over sequences and accumulate composition statistics into `comp`</span>
<span class="k">for</span> <span class="n">seq</span> <span class="k">in</span> <span class="n">seqs</span>
    <span class="n">merge!</span><span class="p">(</span><span class="n">comp</span><span class="p">,</span> <span class="n">composition</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
<span class="k">end</span>

<span class="c"># or functional programming style in one line</span>
<span class="n">foldl</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">merge</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">composition</span><span class="p">(</span><span class="n">y</span><span class="p">)),</span> <span class="n">composition</span><span class="p">(</span><span class="n">dna</span><span class="s">&quot;&quot;</span><span class="p">),</span> <span class="n">seqs</span><span class="p">)</span>
</pre></div>


<p><code>composition</code> is also applicable to a <em>k</em>-mer iterator:</p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">comp</span> <span class="o">=</span> <span class="n">composition</span><span class="p">(</span><span class="n">each</span><span class="p">(</span><span class="n">DNAKmer</span><span class="p">{</span><span class="mi">4</span><span class="p">},</span> <span class="n">dna</span><span class="s">&quot;ACGT&quot;</span><span class="o">^</span><span class="mi">100</span><span class="p">));</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">comp</span><span class="p">[</span><span class="n">DNAKmer</span><span class="p">(</span><span class="s">&quot;ACGT&quot;</span><span class="p">)]</span>
<span class="mi">100</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">comp</span><span class="p">[</span><span class="n">DNAKmer</span><span class="p">(</span><span class="s">&quot;CGTA&quot;</span><span class="p">)]</span>
<span class="mi">99</span>
</pre></div>


<p><a id='Sequence-records-1'></a></p>
<h2 id="sequence-records">Sequence records</h2>
<p>The <code>SeqRecord</code> type is used to represent a named sequence, optionally with accompanying metadata. It is defined as: sequence. It is as:</p>
<div class="codehilite"><pre><span></span><span class="k">type</span><span class="nc"> SeqRecord</span><span class="p">{</span><span class="n">S</span><span class="p">,</span> <span class="n">T</span><span class="p">}</span>
    <span class="n">name</span><span class="p">::</span><span class="n">String</span>
    <span class="n">seq</span><span class="p">::</span><span class="n">S</span>
    <span class="n">metadata</span><span class="p">::</span><span class="n">T</span>
<span class="k">end</span>
</pre></div>


<p>The type of the <code>metadata</code> field depends on the source of the sequence record. For example, if a record is read from a FASTA file, metadata contains the description field. If from a FASTQ file, a quality scores assigned to base calls during sequencing.</p>
<p>The following accessors are defined for the <code>SeqRecord</code> type:</p>
<p><a id='Bio.Seq.seqname' href='#Bio.Seq.seqname'>#</a>
<strong><code>Bio.Seq.seqname</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>seqname(record)
</pre></div>


<p>Return the sequence name of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/seqrecord.jl#L27-L31' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.sequence' href='#Bio.Seq.sequence'>#</a>
<strong><code>Bio.Seq.sequence</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>sequence(record)
</pre></div>


<p>Return the sequence of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/seqrecord.jl#L36-L40' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Seq.metadata' href='#Bio.Seq.metadata'>#</a>
<strong><code>Bio.Seq.metadata</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>metadata(record)
</pre></div>


<p>Return the metadata of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/87bde7925704485c7306b1a8b7a82011a6706908/src/seq/seqrecord.jl#L45-L49' class='documenter-source'>source</a><br></p>
<p><a id='Sequence-demultiplexing-1'></a></p>
<h2 id="sequence-demultiplexing">Sequence demultiplexing</h2>
<p>Multiplex sequencing is a technology to sequence mutliple samples at the same time on a high-throughput DNA sequencer. Samples are distinguished by the short prefix of a DNA sequence called DNA barcode. The <code>Bio.Seq</code> offers the <code>Demultiplexer</code> type and the <code>demultiplex</code> function to identify the DNA barcode of a longer DNA sequence allowing small errors.</p>
<p>In the following example, four kinds of DNA sequences of length 4 are used as DNA barcodes. <code>Demultiplexer</code> takes these barcodes as its first argument with a few options:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">barcodes</span> <span class="o">=</span> <span class="n">DNASequence</span><span class="p">[</span><span class="s">&quot;ATGG&quot;</span><span class="p">,</span> <span class="s">&quot;CAGA&quot;</span><span class="p">,</span> <span class="s">&quot;GGAA&quot;</span><span class="p">,</span> <span class="s">&quot;TACG&quot;</span><span class="p">];</span>

<span class="gp">julia&gt;</span> <span class="n">dplxr</span> <span class="o">=</span> <span class="n">Demultiplexer</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">n_max_errors</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="p">:</span><span class="n">hamming</span><span class="p">)</span>
<span class="go">Bio.Seq.Demultiplexer{Bio.Seq.BioSequence{Bio.Seq.DNAAlphabet{4}}}:</span>
<span class="go">  distance: hamming</span>
<span class="go">  number of barcodes: 4</span>
<span class="go">  number of correctable errors: 1</span>
</pre></div>


<p><code>n_max_errors</code> specifies the number of maximum correctable errors in a barcode. The type of correctable errors depends on the <code>distance</code> parameter. When <code>distance = :hamming</code> as shown above only substitutions are correctable. When <code>distance = :levenshtein</code> substitutions, deletions, and insertions are correctable. The user is responsible for keeping enough distances among barcodes; <code>Demultiplexer</code> will throw an exception if two barcodes are within <code>n_max_errors * 2</code>.</p>
<p>The <code>demultiplex</code> function takes a demultiplexer object and a DNA sequence, and returns a tuple of a barcode index and a distance between the original barcode sequence and the prefix sequence:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">demultiplex</span><span class="p">(</span><span class="n">dplxr</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;ATGGCGNT&quot;</span><span class="p">)</span>  <span class="c"># 1st barcode with no errors</span>
<span class="go">(1,0)</span>

<span class="gp">julia&gt;</span> <span class="n">demultiplex</span><span class="p">(</span><span class="n">dplxr</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;CAGGCGNT&quot;</span><span class="p">)</span>  <span class="c"># 2nd barcode with one error</span>
<span class="go">(2,1)</span>

<span class="gp">julia&gt;</span> <span class="n">demultiplex</span><span class="p">(</span><span class="n">dplxr</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;GGAACGNT&quot;</span><span class="p">)</span>  <span class="c"># 3rd barcode with no errors</span>
<span class="go">(3,0)</span>

<span class="gp">julia&gt;</span> <span class="n">demultiplex</span><span class="p">(</span><span class="n">dplxr</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TGACCGNT&quot;</span><span class="p">)</span>  <span class="c"># no matching barcode</span>
<span class="go">(0,-1)</span>
</pre></div>


<p>The optional third argument controls the search strategy. <code>demultiplex</code> uses an index to search the closest barcode within <code>n_max_errors</code> in the barcode set and returns it if any by default. If the third argument is <code>true</code> it falls back to a linear search after the index search and returns one of the closest barcodes at random. The next example shows the difference of these two strategies:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">demultiplex</span><span class="p">(</span><span class="n">dplxr</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TGACCGNT&quot;</span><span class="p">,</span> <span class="n">false</span><span class="p">)</span>  <span class="c"># linear search off (default)</span>
<span class="go">(0,-1)</span>

<span class="gp">julia&gt;</span> <span class="n">demultiplex</span><span class="p">(</span><span class="n">dplxr</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;TGACCGNT&quot;</span><span class="p">,</span> <span class="n">true</span><span class="p">)</span>   <span class="c"># linear search on</span>
<span class="go">(3,2)</span>
</pre></div></div>
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Copyright (c) 2016 The BioJulia Organisation<br></small>
        
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../../js/jquery-1.10.2.min.js"></script>
    <script src="../../js/bootstrap-3.0.3.min.js"></script>
    <script src="../../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '../..';
    </script>
    <script data-main="../../mkdocs/js/search.js" src="../../mkdocs/js/require.js"></script>
    <script src="../../js/base.js"></script>
    <script>
        ((window.gitter = {}).chat = {}).options = {
            room: 'BioJulia/Bio.jl'
        };
    </script>
    <script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <script src="../../assets/mathjaxhelper.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
