<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0"> 
    <meta name="author" content="The BioJulia Organisation"> 
    <link rel="canonical" href="https://biojulia.github.io/Bio.jl/man/var/"> 
    <link rel="shortcut icon" href="../../img/favicon.ico">

    <title>Genetic Variation - Bio.jl</title>

    <link href="../../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../../css/base.css" rel="stylesheet">
    <link href="../../css/biojulia.css" rel="stylesheet">
    <link rel="stylesheet" href="../../css/highlight.css">
    <link rel="stylesheet" href="../../css/friendly.css" type="text/css">
    <link href="../../assets/Documenter.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="../../home/">Bio.jl</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../../home/">Home</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Manual <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../frontpage/">Front Page</a>
</li>

                        
                            
<li >
    <a href="../install/">Installation</a>
</li>

                        
                            
<li >
    <a href="../reading/">IO API</a>
</li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Biological Sequences</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../seq/module/">Overview</a>
</li>

        
            
<li >
    <a href="../seq/symbols/">Biological Symbols</a>
</li>

        
            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Sequence types</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../seq/sequences/sequences/">Overview</a>
</li>

        
            
<li >
    <a href="../seq/sequences/bioseq/">BioSequence</a>
</li>

        
            
<li >
    <a href="../seq/sequences/refseq/">Reference Sequences</a>
</li>

        
            
<li >
    <a href="../seq/sequences/kmer/">Nucleic acid k-mers</a>
</li>

        
    </ul>
  </li>

        
            
<li >
    <a href="../seq/search/">Searching</a>
</li>

        
            
<li >
    <a href="../seq/seqrecords/">Sequence Records</a>
</li>

        
            
<li >
    <a href="../seq/demultiplexer/">Demultiplexing</a>
</li>

        
    </ul>
  </li>

                        
                            
<li >
    <a href="../alignments/">Alignments</a>
</li>

                        
                            
<li >
    <a href="../intervals/">Intervals</a>
</li>

                        
                            
<li class="active">
    <a href="./">Genetic Variation</a>
</li>

                        
                            
<li >
    <a href="../phylo/">Phylogenetic Trees</a>
</li>

                        
                            
<li >
    <a href="../structure/">Structure</a>
</li>

                        
                            
<li >
    <a href="../services/">Services</a>
</li>

                        
                            
<li >
    <a href="../util/">Util</a>
</li>

                        
                            
<li >
    <a href="../tools/">Tools</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Reference <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../../lib/seq/">Seq</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li >
                        <a href="../contributing/">Contributing</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../intervals/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../phylo/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/BioJulia/Bio.jl">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
                <li><a href="https://biojulia.github.io">BioJulia Home</a></li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="first-level active"><a href="#biovar-biological-variation">Bio.Var: Biological Variation.</a></li>
        
            <li class="second-level"><a href="#identifying-and-counting-sequence-sites">Identifying and counting sequence sites</a></li>
            
                <li class="third-level"><a href="#different-types-of-site">Different types of site</a></li>
            
                <li class="third-level"><a href="#the-count_sites-method">The count_sites method</a></li>
            
        
            <li class="second-level"><a href="#file-formats-for-representing-genetic-variation">File formats for representing genetic variation</a></li>
            
                <li class="third-level"><a href="#vcf-and-bcf-file-formats">VCF and BCF file formats</a></li>
            
        
    
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<p><a id='Bio.Var:-Biological-Variation.-1'></a></p>
<h1 id="biovar-biological-variation">Bio.Var: Biological Variation.</h1>
<p><a id='Identifying-and-counting-sequence-sites-1'></a></p>
<h2 id="identifying-and-counting-sequence-sites">Identifying and counting sequence sites</h2>
<p>You can identify and count the number of various types of site in a nucleotide sequence, or pair/set of aligned nucleotide sequences.</p>
<p><a id='Different-types-of-site-1'></a></p>
<h3 id="different-types-of-site">Different types of site</h3>
<p><a id='The-abstract-Site-types-1'></a></p>
<h4 id="the-abstract-site-types">The abstract Site types</h4>
<div class="codehilite"><pre><span></span>Site
Mutation
</pre></div>


<p><a id='The-concrete-types-of-site-case-1'></a></p>
<h4 id="the-concrete-types-of-site-case">The concrete types of site case</h4>
<p>The types of mutations that can currently be counted are summarized below:</p>
<p><a id='Bio.Var.Gap' href='#Bio.Var.Gap'>#</a>
<strong><code>Bio.Var.Gap</code></strong> &mdash; <em>Type</em>.</p>
<p>An <code>Gap</code> site describes a site where either of two aligned sites are a gap symbol '-'.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L22-L25' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.Ambiguous' href='#Bio.Var.Ambiguous'>#</a>
<strong><code>Bio.Var.Ambiguous</code></strong> &mdash; <em>Type</em>.</p>
<p>An <code>Ambiguous</code> site describes a site where either of two aligned sites are an ambiguity symbol.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L16-L19' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.Certain' href='#Bio.Var.Certain'>#</a>
<strong><code>Bio.Var.Certain</code></strong> &mdash; <em>Type</em>.</p>
<p>A <code>Certain</code> site describes a site where both of two aligned sites are not an ambiguity symbol or a gap.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L10-L13' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.Match' href='#Bio.Var.Match'>#</a>
<strong><code>Bio.Var.Match</code></strong> &mdash; <em>Type</em>.</p>
<p>A <code>Match</code> site describes a site where two aligned nucleotides are the same biological symbol.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L28-L31' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.Mismatch' href='#Bio.Var.Mismatch'>#</a>
<strong><code>Bio.Var.Mismatch</code></strong> &mdash; <em>Type</em>.</p>
<p>A <code>Mismatch</code> site describes a site where two aligned nucleotides are not the same biological symbol.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L34-L37' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.Conserved' href='#Bio.Var.Conserved'>#</a>
<strong><code>Bio.Var.Conserved</code></strong> &mdash; <em>Type</em>.</p>
<p>A <code>Mismatch</code> site describes a site where two aligned nucleotides are definately conserved. By definately conserved this means that the symbols of the site are non-ambiguity symbols, and they are the same symbol.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L40-L44' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.Mutated' href='#Bio.Var.Mutated'>#</a>
<strong><code>Bio.Var.Mutated</code></strong> &mdash; <em>Type</em>.</p>
<p>A <code>Mutated</code> site describes a site where two aligned nucleotides are definately mutated. By definately mutated this means that the symbols of the site are non-ambiguity symbols, and they are not the same symbol.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L47-L51' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.Transition' href='#Bio.Var.Transition'>#</a>
<strong><code>Bio.Var.Transition</code></strong> &mdash; <em>Type</em>.</p>
<p>A <code>Transition</code> site describes a site where two aligned nucleotides are definately mutated, and the type of mutation is a transition mutation. In other words, the symbols must not be ambiguity symbols, and they must be different such that they constitute a transition mutation: i.e. A&lt;-&gt;G, or C&lt;-&gt;T.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L54-L59' class='documenter-source'>source</a><br></p>
<p><a id='Bio.Var.Transversion' href='#Bio.Var.Transversion'>#</a>
<strong><code>Bio.Var.Transversion</code></strong> &mdash; <em>Type</em>.</p>
<p>A <code>Transversion</code> site describes a site where two aligned nucleotides are definately mutated, and the type of mutation is a transversion mutation. In other words, the symbols must not be ambiguity symbols, and they must be different such that they constitute a transversion mutation: i.e. A&lt;-&gt;C, A&lt;-&gt;T, G&lt;-&gt;T, G&lt;-&gt;C.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_types.jl#L62-L68' class='documenter-source'>source</a><br></p>
<p><a id='The-count_sites-method-1'></a></p>
<h3 id="the-count_sites-method">The <code>count_sites</code> method</h3>
<p><a id='Bio.Var.count_sites' href='#Bio.Var.count_sites'>#</a>
<strong><code>Bio.Var.count_sites</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_sites{T&lt;:Site}(::Type{T}, a::BioSequence, b::BioSequence)
</pre></div>


<p>Mutations are counted using the <code>count_sites</code> method.</p>
<p>If the concrete site case type inherits from NoPairDel, then the result will be a simple integer value of the number of counts of site.</p>
<p>If the concrete site case type inherits from PairDel, then the result will be a tuple. As a result, counting methods for such types also count and return the number of sites which could not be unambiguously determined. This second count is important for some downstream purposes, for example evolutionary/genetic distance computations in which pairwise deletion of ambiguous sites is necessary.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_counting.jl#L17-L31' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>count_sites{T&lt;:NoPairDel}(::Type{T}, a::EachWindowIterator, b::EachWindowIterator)
</pre></div>


<p>A method of the count_sites function for counting sites of SiteType <code>T</code> between two aligned sequences.</p>
<p>Crucially, this method accepts two EachWindowIterators as parameters <code>a</code> and <code>b</code> instead of two <code>BioSequence</code>s. As a result, the two sequences are iterated over with a sliding window, and a site count computed for each window.</p>
<p>This function returns two vectors, one with the ranges</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_counting.jl#L36-L47' class='documenter-source'>source</a><br></p>
<p>Count the number of sites between two nucleotide sequences using a sliding window.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_counting.jl#L76' class='documenter-source'>source</a><br></p>
<p>Count the number of mutations between nucleotide sequences in a pairwise manner.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_counting.jl#L83-L85' class='documenter-source'>source</a><br></p>
<p>Count the number of mutations between nucleotide sequences in a pairwise manner, using a sliding window of a given width and step.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_counting.jl#L95-L98' class='documenter-source'>source</a><br></p>
<p>Count the number of mutations between nucleotide sequences in a pairwise manner.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_counting.jl#L108-L110' class='documenter-source'>source</a><br></p>
<p>Count the number of mutations between nucleotide sequences in a pairwise manner, using a sliding window of a given width and step.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_counting.jl#L121-L124' class='documenter-source'>source</a><br></p>
<p>Count the number of sites of a given SiteType <code>T</code> in a biological sequence.</p>
<p><a target='_blank' href='https://github.com/BioJulia/Bio.jl/tree/bd9c128f4ec2dc41d34844716dd81163b8cf103b/src/var/site_counting/site_counting.jl#L135' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span><span class="n">count_sites</span><span class="p">(</span><span class="n">Match</span><span class="p">,</span> <span class="p">[</span><span class="n">dna</span><span class="s">&quot;ATCGATCG&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;ACCGATCG&quot;</span><span class="p">])</span>

<span class="n">count_sites</span><span class="p">(</span><span class="n">Mismatch</span><span class="p">,</span> <span class="p">[</span><span class="n">dna</span><span class="s">&quot;ATCGATCG&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;ACCGATCG&quot;</span><span class="p">])</span>

<span class="n">count_sites</span><span class="p">(</span><span class="n">Conserved</span><span class="p">,</span> <span class="p">[</span><span class="n">dna</span><span class="s">&quot;ATCGATCG&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;ACCGATCG&quot;</span><span class="p">])</span>

<span class="n">count_sites</span><span class="p">(</span><span class="n">Mutated</span><span class="p">,</span> <span class="p">[</span><span class="n">dna</span><span class="s">&quot;ATCGATCG&quot;</span><span class="p">,</span> <span class="n">dna</span><span class="s">&quot;ACCGATCG&quot;</span><span class="p">])</span>

<span class="n">count_sites</span><span class="p">(</span><span class="n">Transition</span><span class="p">,</span> <span class="p">[</span><span class="n">rna</span><span class="s">&quot;AUCGAUCG&quot;</span><span class="p">,</span> <span class="n">rna</span><span class="s">&quot;ACCGAUCG&quot;</span><span class="p">])</span>

<span class="n">count_sites</span><span class="p">(</span><span class="n">Transversion</span><span class="p">,</span> <span class="p">[</span><span class="n">rna</span><span class="s">&quot;AUCGAUCG&quot;</span><span class="p">,</span> <span class="n">rna</span><span class="s">&quot;ACCGAUCG&quot;</span><span class="p">])</span>
</pre></div>


<p><a id='File-formats-for-representing-genetic-variation-1'></a></p>
<h2 id="file-formats-for-representing-genetic-variation">File formats for representing genetic variation</h2>
<p>This module supports some common file formats to read and write genetic variations.</p>
<p><a id='VCF-and-BCF-file-formats-1'></a></p>
<h3 id="vcf-and-bcf-file-formats">VCF and BCF file formats</h3>
<p>VCF and BCF file formats can be read using <code>VCFReader</code> and <code>BCFReader</code>, respectively:</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">VCFReader</span><span class="p">,</span> <span class="s">&quot;example.vcf&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">record</span> <span class="kp">in</span> <span class="n">reader</span>
    <span class="c"># do something</span>
<span class="k">end</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p>A reader first reads the header section of a file and creates a <code>VCFHeader</code> object. The <code>header</code> function is provided to access the header object of a reader:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">header</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
<span class="go">Bio.Var.VCFHeader:</span>
<span class="go">  metainfo tags: fileformat fileDate source reference contig phasing INFO FILTER FORMAT</span>
<span class="go">     sample IDs: NA00001 NA00002 NA00003</span>

<span class="gp">julia&gt;</span> <span class="n">find</span><span class="p">(</span><span class="n">header</span><span class="p">(</span><span class="n">reader</span><span class="p">),</span> <span class="s">&quot;FORMAT&quot;</span><span class="p">)</span>
<span class="go">4-element Array{Bio.Var.VCFMetaInfo,1}:</span>
<span class="go"> Bio.Var.VCFMetaInfo:</span>
<span class="go">    tag: FORMAT</span>
<span class="go">  value: ID=&quot;GT&quot; Number=&quot;1&quot; Type=&quot;String&quot; Description=&quot;Genotype&quot;</span>
<span class="go"> Bio.Var.VCFMetaInfo:</span>
<span class="go">    tag: FORMAT</span>
<span class="go">  value: ID=&quot;GQ&quot; Number=&quot;1&quot; Type=&quot;Integer&quot; Description=&quot;Genotype Quality&quot;</span>
<span class="go"> Bio.Var.VCFMetaInfo:</span>
<span class="go">    tag: FORMAT</span>
<span class="go">  value: ID=&quot;DP&quot; Number=&quot;1&quot; Type=&quot;Integer&quot; Description=&quot;Read Depth&quot;</span>
<span class="go"> Bio.Var.VCFMetaInfo:</span>
<span class="go">    tag: FORMAT</span>
<span class="go">  value: ID=&quot;HQ&quot; Number=&quot;2&quot; Type=&quot;Integer&quot; Description=&quot;Haplotype Quality&quot;</span>
</pre></div>


<p><code>VCFMetaInfo</code> objects in the header support the following accessor functions:</p>
<table>
<thead>
<tr>
<th align="left">Accessor</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>metainfotag</code></td>
<td align="left">tag string</td>
</tr>
<tr>
<td align="left"><code>metainfoval</code></td>
<td align="left">value string</td>
</tr>
<tr>
<td align="left"><code>keys</code></td>
<td align="left">keys of fields between '&lt;' and '&gt;'</td>
</tr>
<tr>
<td align="left"><code>values</code></td>
<td align="left">values of fields between '&lt;' and '&gt;'</td>
</tr>
<tr>
<td align="left"><code>[&lt;key&gt;]</code></td>
<td align="left">value of a field with <code>key</code></td>
</tr>
</tbody>
</table>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">metainfo</span> <span class="o">=</span> <span class="n">VCFMetaInfo</span><span class="p">(</span><span class="s">&quot;##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=</span><span class="se">\&quot;</span><span class="s">Genotype</span><span class="se">\&quot;</span><span class="s">&gt;&quot;</span><span class="p">)</span>
<span class="go">Bio.Var.VCFMetaInfo:</span>
<span class="go">    tag: FORMAT</span>
<span class="go">  value: ID=&quot;GT&quot; Number=&quot;1&quot; Type=&quot;String&quot; Description=&quot;Genotype&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">metainfotag</span><span class="p">(</span><span class="n">metainfo</span><span class="p">)</span>
<span class="go">&quot;FORMAT&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">metainfoval</span><span class="p">(</span><span class="n">metainfo</span><span class="p">)</span>
<span class="go">&quot;&lt;ID=GT,Number=1,Type=String,Description=\&quot;Genotype\&quot;&gt;&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">keys</span><span class="p">(</span><span class="n">metainfo</span><span class="p">)</span>
<span class="go">4-element Array{String,1}:</span>
<span class="go"> &quot;ID&quot;</span>
<span class="go"> &quot;Number&quot;</span>
<span class="go"> &quot;Type&quot;</span>
<span class="go"> &quot;Description&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">metainfo</span><span class="p">[</span><span class="s">&quot;ID&quot;</span><span class="p">]</span>
<span class="go">&quot;GT&quot;</span>
</pre></div>


<p><code>VCFRecord</code> and <code>BCFRecord</code> objects support the following accessor functions (see the docstring of each accessor for the details):</p>
<table>
<thead>
<tr>
<th align="left">Accessor</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>chromosome</code></td>
<td align="left">chromosome name</td>
</tr>
<tr>
<td align="left"><code>leftposition</code></td>
<td align="left">reference position</td>
</tr>
<tr>
<td align="left"><code>identifier</code></td>
<td align="left">unique identifiers</td>
</tr>
<tr>
<td align="left"><code>reference</code></td>
<td align="left">reference bases</td>
</tr>
<tr>
<td align="left"><code>alternate</code></td>
<td align="left">alternate bases</td>
</tr>
<tr>
<td align="left"><code>quality</code></td>
<td align="left">Phred-scaled quality score</td>
</tr>
<tr>
<td align="left"><code>filter_</code></td>
<td align="left">filter status</td>
</tr>
<tr>
<td align="left"><code>information</code></td>
<td align="left">additional information</td>
</tr>
<tr>
<td align="left"><code>infokeys</code></td>
<td align="left">keys of additional information</td>
</tr>
<tr>
<td align="left"><code>format</code></td>
<td align="left">genotype format</td>
</tr>
<tr>
<td align="left"><code>genotype</code></td>
<td align="left">genotype information</td>
</tr>
</tbody>
</table>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">record</span> <span class="o">=</span> <span class="n">VCFRecord</span><span class="p">(</span><span class="s">&quot;20</span><span class="se">\t</span><span class="s">14370</span><span class="se">\t</span><span class="s">rs6054257</span><span class="se">\t</span><span class="s">G</span><span class="se">\t</span><span class="s">A</span><span class="se">\t</span><span class="s">29</span><span class="se">\t</span><span class="s">PASS</span><span class="se">\t</span><span class="s">NS=3;DP=14;AF=0.5;DB;H2</span><span class="se">\t</span><span class="s">GT:GQ:DP:HQ</span><span class="se">\t</span><span class="s">0|0:48:1:51,51</span><span class="se">\t</span><span class="s">1|0:48:8:51,51&quot;</span><span class="p">)</span>

<span class="go">Bio.Var.VCFRecord:</span>
<span class="go">   chromosome: 20</span>
<span class="go">     position: 14370</span>
<span class="go">   identifier: rs6054257</span>
<span class="go">    reference: G</span>
<span class="go">    alternate: A</span>
<span class="go">      quality: 29.0</span>
<span class="go">       filter: PASS</span>
<span class="go">  information: NS=3 DP=14 AF=0.5 DB H2</span>
<span class="go">       format: GT GQ DP HQ</span>
<span class="go">     genotype: [1] 0|0 48 1 51,51 [2] 1|0 48 8 51,51</span>

<span class="gp">julia&gt;</span> <span class="n">chromosome</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">Nullable{String}(&quot;20&quot;)</span>

<span class="gp">julia&gt;</span> <span class="n">leftposition</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">Nullable{Int64}(14370)</span>

<span class="gp">julia&gt;</span> <span class="n">identifier</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">1-element Array{String,1}:</span>
<span class="go"> &quot;rs6054257&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">reference</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">Nullable{String}(&quot;G&quot;)</span>

<span class="gp">julia&gt;</span> <span class="n">alternate</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">1-element Array{String,1}:</span>
<span class="go"> &quot;A&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">quality</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">Nullable{Float64}(29.0)</span>

<span class="gp">julia&gt;</span> <span class="n">filter_</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">1-element Array{String,1}:</span>
<span class="go"> &quot;PASS&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">information</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">5-element Array{Pair{String,String},1}:</span>
<span class="go"> &quot;NS&quot;=&gt;&quot;3&quot;</span>
<span class="go"> &quot;DP&quot;=&gt;&quot;14&quot;</span>
<span class="go"> &quot;AF&quot;=&gt;&quot;0.5&quot;</span>
<span class="go"> &quot;DB&quot;=&gt;&quot;&quot;</span>
<span class="go"> &quot;H2&quot;=&gt;&quot;&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">information</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="s">&quot;AF&quot;</span><span class="p">)</span>
<span class="go">&quot;0.5&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">format</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">4-element Array{String,1}:</span>
<span class="go"> &quot;GT&quot;</span>
<span class="go"> &quot;GQ&quot;</span>
<span class="go"> &quot;DP&quot;</span>
<span class="go"> &quot;HQ&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">genotype</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">2-element Array{Array{String,1},1}:</span>
<span class="go"> String[&quot;0|0&quot;,&quot;48&quot;,&quot;1&quot;,&quot;51,51&quot;]</span>
<span class="go"> String[&quot;1|0&quot;,&quot;48&quot;,&quot;8&quot;,&quot;51,51&quot;]</span>

<span class="gp">julia&gt;</span> <span class="n">genotype</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">4-element Array{String,1}:</span>
<span class="go"> &quot;0|0&quot;</span>
<span class="go"> &quot;48&quot;</span>
<span class="go"> &quot;1&quot;</span>
<span class="go"> &quot;51,51&quot;</span>

<span class="gp">julia&gt;</span> <span class="n">genotype</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="mi">1</span><span class="o">:</span><span class="mi">2</span><span class="p">,</span> <span class="s">&quot;GT&quot;</span><span class="p">)</span>
<span class="go">2-element Array{String,1}:</span>
<span class="go"> &quot;0|0&quot;</span>
<span class="go"> &quot;1|0&quot;</span>
</pre></div></div>
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Copyright (c) 2016 The BioJulia Organisation<br></small>
        
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../../js/jquery-1.10.2.min.js"></script>
    <script src="../../js/bootstrap-3.0.3.min.js"></script>
    <script src="../../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '../..';
    </script>
    <script data-main="../../mkdocs/js/search.js" src="../../mkdocs/js/require.js"></script>
    <script src="../../js/base.js"></script>
    <script>
        ((window.gitter = {}).chat = {}).options = {
            room: 'BioJulia/Bio.jl'
        };
    </script>
    <script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <script src="../../assets/mathjaxhelper.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
